Index: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilter.java
===================================================================
--- lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilter.java	(revision 1458840)
+++ lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilter.java	(working copy)
@@ -24,7 +24,6 @@
 import org.apache.lucene.analysis.tokenattributes.KeywordAttribute;
 import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
 import org.apache.lucene.analysis.util.CharArrayMap;
-import org.apache.lucene.util.Version;
 
 /**
  * Provides the ability to override any {@link KeywordAttribute} aware stemmer
@@ -44,10 +43,10 @@
    * so that they will not be stemmed with stemmers down the chain.
    * </p>
    */
-  public StemmerOverrideFilter(Version matchVersion, TokenStream input,
+  public StemmerOverrideFilter(TokenStream input,
       CharArrayMap<String> dictionary) {
     super(input);
-    this.dictionary = CharArrayMap.copy(matchVersion, dictionary);
+    this.dictionary = dictionary;
   }
 
   @Override
Index: lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilterFactory.java
===================================================================
--- lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilterFactory.java	(revision 1458840)
+++ lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilterFactory.java	(working copy)
@@ -66,6 +66,6 @@
 
   @Override
   public TokenStream create(TokenStream input) {
-    return dictionary == null ? input : new StemmerOverrideFilter(luceneMatchVersion, input, dictionary);
+    return dictionary == null ? input : new StemmerOverrideFilter(input, dictionary);
   }
 }
Index: lucene/CHANGES.txt
===================================================================
--- lucene/CHANGES.txt	(revision 1458840)
+++ lucene/CHANGES.txt	(working copy)
@@ -154,6 +154,9 @@
 * LUCENE-4854: Speed up TermsEnum of FieldCache.getDocTermOrds.
   (Robert Muir)
 
+* LUCENE-4857: Don't unnecessarily copy stem override map in
+  StemmerOverrideFilter. (Simon Willnauer)
+
 ======================= Lucene 4.2.0 =======================
 
 Changes in backwards compatibility policy
