Index: lucene/core/src/test/org/apache/lucene/search/payloads/TestPayloadNearQuery.java
===================================================================
--- lucene/core/src/test/org/apache/lucene/search/payloads/TestPayloadNearQuery.java	(revision 1361309)
+++ lucene/core/src/test/org/apache/lucene/search/payloads/TestPayloadNearQuery.java	(working copy)
@@ -180,6 +180,7 @@
     clauses[0] = q1;
     clauses[1] = q2;
     query = new PayloadNearQuery(clauses, 10, false); 
+    QueryUtils.check(random(), query, searcher);
     //System.out.println(query.toString());
     assertEquals(12, searcher.search(query, null, 100).totalHits);
     /*
