Index: lucene/queries/src/java/org/apache/lucene/queries/CustomScoreQuery.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/queries/src/java/org/apache/lucene/queries/CustomScoreQuery.java	(revision b5b1fc6c1167555f87cd490c16638f0eb025e336)
+++ lucene/queries/src/java/org/apache/lucene/queries/CustomScoreQuery.java	(revision )
@@ -387,6 +387,16 @@
     this.strict = strict;
   }
 
+  /** The sub-query that CustomScoreQuery wraps, affecting both the score and which documents match. */
+  public Query getSubQuery() {
+    return subQuery;
+  }
+
+  /** The scoring queries that only affect the score of CustomScoreQuery. */
+  public Query[] getScoringQueries() {
+    return scoringQueries;
+  }
+
   /**
    * A short name of this query, used in {@link #toString(String)}.
    */
