Index: D:/programming/projects/componence/lucene-dev/src/java/org/apache/lucene/search/BooleanQuery.java
===================================================================
--- D:/programming/projects/componence/lucene-dev/src/java/org/apache/lucene/search/BooleanQuery.java	(revision 345214)
+++ D:/programming/projects/componence/lucene-dev/src/java/org/apache/lucene/search/BooleanQuery.java	(working copy)
@@ -191,6 +191,14 @@
     return (BooleanClause[])clauses.toArray(new BooleanClause[0]);
   }
 
+  /**
+   * Returns count of clauses in this query 
+   * @return count of clauses in this query
+   */
+  public int getClausesCount() {
+    return clauses.size();
+  }
+
   private class BooleanWeight implements Weight {
     protected Similarity similarity;
     protected Vector weights = new Vector();
