Index: D:/dev/lucene/lucene-benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/TrecDocMaker.java
===================================================================
--- D:/dev/lucene/lucene-benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/TrecDocMaker.java	(revision 602663)
+++ D:/dev/lucene/lucene-benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/TrecDocMaker.java	(working copy)
@@ -68,7 +68,10 @@
     super.setConfig(config);
     File workDir = new File(config.get("work.dir","work"));
     String d = config.get("docs.dir","trec");
-    dataDir = new File(workDir,d);
+    dataDir = new File(d);
+    if (!dataDir.isAbsolute()) {
+      dataDir = new File(workDir, d);
+    }
     collectFiles(dataDir,inputFiles);
     if (inputFiles.size()==0) {
       throw new RuntimeException("No txt files in dataDir: "+dataDir.getAbsolutePath());
