Index: IndexFiles.java
===================================================================
RCS file: /home/cvspublic/jakarta-lucene/src/demo/org/apache/lucene/demo/IndexFiles.java,v
retrieving revision 1.2
diff -u -r1.2 IndexFiles.java
--- IndexFiles.java	29 Jan 2004 12:25:10 -0000	1.2
+++ IndexFiles.java	29 Mar 2004 19:42:06 -0000
@@ -27,6 +27,13 @@
 class IndexFiles {
   public static void main(String[] args) throws IOException {
     try {
+    	
+      String usage = "IndexFiles <root_directory>";
+      if (args.length == 0) {
+        System.err.println("Usage: " + usage);
+        return;
+      }
+
       Date start = new Date();
 
       IndexWriter writer = new IndexWriter("index", new StandardAnalyzer(), true);
