Index: solr/contrib/clustering/src/test/java/org/apache/solr/handler/clustering/AbstractClusteringTest.java
===================================================================
--- solr/contrib/clustering/src/test/java/org/apache/solr/handler/clustering/AbstractClusteringTest.java	(revision 993156)
+++ solr/contrib/clustering/src/test/java/org/apache/solr/handler/clustering/AbstractClusteringTest.java	(working copy)
@@ -25,7 +25,7 @@
 /**
  *
  */
-public class AbstractClusteringTest extends SolrTestCaseJ4 {
+public abstract class AbstractClusteringTest extends SolrTestCaseJ4 {
   protected static int numberOfDocs = 0;
 
   @BeforeClass
Index: lucene/src/test/org/apache/lucene/util/LuceneTestCaseJ4.java
===================================================================
--- lucene/src/test/org/apache/lucene/util/LuceneTestCaseJ4.java	(revision 993156)
+++ lucene/src/test/org/apache/lucene/util/LuceneTestCaseJ4.java	(working copy)
@@ -360,6 +360,7 @@
     ConcurrentMergeScheduler.setTestMode();
     savedBoolMaxClauseCount = BooleanQuery.getMaxClauseCount();
     seed = null;
+    TEMP_DIR.mkdirs();
   }
 
 
Index: lucene/src/test/org/apache/lucene/util/LuceneTestCase.java
===================================================================
--- lucene/src/test/org/apache/lucene/util/LuceneTestCase.java	(revision 993156)
+++ lucene/src/test/org/apache/lucene/util/LuceneTestCase.java	(working copy)
@@ -158,6 +155,7 @@
       ? LuceneTestCaseJ4.randomTimeZone(seedRnd)
       : TimeZone.getTimeZone(TEST_TIMEZONE);
     TimeZone.setDefault(timeZone);
+    TEMP_DIR.mkdirs();
   }
 
   /**
