Index: lucene/facet/src/java/org/apache/lucene/facet/search/DrillSidewaysScorer.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/facet/search/DrillSidewaysScorer.java	(revision 1470728)
+++ lucene/facet/src/java/org/apache/lucene/facet/search/DrillSidewaysScorer.java	(working copy)
@@ -90,15 +90,14 @@
 
     DocsEnum[][] docsEnums = new DocsEnum[numDims][];
     Collector[] sidewaysCollectors = new Collector[numDims];
-    int maxFreq = 0;
+    long drillDownCost = 0;
     for(int dim=0;dim<numDims;dim++) {
       docsEnums[dim] = dims[dim].docsEnums;
       sidewaysCollectors[dim] = dims[dim].sidewaysCollector;
-      maxFreq = Math.max(maxFreq, dims[dim].freq);
+      drillDownCost += dims[dim].freq;
     }
 
-    // TODO: if we add cost API to Scorer, switch to that!
-    int estBaseHitCount = context.reader().maxDoc() / (1+baseDocID);
+    long baseQueryCost = baseScorer.cost();
 
     /*
     System.out.println("\nbaseDocID=" + baseDocID + " est=" + estBaseHitCount);
@@ -110,10 +109,10 @@
     }
     */
 
-    if (estBaseHitCount < maxFreq/10) {
+    if (baseQueryCost < drillDownCost/10) {
       //System.out.println("baseAdvance");
       doBaseAdvanceScoring(collector, docsEnums, sidewaysCollectors);
-    } else if (numDims > 1 && (dims[1].freq < estBaseHitCount/10)) {
+    } else if (numDims > 1 && (dims[1].freq < drillDownCost/10)) {
       //System.out.println("drillDownAdvance");
       doDrillDownAdvanceScoring(collector, docsEnums, sidewaysCollectors);
     } else {
