Index: lucene/core/src/test/org/apache/lucene/index/TestCheckIndex.java
===================================================================
--- lucene/core/src/test/org/apache/lucene/index/TestCheckIndex.java	(revision 1361702)
+++ lucene/core/src/test/org/apache/lucene/index/TestCheckIndex.java	(working copy)
@@ -119,9 +119,19 @@
     // common-build.xml sets lucene.version
     final String version = System.getProperty("lucene.version");
     assertNotNull( "null version", version);
+    final String constantVersion;
+    String parts[] = Constants.LUCENE_MAIN_VERSION.split("\\.");
+    if (parts.length == 4) {
+      // alpha/beta version: pull the real portion
+      assert parts[2].equals("0");
+      constantVersion = parts[0] + "." + parts[1];
+    } else {
+      // normal version
+      constantVersion = Constants.LUCENE_MAIN_VERSION;
+    }
     assertTrue("Invalid version: "+version,
-               version.equals(Constants.LUCENE_MAIN_VERSION+"-SNAPSHOT") ||
-               version.equals(Constants.LUCENE_MAIN_VERSION));
+               version.equals(constantVersion+"-SNAPSHOT") ||
+               version.equals(constantVersion));
     assertTrue(Constants.LUCENE_VERSION + " should start with: "+version,
                Constants.LUCENE_VERSION.startsWith(version));
   }
Index: lucene/core/src/java/org/apache/lucene/util/Constants.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/util/Constants.java	(revision 1361702)
+++ lucene/core/src/java/org/apache/lucene/util/Constants.java	(working copy)
@@ -107,7 +107,7 @@
   /**
    * This is the internal Lucene version, recorded into each segment.
    */
-  public static final String LUCENE_MAIN_VERSION = ident("4.0");
+  public static final String LUCENE_MAIN_VERSION = ident("4.0.0.1");
 
   /**
    * This is the Lucene version for display purposes.
@@ -117,9 +117,14 @@
     Package pkg = LucenePackage.get();
     String v = (pkg == null) ? null : pkg.getImplementationVersion();
     if (v == null) {
-      v = LUCENE_MAIN_VERSION + "-SNAPSHOT";
-    } else if (!v.startsWith(LUCENE_MAIN_VERSION)) {
-      v = LUCENE_MAIN_VERSION + "-SNAPSHOT " + v;
+      String parts[] = LUCENE_MAIN_VERSION.split("\\.");
+      if (parts.length == 4) {
+        // alpha/beta
+        assert parts[2].equals("0");
+        v = parts[0] + "." + parts[1] + "-SNAPSHOT";
+      } else {
+        v = LUCENE_MAIN_VERSION + "-SNAPSHOT";
+      }
     }
     LUCENE_VERSION = ident(v);
   }
