Index: lucene/src/java/org/apache/lucene/index/DocInverterPerField.java
===================================================================
--- lucene/src/java/org/apache/lucene/index/DocInverterPerField.java	(revision 932398)
+++ lucene/src/java/org/apache/lucene/index/DocInverterPerField.java	(revision )
@@ -78,7 +78,7 @@
 
         final boolean anyToken;
         
-        if (fieldState.length > 0)
+        if (i > 0)
           fieldState.position += docState.analyzer.getPositionIncrementGap(fieldInfo.name);
 
         if (!field.isTokenized()) {		  // un-tokenized field
