Index: dev-tools/maven/pom.xml.template
===================================================================
--- dev-tools/maven/pom.xml.template	(revision 1689631)
+++ dev-tools/maven/pom.xml.template	(working copy)
@@ -228,6 +228,11 @@
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-javadoc-plugin</artifactId>
           <version>2.9.1</version>
+          <configuration>
+            <quiet>true</quiet>
+            <additionalparam>-Xdoclint:all</additionalparam>
+            <additionalparam>-Xdoclint:-missing</additionalparam>
+          </configuration>
         </plugin>
         <plugin>
           <groupId>org.apache.maven.plugins</groupId>
