Index: src/demo/org/apache/lucene/demo/html/HTMLParser.java
===================================================================
--- src/demo/org/apache/lucene/demo/html/HTMLParser.java	(revision 811975)
+++ src/demo/org/apache/lucene/demo/html/HTMLParser.java	(working copy)
@@ -128,7 +128,7 @@
       title.append(text);
     else {
       addToSummary(text);
-      if (!titleComplete && !title.equals("")) {  // finished title
+      if (!titleComplete && !(title.length() == 0)) {  // finished title
         synchronized(this) {
           titleComplete = true;                   // tell waiting threads
           notifyAll();
Index: src/demo/org/apache/lucene/demo/html/HTMLParser.jj
===================================================================
--- src/demo/org/apache/lucene/demo/html/HTMLParser.jj	(revision 811975)
+++ src/demo/org/apache/lucene/demo/html/HTMLParser.jj	(working copy)
@@ -155,7 +155,7 @@
       title.append(text);
     else {
       addToSummary(text);
-      if (!titleComplete && !title.equals("")) {  // finished title
+      if (!titleComplete && !(title.length() == 0)) {  // finished title
 	synchronized(this) {
 	  titleComplete = true;			  // tell waiting threads
 	  notifyAll();

