diff --git a/lucene/CHANGES.txt b/lucene/CHANGES.txt
index a0415ddb78..19ca27de0d 100644
--- a/lucene/CHANGES.txt
+++ b/lucene/CHANGES.txt
@@ -7,6 +7,9 @@ http://s.apache.org/luceneversions
 
 API Changes:
 
+* LUCENE-8467: RAMDirectory, RAMFile, RAMInputStream, RAMOutputStream are deprecated
+  (Dawid Weiss)
+
 * LUCENE-8356: StandardFilter is deprecated (Alan Woodward)
 
 * LUCENE-8373: ENGLISH_STOP_WORD_SET on StandardAnalyzer is deprecated.  Instead
diff --git a/lucene/MIGRATE.txt b/lucene/MIGRATE.txt
index 86539707af..a7cbfb8a43 100644
--- a/lucene/MIGRATE.txt
+++ b/lucene/MIGRATE.txt
@@ -149,3 +149,11 @@ attributes.
 
 This query matches only documents that have a value for the specified doc
 values field.
+
+## RAMDirectory, RAMFile, RAMInputStream, RAMOutputStream are deprecated ##
+
+This RAM-based directory implementation is an old piece of code that uses inefficient
+thread synchronization primitives and can be confused as "faster" than the NIO-based
+MMapDirectory. It is deprecated and scheduled for removal in future versions of 
+Lucene. (LUCENE-8467, LUCENE-8438)
+
diff --git a/lucene/core/src/java/org/apache/lucene/store/RAMDirectory.java b/lucene/core/src/java/org/apache/lucene/store/RAMDirectory.java
index 3c2bed2894..8d2dcf34dc 100644
--- a/lucene/core/src/java/org/apache/lucene/store/RAMDirectory.java
+++ b/lucene/core/src/java/org/apache/lucene/store/RAMDirectory.java
@@ -49,7 +49,12 @@ import org.apache.lucene.util.Accountables;
  * {@link MMapDirectory}, which is a high-performance directory
  * implementation working directly on the file system cache of the
  * operating system, so copying data to Java heap space is not useful.
+ * 
+ * @deprecated This class uses inefficient synchronization and is discouraged
+ * in favor of {@link MMapDirectory}. It will be removed in future versions 
+ * of Lucene.
  */
+@Deprecated
 public class RAMDirectory extends BaseDirectory implements Accountable {
   protected final Map<String,RAMFile> fileMap = new ConcurrentHashMap<>();
   protected final AtomicLong sizeInBytes = new AtomicLong();
diff --git a/lucene/core/src/java/org/apache/lucene/store/RAMFile.java b/lucene/core/src/java/org/apache/lucene/store/RAMFile.java
index d8ae693519..d4fd014e0a 100644
--- a/lucene/core/src/java/org/apache/lucene/store/RAMFile.java
+++ b/lucene/core/src/java/org/apache/lucene/store/RAMFile.java
@@ -24,7 +24,13 @@ import org.apache.lucene.util.Accountable;
 
 /** 
  * Represents a file in RAM as a list of byte[] buffers.
- * @lucene.internal */
+ * 
+ * @lucene.internal 
+ * @deprecated This class uses inefficient synchronization and is discouraged
+ * in favor of {@link MMapDirectory}. It will be removed in future versions 
+ * of Lucene.
+ */
+@Deprecated
 public class RAMFile implements Accountable {
   protected final ArrayList<byte[]> buffers = new ArrayList<>();
   long length;
diff --git a/lucene/core/src/java/org/apache/lucene/store/RAMInputStream.java b/lucene/core/src/java/org/apache/lucene/store/RAMInputStream.java
index 7c8c7a3564..cb07b6e42c 100644
--- a/lucene/core/src/java/org/apache/lucene/store/RAMInputStream.java
+++ b/lucene/core/src/java/org/apache/lucene/store/RAMInputStream.java
@@ -22,9 +22,15 @@ import java.io.IOException;
 
 import static org.apache.lucene.store.RAMOutputStream.BUFFER_SIZE;
 
-/** A memory-resident {@link IndexInput} implementation. 
- *  
- *  @lucene.internal */
+/** 
+ * A memory-resident {@link IndexInput} implementation. 
+ *
+ * @lucene.internal 
+ * @deprecated This class uses inefficient synchronization and is discouraged
+ * in favor of {@link MMapDirectory}. It will be removed in future versions 
+ * of Lucene.
+ */
+@Deprecated
 public class RAMInputStream extends IndexInput implements Cloneable {
 
   private final RAMFile file;
diff --git a/lucene/core/src/java/org/apache/lucene/store/RAMOutputStream.java b/lucene/core/src/java/org/apache/lucene/store/RAMOutputStream.java
index 3477fe449c..02dc843969 100644
--- a/lucene/core/src/java/org/apache/lucene/store/RAMOutputStream.java
+++ b/lucene/core/src/java/org/apache/lucene/store/RAMOutputStream.java
@@ -30,7 +30,11 @@ import org.apache.lucene.util.Accountables;
  * A memory-resident {@link IndexOutput} implementation.
  *
  * @lucene.internal
+ * @deprecated This class uses inefficient synchronization and is discouraged
+ * in favor of {@link MMapDirectory}. It will be removed in future versions 
+ * of Lucene.
  */
+@Deprecated
 public class RAMOutputStream extends IndexOutput implements Accountable {
   static final int BUFFER_SIZE = 1024;
 
