From daa179addd976edeb9229a7461aa99bee2efb352 Mon Sep 17 00:00:00 2001
From: Jason Gerlowski <jason.gerlowski@lucidworks.com>
Date: Mon, 4 Nov 2019 08:19:42 -0500
Subject: [PATCH] LUCENE-9026: Add getters for DocValuesTermsQuery values

---
 .../org/apache/lucene/search/DocValuesTermsQuery.java     | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/DocValuesTermsQuery.java b/lucene/sandbox/src/java/org/apache/lucene/search/DocValuesTermsQuery.java
index 111148feec5..4299ec29bd8 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/DocValuesTermsQuery.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/DocValuesTermsQuery.java
@@ -167,6 +167,14 @@ public class DocValuesTermsQuery extends Query implements Accountable {
     return builder.toString();
   }
 
+  public String getField() {
+    return field;
+  }
+
+  public PrefixCodedTerms getTerms() {
+    return termData;
+  }
+
   @Override
   public long ramBytesUsed() {
     return BASE_RAM_BYTES +
-- 
2.17.0

