Index: contrib/benchmark/src/java/org/apache/lucene/benchmark/standard/StandardBenchmarker.java
===================================================================
--- contrib/benchmark/src/java/org/apache/lucene/benchmark/standard/StandardBenchmarker.java	(revision 518792)
+++ contrib/benchmark/src/java/org/apache/lucene/benchmark/standard/StandardBenchmarker.java	(working copy)
@@ -275,6 +275,8 @@
         {
             body.append(line).append(' ');
         }
+        reader.close();
+        
         Date date = format.parse(dateStr.trim());
 
         doc.add(new Field("date", DateTools.dateToString(date, DateTools.Resolution.SECOND), Field.Store.YES, Field.Index.UN_TOKENIZED));
Index: contrib/benchmark/src/java/org/apache/lucene/benchmark/utils/ExtractReuters.java
===================================================================
--- contrib/benchmark/src/java/org/apache/lucene/benchmark/utils/ExtractReuters.java	(revision 518792)
+++ contrib/benchmark/src/java/org/apache/lucene/benchmark/utils/ExtractReuters.java	(working copy)
@@ -135,6 +135,7 @@
                     buffer.setLength(0);
                 }
             }
+            reader.close();
         }
 
         catch (
Index: contrib/benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/ReutersDocMaker.java
===================================================================
--- contrib/benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/ReutersDocMaker.java	(revision 518792)
+++ contrib/benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/ReutersDocMaker.java	(working copy)
@@ -79,6 +79,7 @@
     while ((line = reader.readLine()) != null) {
       bodyBuf.append(line).append(' ');
     }
+    reader.close();
     
     addBytes(f.length());
 
Index: contrib/benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/TrecDocMaker.java
===================================================================
--- contrib/benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/TrecDocMaker.java	(revision 518792)
+++ contrib/benchmark/src/java/org/apache/lucene/benchmark/byTask/feeds/TrecDocMaker.java	(working copy)
@@ -26,8 +26,10 @@
 import java.io.Reader;
 import java.io.StringReader;
 import java.text.DateFormat;
+import java.text.ParseException;
 import java.text.SimpleDateFormat;
 import java.util.ArrayList;
+import java.util.Date;
 import java.util.Locale;
 import java.util.Properties;
 import java.util.zip.GZIPInputStream;
@@ -159,7 +161,8 @@
     read("</DOCHDR>",null,false,false); 
     // 6. collect until end of doc
     sb = read("</DOC>",null,false,true);
-    // this is the next document, so parse it  
+    // this is the next document, so parse it 
+    // TODO use a more robust html parser (current one aborts parsing quite easily). 
     HTMLParser p = new HTMLParser(new StringReader(sb.toString()));
     // title
     String title = p.getTitle();
@@ -175,11 +178,18 @@
         bodyBuf.append(c,0,n);
       }
     }
+    r.close();
     addBytes(bodyBuf.length());
     
     DocData dd = new DocData();
-    
-    dd.date = dateFormat.parse(dateStr.trim());
+
+    try {
+      dd.date = dateFormat.parse(dateStr.trim());
+    } catch (ParseException e) {
+      // do not fail test just because a date could not be parsed
+      System.out.println("ignoring date parse exception (assigning 'now') for: "+dateStr);
+      dd.date = new Date(); // now 
+    }
     dd.name = name;
     dd.title = title;
     dd.body = bodyBuf.toString();
