diff --git a/lucene/core/src/java/org/apache/lucene/util/AttributeSource.java b/lucene/core/src/java/org/apache/lucene/util/AttributeSource.java
index e962fedc1de..ad158e0ec33 100644
--- a/lucene/core/src/java/org/apache/lucene/util/AttributeSource.java
+++ b/lucene/core/src/java/org/apache/lucene/util/AttributeSource.java
@@ -292,6 +292,8 @@ public class AttributeSource {
   /**
    * Captures the state of all Attributes. The return value can be passed to
    * {@link #restoreState} to restore the state of this or another AttributeSource.
+   * 
+   * Be careful, this method comes with a cost of copying all attributes in the source.
    */
   public final State captureState() {
     final State state = this.getCurrentState();
