Index: src/java/org/apache/lucene/analysis/StopFilter.java
===================================================================
--- src/java/org/apache/lucene/analysis/StopFilter.java	(revision 543080)
+++ src/java/org/apache/lucene/analysis/StopFilter.java	(working copy)
@@ -103,12 +103,23 @@
    * Returns the next input Token whose termText() is not a stop word.
    */
   public final Token next() throws IOException {
+    boolean flag = false ;
     // return the first non-stop word found
     for (Token token = input.next(); token != null; token = input.next())
     {
         String termText = ignoreCase ? token.termText.toLowerCase() : token.termText;
-        if (!stopWords.contains(termText))
-          return token;
+        if(stopWords.contains(termText)) {
+        	if( token.getPositionIncrement() == 1 ) {
+        		flag = true ;
+        	}
+        }else if( flag ){
+        	if(token.getPositionIncrement() == 1){
+        		flag = false ;
+                return token;
+        	}
+        }else{
+            return token;
+        }
     }
     // reached EOS -- return null
     return null;
