Index: lucene/src/java/org/apache/lucene/index/MultiReader.java
===================================================================
--- lucene/src/java/org/apache/lucene/index/MultiReader.java	(revision 1212754)
+++ lucene/src/java/org/apache/lucene/index/MultiReader.java	(working copy)
@@ -113,6 +113,21 @@
   }
   
   /**
+   * If the index has changed since it was opened, open and return a new reader;
+   * else, return {@code null}.
+   * 
+   * @see #openIfChanged(IndexReader, boolean)
+   * @deprecated Write support will be removed in Lucene 4.0.
+   * Use {@link #doOpenIfChanged()} instead
+   */
+  @Deprecated @Override
+  protected IndexReader doOpenIfChanged(boolean openReadOnly) throws CorruptIndexException, IOException {
+    if (!openReadOnly)
+      throw new UnsupportedOperationException("MultiReader does not support reopening in read/write mode");
+    return doReopen(false);
+  }
+
+  /**
    * Clones the subreaders.
    * (see {@link IndexReader#clone()}).
    * <br>
Index: lucene/src/java/org/apache/lucene/index/ParallelReader.java
===================================================================
--- lucene/src/java/org/apache/lucene/index/ParallelReader.java	(revision 1212754)
+++ lucene/src/java/org/apache/lucene/index/ParallelReader.java	(working copy)
@@ -173,7 +173,22 @@
     // doReopen calls ensureOpen
     return doReopen(false);
   }
-    
+  
+  /**
+   * If the index has changed since it was opened, open and return a new reader;
+   * else, return {@code null}.
+   * 
+   * @see #openIfChanged(IndexReader, boolean)
+   * @deprecated Write support will be removed in Lucene 4.0.
+   * Use {@link #doOpenIfChanged()} instead
+   */
+  @Deprecated @Override
+  protected IndexReader doOpenIfChanged(boolean openReadOnly) throws CorruptIndexException, IOException {
+    if (!openReadOnly)
+      throw new UnsupportedOperationException("ParallelReader does not support reopening in read/write mode");
+    return doReopen(false);
+  }
+
   private IndexReader doReopen(boolean doClone) throws CorruptIndexException, IOException {
     ensureOpen();
     
