Index: GermanStemmer.java
===================================================================
RCS file: /home/cvspublic/jakarta-lucene/src/java/org/apache/lucene/analysis/de/GermanStemmer.java,v
retrieving revision 1.7
diff -u -r1.7 GermanStemmer.java
--- GermanStemmer.java	9 Oct 2003 00:08:52 -0000	1.7
+++ GermanStemmer.java	29 Mar 2004 15:50:41 -0000
@@ -70,11 +70,6 @@
     private StringBuffer sb = new StringBuffer();
 
     /**
-     * Indicates if a term is handled as a noun.
-     */
-    private boolean uppercase = false;
-
-    /**
      * Amount of characters that are removed with <tt>substitute()</tt> while stemming.
      */
     private int substCount = 0;
@@ -87,8 +82,6 @@
      */
     protected String stem( String term )
     {
-	// Mark a possible noun.
-	uppercase = Character.isUpperCase( term.charAt( 0 ) );
 	// Use lowercase for medium stemming.
 	term = term.toLowerCase();
 	if ( !isStemmable( term ) )
@@ -153,7 +146,7 @@
 		buffer.deleteCharAt( buffer.length() - 1 );
 	    }
 	    // "t" occurs only as suffix of verbs.
-	    else if ( buffer.charAt( buffer.length() - 1 ) == 't' && !uppercase ) {
+	    else if ( buffer.charAt( buffer.length() - 1 ) == 't' ) {
 		buffer.deleteCharAt( buffer.length() - 1 );
 	    }
 	    else {
