Index: src/java/org/apache/lucene/analysis/TokenFilter.java
===================================================================
--- src/java/org/apache/lucene/analysis/TokenFilter.java	(revision 673054)
+++ src/java/org/apache/lucene/analysis/TokenFilter.java	(working copy)
@@ -39,4 +39,9 @@
     input.close();
   }
 
+  /** Reset the filter as well as the input TokenStream. */
+  public void reset() throws IOException {
+	super.reset();
+	input.reset();
+  }
 }
