Index: src/java/org/apache/lucene/analysis/CharArraySet.java
===================================================================
--- src/java/org/apache/lucene/analysis/CharArraySet.java	(revision 564920)
+++ src/java/org/apache/lucene/analysis/CharArraySet.java	(working copy)
@@ -55,7 +55,7 @@
     int pos = code & mask;
     char[] text2 = entries[pos];
     if (text2 != null && !equals(text, len, text2)) {
-      final int inc = code*1347|1;
+      final int inc = ((code>>8)+code)|1;
       do {
         code += inc;
         pos = code & mask;
@@ -79,7 +79,7 @@
     int pos = code & mask;
     char[] text2 = entries[pos];
     if (text2 != null) {
-      final int inc = code*1347|1;
+      final int inc = ((code>>8)+code)|1;
       do {
         code += inc;
         pos = code & mask;
@@ -120,7 +120,7 @@
         int code = getHashCode(text, text.length);
         int pos = code & mask;
         if (newEntries[pos] != null) {
-          final int inc = code*1347|1;
+          final int inc = ((code>>8)+code)|1;
           do {
             code += inc;
             pos = code & mask;
Index: src/java/org/apache/lucene/index/DocumentsWriter.java
===================================================================
--- src/java/org/apache/lucene/index/DocumentsWriter.java	(revision 564920)
+++ src/java/org/apache/lucene/index/DocumentsWriter.java	(working copy)
@@ -1375,7 +1375,7 @@
         if (p != null && !postingEquals(tokenText, tokenTextLen)) {
           // Conflict: keep searching different locations in
           // the hash table.
-          final int inc = code*1347|1;
+          final int inc = ((code>>8)+code)|1;
           do {
             code += inc;
             hashPos = code & postingsHashMask;
@@ -1550,7 +1550,7 @@
             int hashPos = code & postingsHashMask;
             assert hashPos >= 0;
             if (newHash[hashPos] != null) {
-              final int inc = code*1347|1;
+              final int inc = ((code>>8)+code)|1;
               do {
                 code += inc;
                 hashPos = code & postingsHashMask;
