diff --git solr/core/src/java/org/apache/solr/handler/component/FieldFacetStats.java solr/core/src/java/org/apache/solr/handler/component/FieldFacetStats.java
index 0a8703e..68af9d0 100644
--- solr/core/src/java/org/apache/solr/handler/component/FieldFacetStats.java
+++ solr/core/src/java/org/apache/solr/handler/component/FieldFacetStats.java
@@ -55,7 +55,6 @@ public class FieldFacetStats {
   List<HashMap<String, Integer>> facetStatsTerms;
 
   final AtomicReader topLevelReader;
-  AtomicReaderContext leave;
   final ValueSource valueSource;
   AtomicReaderContext context;
   FunctionValues values;
@@ -104,16 +103,9 @@ public class FieldFacetStats {
     }
     
     int term = topLevelSortedValues.getOrd(docID);
-    int arrIdx = term;
-    if (arrIdx >= 0 && arrIdx < topLevelSortedValues.getValueCount()) {
-      final BytesRef br;
-      if (term == -1) {
-        br = null;
-      } else {
-        br = tempBR;
-        topLevelSortedValues.lookupOrd(term, tempBR);
-      }
-      String key = br == null ? null : br.utf8ToString();
+    if (term >= 0 && term < topLevelSortedValues.getValueCount()) {
+      topLevelSortedValues.lookupOrd(term, tempBR);
+      String key = tempBR.utf8ToString();
       while (facetStatsTerms.size() <= statsTermNum) {
         facetStatsTerms.add(new HashMap<String, Integer>());
       }
