### Eclipse Workspace Patch 1.0
#P lucene
Index: src/java/org/apache/lucene/analysis/KeywordMarkerFilter.java
===================================================================
--- src/java/org/apache/lucene/analysis/KeywordMarkerFilter.java	(revision 1065046)
+++ src/java/org/apache/lucene/analysis/KeywordMarkerFilter.java	(working copy)
@@ -71,10 +71,12 @@
   @Override
   public final boolean incrementToken() throws IOException {
     if (input.incrementToken()) {
-      keywordAttr.setKeyword(keywordSet.contains(termAtt.buffer(), 0,
-          termAtt.length()));
+      if (keywordSet.contains(termAtt.buffer(), 0, termAtt.length())) { 
+        keywordAttr.setKeyword(true);
+      }
       return true;
-    } else
+    } else {
       return false;
+    }
   }
 }

