Index: .
===================================================================
--- .	(revision 1359946)
+++ .	(working copy)

Property changes on: .
___________________________________________________________________
Modified: svn:mergeinfo
   Merged /lucene/dev/trunk:r1359946
Index: lucene
===================================================================
--- lucene	(revision 1359946)
+++ lucene	(working copy)

Property changes on: lucene
___________________________________________________________________
Modified: svn:mergeinfo
   Merged /lucene/dev/trunk/lucene:r1359946
Index: lucene/suggest
===================================================================
--- lucene/suggest	(revision 1359946)
+++ lucene/suggest	(working copy)

Property changes on: lucene/suggest
___________________________________________________________________
Modified: svn:mergeinfo
   Merged /lucene/dev/trunk/lucene/suggest:r1359946
Index: lucene/suggest/src/java/org/apache/lucene/search/suggest/fst/Sort.java
===================================================================
--- lucene/suggest/src/java/org/apache/lucene/search/suggest/fst/Sort.java	(revision 1359946)
+++ lucene/suggest/src/java/org/apache/lucene/search/suggest/fst/Sort.java	(working copy)
@@ -224,10 +224,12 @@
 
     // One partition, try to rename or copy if unsuccessful.
     if (merges.size() == 1) {     
+      File single = merges.get(0);
       // If simple rename doesn't work this means the output is
       // on a different volume or something. Copy the input then.
-      if (!merges.get(0).renameTo(output)) {
-        copy(merges.get(0), output);
+      if (!single.renameTo(output)) {
+        copy(single, output);
+        single.delete();
       }
     } else { 
       // otherwise merge the partitions with a priority queue.                  
Index: lucene/CHANGES.txt
===================================================================
--- lucene/CHANGES.txt	(revision 1359946)
+++ lucene/CHANGES.txt	(working copy)
@@ -44,7 +44,8 @@
 
 * LUCENE-4209: Fix FSTCompletionLookup to close its sorter, so that it won't
   leave temp files behind in /tmp. Fix SortedTermFreqIteratorWrapper to not
-  leave temp files behind in /tmp on Windows.  (Uwe Schindler, Robert Muir)
+  leave temp files behind in /tmp on Windows. Fix Sort to not leave
+  temp files behind when /tmp is a separate volume. (Uwe Schindler, Robert Muir)
 
 Build
 

Property changes on: lucene/CHANGES.txt
___________________________________________________________________
Modified: svn:mergeinfo
   Merged /lucene/dev/trunk/lucene/CHANGES.txt:r1359946
