Index: dev-tools/scripts/smokeTestRelease.py
===================================================================
--- dev-tools/scripts/smokeTestRelease.py	(revision 1389187)
+++ dev-tools/scripts/smokeTestRelease.py	(working copy)
@@ -226,10 +226,9 @@
   if mavenURL is None:
     raise RuntimeError('%s is missing maven' % project)
 
-  if project == 'lucene':
-    if changesURL is None:
-      raise RuntimeError('%s is missing changes-%s' % (project, version))
-    testChanges(project, version, changesURL)
+  if changesURL is None:
+    raise RuntimeError('%s is missing changes-%s' % (project, version))
+  testChanges(project, version, changesURL)
 
   for artifact, urlString in artifacts:
     print('  download %s...' % artifact)
Index: build.xml
===================================================================
--- build.xml	(revision 1389187)
+++ build.xml	(working copy)
@@ -275,6 +275,9 @@
        <fileset dir="solr/package"/>
      </copy>
      <copy file="${fakeRelease}/lucene/KEYS" todir="${fakeRelease}/solr"/>
+     <copy todir="${fakeRelease}/solr/changes">
+       <fileset dir="solr/build/docs/changes"/>
+     </copy>
      <makeurl file="${fakeRelease}" validate="false" property="fakeRelease.uri"/>
      <exec executable="${python32.exe}" failonerror="true">
        <arg value="-u"/>
