Index: lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/RandomGeoPolygonTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/RandomGeoPolygonTest.java	(revision c1f42906ef1e0929f3887736dd9bb5bd0f9975c8)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/RandomGeoPolygonTest.java	(revision )
@@ -120,11 +120,12 @@
       referencePoint = randomGeoPoint(planetModel);
     }
     final int n = random().nextInt(4) + 4;
-    final List<GeoPoint> points = new ArrayList<>(n);
+
     List<GeoPoint> orderedPoints = null;
     GeoPolygon polygon = null;
     GeoPolygon largePolygon = null;
     do {
+      final List<GeoPoint> points = new ArrayList<>(n);
       double maxDistance = random().nextDouble() * limitDistance;
       //if distance is too small we can fail
       //building the polygon.
