Index: contrib/spatial/src/test/org/apache/lucene/spatial/tier/DistanceCheck.java
===================================================================
--- contrib/spatial/src/test/org/apache/lucene/spatial/tier/DistanceCheck.java	(revision 787424)
+++ contrib/spatial/src/test/org/apache/lucene/spatial/tier/DistanceCheck.java	Sun Dec 20 21:41:07 CET 2009
@@ -18,7 +18,10 @@
 
 import java.text.DecimalFormat;
 
-
+/**
+ * @deprecated This class does not test any functionality and only uses deprecated functionality.  This will be removed.
+ */
+@Deprecated
 public class DistanceCheck {
 
   /**
Index: contrib/spatial/src/java/org/apache/lucene/spatial/tier/DistanceHandler.java
===================================================================
--- contrib/spatial/src/java/org/apache/lucene/spatial/tier/DistanceHandler.java	(revision 834847)
+++ contrib/spatial/src/java/org/apache/lucene/spatial/tier/DistanceHandler.java	Sun Dec 20 21:39:30 CET 2009
@@ -28,7 +28,9 @@
  * flux and might change in incompatible ways in the next
  * release.</font>
  *
+ * @deprecated This class has been deprecated since its purpose is unclear and is unneeded.  This will be removed.
  */
+@Deprecated
 public class DistanceHandler {
 
   public enum Precision {EXACT, TWOFEET, TWENTYFEET, TWOHUNDREDFEET};
Index: contrib/spatial/src/test/org/apache/lucene/spatial/tier/TestDistance.java
===================================================================
--- contrib/spatial/src/test/org/apache/lucene/spatial/tier/TestDistance.java	(revision 825022)
+++ contrib/spatial/src/test/org/apache/lucene/spatial/tier/TestDistance.java	Sun Dec 20 21:40:42 CET 2009
@@ -33,8 +33,9 @@
 
 
 /**
- *
+ * @deprecated This test class tests deprecated behavior.  This will be removed.
  */
+@Deprecated
 public class TestDistance extends TestCase{
 
   
Index: contrib/spatial/src/java/org/apache/lucene/spatial/tier/DistanceUtils.java
===================================================================
--- contrib/spatial/src/java/org/apache/lucene/spatial/tier/DistanceUtils.java	(revision 810951)
+++ contrib/spatial/src/java/org/apache/lucene/spatial/tier/DistanceUtils.java	Sun Dec 20 21:38:36 CET 2009
@@ -27,7 +27,11 @@
  * <p><font color="red"><b>NOTE:</b> This API is still in
  * flux and might change in incompatible ways in the next
  * release.</font>
+ *
+ * @deprecated This class has been deprecated since it is only used by other deprecated classes and is unnecessary.  This
+ *             will be removed.
  */
+@Deprecated
 public class DistanceUtils {
 
   static DistanceUtils instance = new DistanceUtils();
Index: contrib/spatial/src/java/org/apache/lucene/spatial/tier/InvalidGeoException.java
===================================================================
--- contrib/spatial/src/java/org/apache/lucene/spatial/tier/InvalidGeoException.java	(revision 810951)
+++ contrib/spatial/src/java/org/apache/lucene/spatial/tier/InvalidGeoException.java	Sun Dec 20 21:39:01 CET 2009
@@ -21,7 +21,10 @@
  * <p><font color="red"><b>NOTE:</b> This API is still in
  * flux and might change in incompatible ways in the next
  * release.</font>
+ *
+ * @deprecated This Exception has been deprecated since it is unneeded.  This will be removed
  */
+@Deprecated
 public class InvalidGeoException extends Exception {
 
   /**
