Index: lucene/test-framework/src/java/org/apache/lucene/util/TestRuleTemporaryFilesCleanup.java
===================================================================
--- lucene/test-framework/src/java/org/apache/lucene/util/TestRuleTemporaryFilesCleanup.java	(revision 1649269)
+++ lucene/test-framework/src/java/org/apache/lucene/util/TestRuleTemporaryFilesCleanup.java	(working copy)
@@ -205,7 +205,7 @@
               "Failed to get a temporary name too many times, check your temp directory and consider manually cleaning it: "
                 + javaTempDir.toAbsolutePath());            
         }
-        f = javaTempDir.resolve(prefix + "-" + ctx.getRunnerSeedAsString() 
+        f = javaTempDir.resolve(prefix + " " + ctx.getRunnerSeedAsString() 
               + "-" + String.format(Locale.ENGLISH, "%03d", attempt));
         try {
           Files.createDirectory(f);
