Index: src/java/org/apache/lucene/analysis/tokenattributes/TermAttribute.java
===================================================================
--- src/java/org/apache/lucene/analysis/tokenattributes/TermAttribute.java	(revision 769060)
+++ src/java/org/apache/lucene/analysis/tokenattributes/TermAttribute.java	(working copy)
@@ -171,7 +171,6 @@
   /** Return number of valid characters (length of the term)
    *  in the termBuffer array. */
   public int termLength() {
-    initTermBuffer();
     return termLength;
   }
 
