Index: ./src/java/org/apache/lucene/queryParser/QueryParser.jj
===================================================================
RCS file: /home/cvs/jakarta-lucene/src/java/org/apache/lucene/queryParser/QueryParser.jj,v
retrieving revision 1.29
diff -u -u -r1.29 QueryParser.jj
--- ./src/java/org/apache/lucene/queryParser/QueryParser.jj	20 Mar 2003 18:28:13 -0000	1.29
+++ ./src/java/org/apache/lucene/queryParser/QueryParser.jj	11 May 2003 20:26:39 -0000
@@ -478,7 +478,7 @@
 }
 
 <Boost> TOKEN : {
-<NUMBER:    (<_NUM_CHAR>)+ ( "." (<_NUM_CHAR>)+ )? > : DEFAULT
+<NUMBER:    (["+","-"])?   (<_NUM_CHAR>)+ ( "." (<_NUM_CHAR>)+ )? > : DEFAULT
 }
 
 <RangeIn> TOKEN : {
