Index: src/java/org/apache/lucene/index/MultiLevelSkipListReader.java
===================================================================
--- src/java/org/apache/lucene/index/MultiLevelSkipListReader.java	(revision 553539)
+++ src/java/org/apache/lucene/index/MultiLevelSkipListReader.java	(working copy)
@@ -174,7 +174,7 @@
     Arrays.fill(numSkipped, 0);
     haveSkipped = false;
     for (int i = 1; i < numberOfSkipLevels; i++) {
-      skipStream[0] = null;
+      skipStream[i] = null;
     }
   }
   
