Index: lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizer.java
===================================================================
--- lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizer.java	(revision 0)
+++ lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizer.java	(working copy)
@@ -0,0 +1,56 @@
+package org.apache.lucene.analysis.icu;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.analysis.util.CharTokenizer;
+import org.apache.lucene.util.AttributeFactory;
+
+import com.ibm.icu.lang.UCharacter;
+
+/**
+ * A tokenizer that divides text at Unicode whitespace.
+ * Adjacent sequences of non-Whitespace characters form tokens.
+ * This tokenizer uses {@link UCharacter#isUWhiteSpace(int)}.
+ */
+public final class ICUWhitespaceTokenizer extends CharTokenizer {
+  
+  /**
+   * Construct a new ICUWhitespaceTokenizer.
+   */
+  public ICUWhitespaceTokenizer() {
+  }
+
+  /**
+   * Construct a new ICUWhitespaceTokenizer using a given
+   * {@link org.apache.lucene.util.AttributeFactory}.
+   *
+   * @param factory
+   *          the attribute factory to use for this {@link Tokenizer}
+   */
+  public ICUWhitespaceTokenizer(AttributeFactory factory) {
+    super(factory);
+  }
+  
+  /** Collects only characters which do not satisfy
+   * {@link UCharacter#isUWhiteSpace(int)}.*/
+  @Override
+  protected boolean isTokenChar(int c) {
+    return !UCharacter.isUWhiteSpace(c);
+  }
+}
Index: lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizer.java
===================================================================
--- lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizer.java	(revision 0)
+++ lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizer.java	(working copy)

Property changes on: lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizer.java
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Index: lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizerFactory.java
===================================================================
--- lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizerFactory.java	(revision 0)
+++ lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizerFactory.java	(working copy)
@@ -0,0 +1,48 @@
+package org.apache.lucene.analysis.icu;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.apache.lucene.analysis.util.TokenizerFactory;
+import org.apache.lucene.util.AttributeFactory;
+
+import java.util.Map;
+
+/**
+ * Factory for {@link ICUWhitespaceTokenizer}. 
+ * <pre class="prettyprint">
+ * &lt;fieldType name="text_ws" class="solr.TextField" positionIncrementGap="100"&gt;
+ *   &lt;analyzer&gt;
+ *     &lt;tokenizer class="solr.ICUWhitespaceTokenizerFactory"/&gt;
+ *   &lt;/analyzer&gt;
+ * &lt;/fieldType&gt;</pre>
+ */
+public class ICUWhitespaceTokenizerFactory extends TokenizerFactory {
+
+  /** Creates a new ICUWhitespaceTokenizerFactory */
+  public ICUWhitespaceTokenizerFactory(Map<String,String> args) {
+    super(args);
+    if (!args.isEmpty()) {
+      throw new IllegalArgumentException("Unknown parameters: " + args);
+    }
+  }
+
+  @Override
+  public ICUWhitespaceTokenizer create(AttributeFactory factory) {
+    return new ICUWhitespaceTokenizer(factory);
+  }
+}
Index: lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizerFactory.java
===================================================================
--- lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizerFactory.java	(revision 0)
+++ lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizerFactory.java	(working copy)

Property changes on: lucene/analysis/icu/src/java/org/apache/lucene/analysis/icu/ICUWhitespaceTokenizerFactory.java
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Index: lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUWhitespaceTokenizer.java
===================================================================
--- lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUWhitespaceTokenizer.java	(revision 0)
+++ lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUWhitespaceTokenizer.java	(working copy)
@@ -0,0 +1,44 @@
+package org.apache.lucene.analysis.icu;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import java.io.IOException;
+import java.io.StringReader;
+
+import org.apache.lucene.analysis.BaseTokenStreamTestCase;
+
+public class TestICUWhitespaceTokenizer extends BaseTokenStreamTestCase {
+
+  // clone of test from WhitespaceTokenizer
+  public void testSimple() throws IOException {
+    StringReader reader = new StringReader("Tokenizer \ud801\udc1ctest");
+    ICUWhitespaceTokenizer tokenizer = new ICUWhitespaceTokenizer();
+    tokenizer.setReader(reader);
+    assertTokenStreamContents(tokenizer, new String[] { "Tokenizer",
+    "\ud801\udc1ctest" });
+  }
+  
+  public void testNBSP() throws IOException {
+    StringReader reader = new StringReader("Tokenizer\u00A0test");
+    ICUWhitespaceTokenizer tokenizer = new ICUWhitespaceTokenizer();
+    tokenizer.setReader(reader);
+    assertTokenStreamContents(tokenizer, new String[] { "Tokenizer",
+    "test" });
+  }
+  
+}
Index: lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUWhitespaceTokenizer.java
===================================================================
--- lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUWhitespaceTokenizer.java	(revision 0)
+++ lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUWhitespaceTokenizer.java	(working copy)

Property changes on: lucene/analysis/icu/src/test/org/apache/lucene/analysis/icu/TestICUWhitespaceTokenizer.java
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
