Index: contrib/benchmark/pom.xml.template
===================================================================
--- contrib/benchmark/pom.xml.template	(revision 736281)
+++ contrib/benchmark/pom.xml.template	(working copy)
@@ -39,6 +39,11 @@
       <version>@version@</version>
     </dependency>
     <dependency>
+      <groupId>org.apache.lucene</groupId>
+      <artifactId>lucene-highlighter</artifactId>
+      <version>@version@</version>
+    </dependency>
+    <dependency>
       <groupId>commons-beanutils</groupId>
       <artifactId>commons-beanutils</artifactId>
       <version>${commons-beanutils-version}</version>
Index: contrib/highlighter/pom.xml.template
===================================================================
--- contrib/highlighter/pom.xml.template	(revision 736281)
+++ contrib/highlighter/pom.xml.template	(working copy)
@@ -35,4 +35,11 @@
     This is the highlighter for apache lucene java
   </description>
   <packaging>jar</packaging>
+  <dependencies>
+    <dependency>
+      <groupId>org.apache.lucene</groupId>
+      <artifactId>lucene-memory</artifactId>
+      <version>@version@</version>
+    </dependency>
+  </dependencies>
 </project>
