Index: lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoCircleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoCircleTest.java	(revision 4a12e066bbd257e79d942cad1f96f14f72f8584e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoCircleTest.java	(revision )
@@ -551,4 +551,12 @@
     assertEquals(westPoint.getLongitude(), bounds.getLeftLongitude(), 1e-2);
     assertEquals(eastPoint.getLongitude(), bounds.getRightLongitude(), 1e-2);
   }
+
+  @Test
+  public void testLUCENE8065(){
+    GeoCircle circle1 = GeoCircleFactory.makeExactGeoCircle(PlanetModel.WGS84, 0.03186456479560385, -2.2254294002683617, 1.5702573535090856, 8.184299676008562E-6);
+    GeoCircle circle2 = GeoCircleFactory.makeExactGeoCircle(PlanetModel.WGS84, 0.03186456479560385, -2.2254294002683617 , 1.5698163157923914, 1.0E-5);
+    assertTrue(circle1.getRelationship(circle2) != GeoArea.DISJOINT);
+
+  }
 }
