Index: dev-tools/scripts/buildAndPushRelease.py
===================================================================
--- dev-tools/scripts/buildAndPushRelease.py	(revision 1582190)
+++ dev-tools/scripts/buildAndPushRelease.py	(working copy)
@@ -32,6 +32,8 @@
 # it twice in a row!
 
 LOG = '/tmp/release.log'
+MAX_RUN_RETRIES = 5
+WAIT_SECS_RUN_RETRY = 30
 
 def log(msg):
   f = open(LOG, mode='ab')
@@ -45,6 +47,25 @@
     print(msg)
     raise RuntimeError(msg)
 
+def runWithRetries(command):
+  log('\n\n%s: RUN: %s\n' % (datetime.datetime.now(), command))
+  for attempt in range(MAX_RUN_RETRIES + 1):
+    if os.system('%s >> %s 2>&1' % (command, LOG)):
+      msg = '        FAILED: %s [see log %s]' % (command, LOG)
+      print(msg)
+      if attempt < MAX_RUN_RETRIES:
+        log('        FAILED\n')
+        if WAIT_SECS_RUN_RETRY > 0:
+          msg = '        RETRYING AFTER %d seconds...' % WAIT_SECS_RUN_RETRY
+          print(msg)
+          log(msg + '\n')
+          time.sleep(WAIT_SECS_RUN_RETRY)
+        msg = '    RETRY #%d: %s' % (attempt + 1, command)
+        print(msg)
+        log(msg + '\n')
+      else:
+        raise RuntimeError(msg)
+
 def runAndSendGPGPassword(command, password):
   p = subprocess.Popen(command, shell=True, bufsize=0, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, stdin=subprocess.PIPE)
   f = open(LOG, 'ab')
@@ -145,14 +166,14 @@
 
 def push(version, root, rev, rcNum, username):
   print('Push...')
-  dir = 'lucene-solr-%s-RC%d-rev%s' % (version, rcNum, rev)
-  s = os.popen('ssh %s@people.apache.org "ls -ld public_html/staging_area/%s" 2>&1' % (username, dir)).read()
+  lucene_solr_dir = 'lucene-solr-%s-RC%d-rev%s' % (version, rcNum, rev)
+  dir = 'public_html/staging_area/%s' % lucene_solr_dir
+  rsync = 'rsync --progress --inplace --partial --rsh=ssh'
+  s = os.popen('ssh %s@people.apache.org "ls -ld %s" 2>&1' % (username, dir)).read()
   if 'no such file or directory' not in s.lower():
     print('  Remove old dir...')
-    run('ssh %s@people.apache.org "chmod -R u+rwX public_html/staging_area/%s; rm -rf public_html/staging_area/%s"' % 
-        (username, dir, dir))
-  run('ssh %s@people.apache.org "mkdir -p public_html/staging_area/%s/lucene public_html/staging_area/%s/solr"' % \
-      (username, dir, dir))
+    runWithRetries('ssh %s@people.apache.org "chmod -R u+rwX %s ; rm -rf %s"' % (username, dir, dir))
+  runWithRetries('ssh %s@people.apache.org "mkdir -p %s/lucene %s/solr"' % (username, dir, dir))
   print('  Lucene')
   os.chdir('%s/lucene/dist' % root)
   print('    zip...')
@@ -160,9 +181,11 @@
     os.remove('lucene.tar.bz2')
   run('tar cjf lucene.tar.bz2 *')
   print('    copy...')
-  run('scp lucene.tar.bz2 %s@people.apache.org:public_html/staging_area/%s/lucene' % (username, dir))
+  # Target file must exist to allow 'rsync --partial --inplace' to work
+  runWithRetries('ssh %s@people.apache.org "touch %s/lucene/lucene.tar.bz2"' % (username, dir))
+  runWithRetries('%s lucene.tar.bz2 %s@people.apache.org:%s/lucene/' % (rsync, username, dir))
   print('    unzip...')
-  run('ssh %s@people.apache.org "cd public_html/staging_area/%s/lucene; tar xjf lucene.tar.bz2; rm -f lucene.tar.bz2"' % (username, dir))
+  runWithRetries('ssh %s@people.apache.org "cd %s/lucene; tar xjf lucene.tar.bz2; rm -f lucene.tar.bz2"' % (username, dir))
   os.remove('lucene.tar.bz2')
 
   print('  Solr')
@@ -172,16 +195,18 @@
     os.remove('solr.tar.bz2')
   run('tar cjf solr.tar.bz2 *')
   print('    copy...')
-  run('scp solr.tar.bz2 %s@people.apache.org:public_html/staging_area/%s/solr' % (username, dir))
+  # Target file must exist to allow 'rsync --partial --inplace' to work
+  runWithRetries('ssh %s@people.apache.org "touch %s/solr/solr.tar.bz2"' % (username, dir))
+  runWithRetries('%s solr.tar.bz2 %s@people.apache.org:%s/solr/' % (rsync, username, dir))
   print('    unzip...')
-  run('ssh %s@people.apache.org "cd public_html/staging_area/%s/solr; tar xjf solr.tar.bz2; rm -f solr.tar.bz2"' % (username, dir))
+  runWithRetries('ssh %s@people.apache.org "cd %s/solr; tar xjf solr.tar.bz2; rm -f solr.tar.bz2"' % (username, dir))
   os.remove('solr.tar.bz2')
 
   print('  chmod...')
-  run('ssh %s@people.apache.org "chmod -R a+rX-w public_html/staging_area/%s"' % (username, dir))
+  runWithRetries('ssh %s@people.apache.org "chmod -R a+rX-w %s"' % (username, dir))
 
   print('  done!')
-  url = 'https://people.apache.org/~%s/staging_area/%s' % (username, dir)
+  url = 'https://people.apache.org/~%s/staging_area/%s' % (username, lucene_solr_dir)
   return url
 
 def pushLocal(version, root, rev, rcNum, localDir):
@@ -217,7 +242,7 @@
   os.remove('%s/solr/package/solr.tar.bz2' % root)
 
   print('  KEYS')
-  run('wget http://people.apache.org/keys/group/lucene.asc')
+  runWithRetries('wget http://people.apache.org/keys/group/lucene.asc')
   os.rename('lucene.asc', 'KEYS')
   run('chmod a+r-w KEYS')
   run('cp KEYS ../lucene')
