diff --git a/lucene/core/src/java/org/apache/lucene/search/ReqExclScorer.java b/lucene/core/src/java/org/apache/lucene/search/ReqExclScorer.java
index 5ce6f5e..50a321b 100644
--- a/lucene/core/src/java/org/apache/lucene/search/ReqExclScorer.java
+++ b/lucene/core/src/java/org/apache/lucene/search/ReqExclScorer.java
@@ -154,7 +154,7 @@
         }
       };
     } else {
-      // reqTwoPhaseIterator is MORE costly than exclTwoPhaseIterator, check it first
+      // reqTwoPhaseIterator is MORE costly than exclTwoPhaseIterator, check it last
       return new TwoPhaseIterator(reqApproximation) {
 
         @Override
