Index: src/java/org/apache/lucene/search/spans/SpanScorer.java
===================================================================
--- src/java/org/apache/lucene/search/spans/SpanScorer.java	(revision 693011)
+++ src/java/org/apache/lucene/search/spans/SpanScorer.java	(working copy)
@@ -77,12 +77,12 @@
     }
     doc = spans.doc();
     freq = 0.0f;
-    while (more && doc == spans.doc()) {
+    do {
       int matchLength = spans.end() - spans.start();
       freq += getSimilarity().sloppyFreq(matchLength);
       more = spans.next();
-    }
-    return more || (freq != 0);
+    } while (more && (doc == spans.doc()));
+    return true;
   }
 
   public int doc() { return doc; }
