Index: src/java/org/apache/lucene/index/TermFreqVector.java
===================================================================
RCS file: /home/cvspublic/jakarta-lucene/src/java/org/apache/lucene/index/TermFreqVector.java,v
retrieving revision 1.1
diff -u -r1.1 TermFreqVector.java
--- src/java/org/apache/lucene/index/TermFreqVector.java	20 Feb 2004 20:14:55 -0000	1.1
+++ src/java/org/apache/lucene/index/TermFreqVector.java	24 Feb 2004 18:51:16 -0000
@@ -25,7 +25,7 @@
 
 
   /** Array of term frequencies. Locations of the array correspond one to one
-   *  to the term numbers in the array obtained from <code>getTermNumbers</code>
+   *  to the terms in the array obtained from <code>getTerms</code>
    *  method. Each location in the array contains the number of times this
    *  term occurs in the document or the document field.
    */
@@ -43,7 +43,7 @@
   public String toString(IndexReader ir) throws IOException;
 
 
-  /** Return an index in the term numbers array returned from <code>getTermNumbers</code>
+  /** Return an index in the terms array returned from <code>getTerms</code>
    *  at which the term with the specified <code>termNumber</code> appears. If this
    *  term does not appear in the array, return -1.
    */
