Index: src/java/org/apache/lucene/analysis/TokenFilter.java
===================================================================
--- src/java/org/apache/lucene/analysis/TokenFilter.java	(revision 774732)
+++ src/java/org/apache/lucene/analysis/TokenFilter.java	(working copy)
@@ -35,7 +35,7 @@
   */
 public abstract class TokenFilter extends TokenStream {
   /** The source of tokens for this filter. */
-  protected TokenStream input;
+  protected final TokenStream input;
 
   /** Construct a token stream filtering the given input. */
   protected TokenFilter(TokenStream input) {
