Index: lucene/CHANGES.txt
===================================================================
--- lucene/CHANGES.txt	(revision 1004320)
+++ lucene/CHANGES.txt	(working copy)
@@ -693,8 +693,7 @@
 
 * LUCENE-2123, LUCENE-2261: Move FuzzyQuery rewrite to separate RewriteMode 
   into MultiTermQuery. The number of fuzzy expansions can be specified with
-  the maxExpansions parameter to FuzzyQuery, but the default is limited to
-  BooleanQuery.maxClauseCount() as before. 
+  the maxExpansions parameter to FuzzyQuery.
   (Uwe Schindler, Robert Muir, Mike McCandless)
 
 * LUCENE-2164: ConcurrentMergeScheduler has more control over merge
Index: lucene/src/test/org/apache/lucene/search/TestTermRangeQuery.java
===================================================================
--- lucene/src/test/org/apache/lucene/search/TestTermRangeQuery.java	(revision 1004320)
+++ lucene/src/test/org/apache/lucene/search/TestTermRangeQuery.java	(working copy)
@@ -138,7 +138,7 @@
   }
   
   private void checkBooleanTerms(Searcher searcher, TermRangeQuery query, String... terms) throws IOException {
-    query.setRewriteMethod(new MultiTermQuery.TopTermsScoringBooleanQueryRewrite());
+    query.setRewriteMethod(new MultiTermQuery.TopTermsScoringBooleanQueryRewrite(50));
     final BooleanQuery bq = (BooleanQuery) searcher.rewrite(query);
     final Set<String> allowedTerms = new HashSet<String>(Arrays.asList(terms));
     assertEquals(allowedTerms.size(), bq.clauses().size());
Index: lucene/src/test/org/apache/lucene/search/TestFuzzyQuery.java
===================================================================
--- lucene/src/test/org/apache/lucene/search/TestFuzzyQuery.java	(revision 1004320)
+++ lucene/src/test/org/apache/lucene/search/TestFuzzyQuery.java	(working copy)
@@ -329,7 +329,7 @@
     writer.close();
     
     FuzzyQuery query = new FuzzyQuery(new Term("field", "lucene"));
-    query.setRewriteMethod(new MultiTermQuery.TopTermsBoostOnlyBooleanQueryRewrite());
+    query.setRewriteMethod(new MultiTermQuery.TopTermsBoostOnlyBooleanQueryRewrite(50));
     ScoreDoc[] hits = searcher.search(query, null, 1000).scoreDocs;
     assertEquals(3, hits.length);
     // normally, 'Lucenne' would be the first result as IDF will skew the score.
Index: lucene/src/java/org/apache/lucene/search/MultiTermQuery.java
===================================================================
--- lucene/src/java/org/apache/lucene/search/MultiTermQuery.java	(revision 1004320)
+++ lucene/src/java/org/apache/lucene/search/MultiTermQuery.java	(working copy)
@@ -283,14 +283,6 @@
       this.size = size;
     }
     
-    /** 
-     * Create a TopTermsBooleanQueryRewrite that is limited
-     * to at most {@link BooleanQuery#getMaxClauseCount} terms. 
-     */
-    public TopTermsBooleanQueryRewrite() {
-      this(Integer.MAX_VALUE);
-    }
-    
     /** Return a suitable Query for a MultiTermQuery term. */
     protected abstract Query getQuery(Term term, int docCount);
 
@@ -380,14 +372,6 @@
       TopTermsBooleanQueryRewrite {
 
     /** 
-     * Create a TopTermsScoringBooleanQueryRewrite that is limited
-     * to at most {@link BooleanQuery#getMaxClauseCount} terms. 
-     */
-    public TopTermsScoringBooleanQueryRewrite() {
-      super();
-    }
-
-    /** 
      * Create a TopTermsScoringBooleanQueryRewrite for 
      * at most <code>size</code> terms.
      * <p>
@@ -418,14 +402,6 @@
       TopTermsBooleanQueryRewrite {
     
     /** 
-     * Create a TopTermsBoostOnlyBooleanQueryRewrite that is limited
-     * to at most {@link BooleanQuery#getMaxClauseCount} terms. 
-     */
-    public TopTermsBoostOnlyBooleanQueryRewrite() {
-      super();
-    }
-
-    /** 
      * Create a TopTermsBoostOnlyBooleanQueryRewrite for 
      * at most <code>size</code> terms.
      * <p>
