Index: lucene/core/src/java/org/apache/lucene/codecs/lucene41/Lucene41SkipReader.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/codecs/lucene41/Lucene41SkipReader.java	(revision 1653376)
+++ lucene/core/src/java/org/apache/lucene/codecs/lucene41/Lucene41SkipReader.java	(working copy)
@@ -195,13 +195,13 @@
     // if (DEBUG) {
     //   System.out.println("  delta=" + delta);
     // }
-    docPointer[level] += skipStream.readVInt();
+    docPointer[level] += skipStream.readVLong();
     // if (DEBUG) {
     //   System.out.println("  docFP=" + docPointer[level]);
     // }
 
     if (posPointer != null) {
-      posPointer[level] += skipStream.readVInt();
+      posPointer[level] += skipStream.readVLong();
       // if (DEBUG) {
       //   System.out.println("  posFP=" + posPointer[level]);
       // }
@@ -215,7 +215,7 @@
       }
 
       if (payPointer != null) {
-        payPointer[level] += skipStream.readVInt();
+        payPointer[level] += skipStream.readVLong();
       }
     }
     return delta;
Index: lucene/core/src/java/org/apache/lucene/codecs/lucene41/Lucene41SkipWriter.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/codecs/lucene41/Lucene41SkipWriter.java	(revision 1653376)
+++ lucene/core/src/java/org/apache/lucene/codecs/lucene41/Lucene41SkipWriter.java	(working copy)
@@ -151,7 +151,7 @@
     skipBuffer.writeVInt(delta);
     lastSkipDoc[level] = curDoc;
 
-    skipBuffer.writeVInt((int) (curDocPointer - lastSkipDocPointer[level]));
+    skipBuffer.writeVLong(curDocPointer - lastSkipDocPointer[level]);
     lastSkipDocPointer[level] = curDocPointer;
 
     if (fieldHasPositions) {
@@ -158,7 +158,7 @@
       // if (DEBUG) {
       //   System.out.println("  curPosPointer=" + curPosPointer + " curPosBufferUpto=" + curPosBufferUpto);
       // }
-      skipBuffer.writeVInt((int) (curPosPointer - lastSkipPosPointer[level]));
+      skipBuffer.writeVLong(curPosPointer - lastSkipPosPointer[level]);
       lastSkipPosPointer[level] = curPosPointer;
       skipBuffer.writeVInt(curPosBufferUpto);
 
@@ -167,7 +167,7 @@
       }
 
       if (fieldHasOffsets || fieldHasPayloads) {
-        skipBuffer.writeVInt((int) (curPayPointer - lastSkipPayPointer[level]));
+        skipBuffer.writeVLong(curPayPointer - lastSkipPayPointer[level]);
         lastSkipPayPointer[level] = curPayPointer;
       }
     }
