Index: dev-tools/idea/.idea/libraries/JUnit.xml
===================================================================
--- dev-tools/idea/.idea/libraries/JUnit.xml	(revision 1714857)
+++ dev-tools/idea/.idea/libraries/JUnit.xml	(working copy)
@@ -2,7 +2,7 @@
   <library name="JUnit">
     <CLASSES>
       <root url="jar://$PROJECT_DIR$/lucene/test-framework/lib/junit-4.10.jar!/" />
-      <root url="jar://$PROJECT_DIR$/lucene/test-framework/lib/randomizedtesting-runner-2.2.0.jar!/" />
+      <root url="jar://$PROJECT_DIR$/lucene/test-framework/lib/randomizedtesting-runner-2.3.1.jar!/" />
     </CLASSES>
     <JAVADOC />
     <SOURCES />
Index: lucene/CHANGES.txt
===================================================================
--- lucene/CHANGES.txt	(revision 1714857)
+++ lucene/CHANGES.txt	(working copy)
@@ -294,6 +294,8 @@
 
 Other
 
+* LUCENE-6899: Upgrade randomizedtesting to 2.3.1. (Dawid Weiss)
+
 * LUCENE-6478: Test execution can hang with java.security.debug. (Dawid Weiss)
 
 * LUCENE-6862: Upgrade of RandomizedRunner to version 2.2.0. (Dawid Weiss)
Index: lucene/ivy-versions.properties
===================================================================
--- lucene/ivy-versions.properties	(revision 1714857)
+++ lucene/ivy-versions.properties	(working copy)
@@ -7,7 +7,7 @@
 /cglib/cglib-nodep = 2.2
 /com.adobe.xmp/xmpcore = 5.1.2
 
-com.carrotsearch.randomizedtesting.version = 2.2.0
+com.carrotsearch.randomizedtesting.version = 2.3.1
 /com.carrotsearch.randomizedtesting/junit4-ant = ${com.carrotsearch.randomizedtesting.version}
 /com.carrotsearch.randomizedtesting/randomizedtesting-runner = ${com.carrotsearch.randomizedtesting.version}
 
Index: lucene/licenses/junit4-ant-2.2.0.jar.sha1
===================================================================
--- lucene/licenses/junit4-ant-2.2.0.jar.sha1	(revision 1714857)
+++ lucene/licenses/junit4-ant-2.2.0.jar.sha1	(working copy)
@@ -1 +0,0 @@
-0d401c9c729deccd5db8a5df3102eb18793c2224
Index: lucene/licenses/junit4-ant-2.3.1.jar.sha1
===================================================================
--- lucene/licenses/junit4-ant-2.3.1.jar.sha1	(revision 0)
+++ lucene/licenses/junit4-ant-2.3.1.jar.sha1	(working copy)
@@ -0,0 +1 @@
+054cdf1533c6362812605dab2a984494930108d6

Property changes on: lucene/licenses/junit4-ant-2.3.1.jar.sha1
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Index: lucene/licenses/randomizedtesting-runner-2.2.0.jar.sha1
===================================================================
--- lucene/licenses/randomizedtesting-runner-2.2.0.jar.sha1	(revision 1714857)
+++ lucene/licenses/randomizedtesting-runner-2.2.0.jar.sha1	(working copy)
@@ -1 +0,0 @@
-60de504132241be049564a3a34fd7dcc296e2ef0
Index: lucene/licenses/randomizedtesting-runner-2.3.1.jar.sha1
===================================================================
--- lucene/licenses/randomizedtesting-runner-2.3.1.jar.sha1	(revision 0)
+++ lucene/licenses/randomizedtesting-runner-2.3.1.jar.sha1	(working copy)
@@ -0,0 +1 @@
+a6e5446a14e44802b91983b3e5c2d56e1d1c424a

Property changes on: lucene/licenses/randomizedtesting-runner-2.3.1.jar.sha1
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Index: lucene/spatial/src/test/org/apache/lucene/spatial/DistanceStrategyTest.java
===================================================================
--- lucene/spatial/src/test/org/apache/lucene/spatial/DistanceStrategyTest.java	(revision 1714857)
+++ lucene/spatial/src/test/org/apache/lucene/spatial/DistanceStrategyTest.java	(working copy)
@@ -22,7 +22,6 @@
 import java.util.Arrays;
 import java.util.List;
 
-import com.carrotsearch.randomizedtesting.annotations.Name;
 import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
 import com.spatial4j.core.context.SpatialContext;
 import com.spatial4j.core.shape.Point;
@@ -41,8 +40,7 @@
 import org.junit.Test;
 
 public class DistanceStrategyTest extends StrategyTestCase {
-
-  @ParametersFactory
+  @ParametersFactory(argumentFormatting = "strategy=%s")
   public static Iterable<Object[]> parameters() {
     List<Object[]> ctorArgs = new ArrayList<>();
 
@@ -52,46 +50,29 @@
 
     grid = new QuadPrefixTree(ctx,25);
     strategy = new RecursivePrefixTreeStrategy(grid, "recursive_quad");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     grid = new GeohashPrefixTree(ctx,12);
     strategy = new TermQueryPrefixTreeStrategy(grid, "termquery_geohash");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     grid = new PackedQuadPrefixTree(ctx,25);
     strategy = new RecursivePrefixTreeStrategy(grid, "recursive_packedquad");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     strategy = new PointVectorStrategy(ctx, "pointvector");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     strategy = new BBoxStrategy(ctx, "bbox");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     strategy = new SerializedDVStrategy(ctx, "serialized");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     return ctorArgs;
   }
 
-  // this is a hack for clover!
-  static class Param {
-    SpatialStrategy strategy;
-
-    Param(SpatialStrategy strategy) {
-      this.strategy = strategy;
-    }
-
-    @Override
-    public String toString() {
-      return strategy.getFieldName();
-    }
-  }
-
-//  private String fieldName;
-
-  public DistanceStrategyTest(@Name("strategy") Param param) {
-    SpatialStrategy strategy = param.strategy;
+  public DistanceStrategyTest(String suiteName, SpatialStrategy strategy) {
     this.ctx = strategy.getSpatialContext();
     this.strategy = strategy;
   }
Index: lucene/spatial/src/test/org/apache/lucene/spatial/PortedSolr3Test.java
===================================================================
--- lucene/spatial/src/test/org/apache/lucene/spatial/PortedSolr3Test.java	(revision 1714857)
+++ lucene/spatial/src/test/org/apache/lucene/spatial/PortedSolr3Test.java	(working copy)
@@ -22,7 +22,6 @@
 import java.util.List;
 import java.util.Set;
 
-import com.carrotsearch.randomizedtesting.annotations.Name;
 import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
 import com.spatial4j.core.context.SpatialContext;
 import com.spatial4j.core.distance.DistanceUtils;
@@ -44,7 +43,7 @@
  */
 public class PortedSolr3Test extends StrategyTestCase {
 
-  @ParametersFactory
+  @ParametersFactory(argumentFormatting = "strategy=%s")
   public static Iterable<Object[]> parameters() {
     List<Object[]> ctorArgs = new ArrayList<>();
 
@@ -54,36 +53,23 @@
 
     grid = new GeohashPrefixTree(ctx,12);
     strategy = new RecursivePrefixTreeStrategy(grid, "recursive_geohash");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     grid = new QuadPrefixTree(ctx,25);
     strategy = new RecursivePrefixTreeStrategy(grid, "recursive_quad");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     grid = new GeohashPrefixTree(ctx,12);
     strategy = new TermQueryPrefixTreeStrategy(grid, "termquery_geohash");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     strategy = new PointVectorStrategy(ctx, "pointvector");
-    ctorArgs.add(new Object[]{new Param(strategy)});
+    ctorArgs.add(new Object[]{strategy.getFieldName(), strategy});
 
     return ctorArgs;
   }
-  
-  // this is a hack for clover! (otherwise strategy.toString() used as file name)
-  static class Param {
-    SpatialStrategy strategy;
 
-    Param(SpatialStrategy strategy) { this.strategy = strategy; }
-    
-    @Override
-    public String toString() { return strategy.getFieldName(); }
-  }
-
-//  private String fieldName;
-
-  public PortedSolr3Test(@Name("strategy") Param param) {
-    SpatialStrategy strategy = param.strategy;
+  public PortedSolr3Test(String suiteName, SpatialStrategy strategy) {
     this.ctx = strategy.getSpatialContext();
     this.strategy = strategy;
   }
Index: solr/licenses/junit4-ant-2.2.0.jar.sha1
===================================================================
--- solr/licenses/junit4-ant-2.2.0.jar.sha1	(revision 1714857)
+++ solr/licenses/junit4-ant-2.2.0.jar.sha1	(working copy)
@@ -1 +0,0 @@
-0d401c9c729deccd5db8a5df3102eb18793c2224
Index: solr/licenses/junit4-ant-2.3.1.jar.sha1
===================================================================
--- solr/licenses/junit4-ant-2.3.1.jar.sha1	(revision 0)
+++ solr/licenses/junit4-ant-2.3.1.jar.sha1	(working copy)
@@ -0,0 +1 @@
+054cdf1533c6362812605dab2a984494930108d6

Property changes on: solr/licenses/junit4-ant-2.3.1.jar.sha1
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Index: solr/licenses/randomizedtesting-runner-2.2.0.jar.sha1
===================================================================
--- solr/licenses/randomizedtesting-runner-2.2.0.jar.sha1	(revision 1714857)
+++ solr/licenses/randomizedtesting-runner-2.2.0.jar.sha1	(working copy)
@@ -1 +0,0 @@
-60de504132241be049564a3a34fd7dcc296e2ef0
Index: solr/licenses/randomizedtesting-runner-2.3.1.jar.sha1
===================================================================
--- solr/licenses/randomizedtesting-runner-2.3.1.jar.sha1	(revision 0)
+++ solr/licenses/randomizedtesting-runner-2.3.1.jar.sha1	(working copy)
@@ -0,0 +1 @@
+a6e5446a14e44802b91983b3e5c2d56e1d1c424a

Property changes on: solr/licenses/randomizedtesting-runner-2.3.1.jar.sha1
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
