Index: lucene/queries/src/java/org/apache/lucene/queries/BoostingQuery.java
===================================================================
--- lucene/queries/src/java/org/apache/lucene/queries/BoostingQuery.java	(revision 1701231)
+++ lucene/queries/src/java/org/apache/lucene/queries/BoostingQuery.java	(working copy)
@@ -21,10 +21,10 @@
 import java.util.Objects;
 import java.util.Set;
 
+import org.apache.lucene.index.IndexReader;
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.Term;
 import org.apache.lucene.search.*;
-import org.apache.lucene.util.Bits;
 
 /**
  * The BoostingQuery class can be used to effectively demote results that match a given query. 
@@ -55,6 +55,18 @@
     }
 
     @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+      Query matchRewritten = match.rewrite(reader);
+      Query contextRewritten = context.rewrite(reader);
+      if (match != matchRewritten || context != contextRewritten) {
+        BoostingQuery rewritten = new BoostingQuery(matchRewritten, contextRewritten, boost);
+        rewritten.setBoost(getBoost());
+        return rewritten;
+      }
+      return super.rewrite(reader);
+    }
+
+    @Override
     public Weight createWeight(IndexSearcher searcher, boolean needsScores) throws IOException {
       if (needsScores == false) {
         return match.createWeight(searcher, needsScores);
Index: lucene/queries/src/test/org/apache/lucene/queries/BoostingQueryTest.java
===================================================================
--- lucene/queries/src/test/org/apache/lucene/queries/BoostingQueryTest.java	(revision 1701231)
+++ lucene/queries/src/test/org/apache/lucene/queries/BoostingQueryTest.java	(working copy)
@@ -17,7 +17,15 @@
  * limitations under the License.
  */
 
+import java.io.IOException;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.MultiReader;
 import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.MatchAllDocsQuery;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
 import org.apache.lucene.search.QueryUtils;
 import org.apache.lucene.search.TermQuery;
 import org.apache.lucene.util.LuceneTestCase;
@@ -35,4 +43,13 @@
     BoostingQuery bq2 = new BoostingQuery(q1, q2, 0.1f);
     assertEquals("BoostingQuery with same attributes is not equal", bq1, bq2);
   }
+
+  public void testRewrite() throws IOException {
+    IndexReader reader = new MultiReader();
+    BoostingQuery q = new BoostingQuery(new MatchNoDocsQuery(), new MatchAllDocsQuery(), 3);
+    Query rewritten = q.rewrite(reader);
+    Query expectedRewritten = new BoostingQuery(new BooleanQuery.Builder().build(), new MatchAllDocsQuery(), 3);
+    assertEquals(expectedRewritten, rewritten);
+    assertSame(rewritten, rewritten.rewrite(reader));
+  }
 }
