diff --git a/lucene/CHANGES.txt b/lucene/CHANGES.txt
index 946706ec39..be529e27b1 100644
--- a/lucene/CHANGES.txt
+++ b/lucene/CHANGES.txt
@@ -20,6 +20,8 @@ API Changes
 
 Bug fixes
 
+* LUCENE-8804: Today FieldType attribute map can be modified even after freeze. For all other properties of FieldType, we do "checkIfFrozen()" before making the update to the property but for attribute map, we does not seem to make such check. 
+
 * LUCENE-8663: NRTCachingDirectory.slowFileExists may open a file while 
   it's inaccessible. (Dawid Weiss)
 
diff --git a/lucene/core/src/java/org/apache/lucene/document/FieldType.java b/lucene/core/src/java/org/apache/lucene/document/FieldType.java
index b450eee385..8dd802c697 100644
--- a/lucene/core/src/java/org/apache/lucene/document/FieldType.java
+++ b/lucene/core/src/java/org/apache/lucene/document/FieldType.java
@@ -361,6 +361,7 @@ public class FieldType implements IndexableFieldType  {
    * @lucene.experimental
    */
   public String putAttribute(String key, String value) {
+    checkIfFrozen();
     if (attributes == null) {
       attributes = new HashMap<>();
     }
diff --git a/lucene/core/src/test/org/apache/lucene/document/TestFieldType.java b/lucene/core/src/test/org/apache/lucene/document/TestFieldType.java
index b770e62cb5..a9ac3c1271 100644
--- a/lucene/core/src/test/org/apache/lucene/document/TestFieldType.java
+++ b/lucene/core/src/test/org/apache/lucene/document/TestFieldType.java
@@ -72,7 +72,33 @@ public class TestFieldType extends LuceneTestCase {
     assertEquals("pointDataDimensionCount=1,pointIndexDimensionCount=1,pointNumBytes=4", ft.toString());
   }
 
-  private static Object randomValue(Class<?> clazz) {
+  /**
+   * FieldType's attribute map should not be modifiable/add after freeze
+   */
+  public void testAttributeMapFrozen() {
+    try {
+      FieldType ft = new FieldType();
+      ft.putAttribute("dummy", "d");
+      ft.freeze();
+      ft.putAttribute("dummy", "a");
+      fail("This FieldType is already frozen and should not have been changed");
+    } catch (IllegalStateException ex) {
+      // expected
+    }
+  }
+
+  /**
+   * FieldType's attribute map can be changed if not frozen
+   */
+  public void testAttributeMapNotFrozen() {
+    FieldType ft = new FieldType();
+    ft.putAttribute("dummy", "d");
+    ft.putAttribute("dummy", "a");
+    assertEquals(ft.getAttributes().size(), 1);
+    assertEquals(ft.getAttributes().get("dummy"), "a");
+  }
+
+    private static Object randomValue(Class<?> clazz) {
     if (clazz.isEnum()) {
       return RandomPicks.randomFrom(random(), clazz.getEnumConstants());
     } else if (clazz == boolean.class) {
