diff -Nur lucene-4.2/core/src/test/org/apache/lucene/index/TestIndexWriterOnJRECrash.java lucene-4.2-new/core/src/test/org/apache/lucene/index/TestIndexWriterOnJRECrash.java
--- lucene-4.2/core/src/test/org/apache/lucene/index/TestIndexWriterOnJRECrash.java	2013-04-01 20:56:26.114960731 -0300
+++ lucene-4.2-new/core/src/test/org/apache/lucene/index/TestIndexWriterOnJRECrash.java	2013-04-02 22:27:59.924221189 -0300
@@ -52,7 +52,8 @@
   public void testNRTThreads() throws Exception {
     String vendor = Constants.JAVA_VENDOR;
     assumeTrue(vendor + " JRE not supported.", 
-        vendor.startsWith("Oracle") || vendor.startsWith("Sun") || vendor.startsWith("Apple"));
+        vendor.startsWith("Oracle") || vendor.startsWith("Sun") || vendor.startsWith("Apple")
+        || vendor.startsWith("IBM"));
     
     // if we are not the fork
     if (System.getProperty("tests.crashmode") == null) {
