Index: lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/XYZSolidTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/XYZSolidTest.java	(revision 37e8fea22e532f9b9deb9ce040427cb8abc4a455)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/XYZSolidTest.java	(date 1534484140000)
@@ -216,5 +216,20 @@
     // MHL for degenerate Z
     
   }
+
+  @Test
+  public void testLUCENE8457() {
+    GeoShape shape = GeoBBoxFactory.makeGeoBBox(PlanetModel.WGS84, Math.PI, 1.2487354264870392, 0.0, 3.5181789305199657E-12);
+    XYZBounds bounds = new XYZBounds();
+    shape.getBounds(bounds);
+    XYZSolid solid = XYZSolidFactory.makeXYZSolid(PlanetModel.WGS84, bounds);
+
+    GeoPoint point = new GeoPoint(PlanetModel.WGS84, 1.4812439919751819, -3.141592653589793);
+    //if the point is within the shape, it must be within the solid
+    if (shape.isWithin(point)) {
+      assertTrue(solid.isWithin(point));
+    }
+
+  }
   
 }
