Index: src/java/org/apache/lucene/search/FieldSortedHitQueue.java
===================================================================
RCS file: /home/cvspublic/jakarta-lucene/src/java/org/apache/lucene/search/FieldSortedHitQueue.java,v
retrieving revision 1.11
diff -u -r1.11 FieldSortedHitQueue.java
--- src/java/org/apache/lucene/search/FieldSortedHitQueue.java	21 Jul 2004 19:05:46 -0000	1.11
+++ src/java/org/apache/lucene/search/FieldSortedHitQueue.java	29 Sep 2004 10:13:46 -0000
@@ -197,10 +197,9 @@
   static ScoreDocComparator comparatorInt (final IndexReader reader, final String fieldname)
   throws IOException {
     final String field = fieldname.intern();
+    final int[] fieldOrder = FieldCache.DEFAULT.getInts (reader, field);
     return new ScoreDocComparator() {
 
-      final int[] fieldOrder = FieldCache.DEFAULT.getInts (reader, field);
-
       public final int compare (final ScoreDoc i, final ScoreDoc j) {
         final int fi = fieldOrder[i.doc];
         final int fj = fieldOrder[j.doc];
@@ -229,10 +228,9 @@
   static ScoreDocComparator comparatorFloat (final IndexReader reader, final String fieldname)
   throws IOException {
     final String field = fieldname.intern();
+    final float[] fieldOrder = FieldCache.DEFAULT.getFloats (reader, field);
     return new ScoreDocComparator () {
 
-      protected final float[] fieldOrder = FieldCache.DEFAULT.getFloats (reader, field);
-
       public final int compare (final ScoreDoc i, final ScoreDoc j) {
         final float fi = fieldOrder[i.doc];
         final float fj = fieldOrder[j.doc];
@@ -261,8 +259,8 @@
   static ScoreDocComparator comparatorString (final IndexReader reader, final String fieldname)
   throws IOException {
     final String field = fieldname.intern();
+    final FieldCache.StringIndex index = FieldCache.DEFAULT.getStringIndex (reader, field);
     return new ScoreDocComparator () {
-      final FieldCache.StringIndex index = FieldCache.DEFAULT.getStringIndex (reader, field);
 
       public final int compare (final ScoreDoc i, final ScoreDoc j) {
         final int fi = index.order[i.doc];
@@ -293,8 +291,8 @@
   throws IOException {
     final Collator collator = Collator.getInstance (locale);
     final String field = fieldname.intern();
+    final String[] index = FieldCache.DEFAULT.getStrings (reader, field);
     return new ScoreDocComparator() {
-      final String[] index = FieldCache.DEFAULT.getStrings (reader, field);
 
       public final int compare (final ScoreDoc i, final ScoreDoc j) {
         return collator.compare (index[i.doc], index[j.doc]);
