Index: lucene/core/src/java/org/apache/lucene/index/TwoPhaseCommitTool.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/index/TwoPhaseCommitTool.java	(revision 1677364)
+++ lucene/core/src/java/org/apache/lucene/index/TwoPhaseCommitTool.java	(working copy)
@@ -55,8 +55,8 @@
     
   }
 
-  /** rollback all objects, discarding any exceptions that occur. */
-  private static void rollback(TwoPhaseCommit... objects) {
+  /** rollback all objects, suppressing any exceptions that occur. */
+  private static void rollback(Throwable priorException, TwoPhaseCommit... objects) {
     for (TwoPhaseCommit tpc : objects) {
       // ignore any exception that occurs during rollback - we want to ensure
       // all objects are rolled-back.
@@ -63,7 +63,9 @@
       if (tpc != null) {
         try {
           tpc.rollback();
-        } catch (Throwable t) {}
+        } catch (Throwable t) {
+          priorException.addSuppressed(t);
+        }
       }
     }
   }
@@ -103,7 +105,7 @@
     } catch (Throwable t) {
       // first object that fails results in rollback all of them and
       // throwing an exception.
-      rollback(objects);
+      rollback(t, objects);
       throw new PrepareCommitFailException(t, tpc);
     }
     
@@ -118,7 +120,7 @@
     } catch (Throwable t) {
       // first object that fails results in rollback all of them and
       // throwing an exception.
-      rollback(objects);
+      rollback(t, objects);
       throw new CommitFailException(t, tpc);
     }
   }
