Index: lucene/test-framework/src/java/org/apache/lucene/index/AssertingAtomicReader.java
===================================================================
--- lucene/test-framework/src/java/org/apache/lucene/index/AssertingAtomicReader.java	(revision 1386719)
+++ lucene/test-framework/src/java/org/apache/lucene/index/AssertingAtomicReader.java	(working copy)
@@ -223,13 +223,18 @@
       super(in);
       int docid = in.docID();
       assert docid == -1 || docid == DocIdSetIterator.NO_MORE_DOCS : "invalid initial doc id: " + docid;
+      if (docid == NO_MORE_DOCS) {
+        state = DocsEnumState.FINISHED;
+      }
     }
 
     @Override
     public int nextDoc() throws IOException {
       assert state != DocsEnumState.FINISHED : "nextDoc() called after NO_MORE_DOCS";
+      int prevDoc = docID();
       int nextDoc = super.nextDoc();
       assert nextDoc >= 0 : "invalid doc id: " + nextDoc;
+      assert nextDoc > prevDoc : "docsenum went backwards: prevDoc=" + prevDoc + ",doc=" + nextDoc; 
       if (nextDoc == DocIdSetIterator.NO_MORE_DOCS) {
         state = DocsEnumState.FINISHED;
       } else {
@@ -275,13 +280,18 @@
       super(in);
       int docid = in.docID();
       assert docid == -1 || docid == DocIdSetIterator.NO_MORE_DOCS : "invalid initial doc id: " + docid;
+      if (docid == DocIdSetIterator.NO_MORE_DOCS) {
+        state = DocsEnumState.FINISHED;
+      }
     }
 
     @Override
     public int nextDoc() throws IOException {
       assert state != DocsEnumState.FINISHED : "nextDoc() called after NO_MORE_DOCS";
+      int prevDoc = docID();
       int nextDoc = super.nextDoc();
       assert nextDoc >= 0 : "invalid doc id: " + nextDoc;
+      assert nextDoc > prevDoc : "docsenum went backwards: prevDoc=" + prevDoc + ",doc=" + nextDoc; 
       positionCount = 0;
       if (nextDoc == DocIdSetIterator.NO_MORE_DOCS) {
         state = DocsEnumState.FINISHED;
Index: lucene/core/src/java/org/apache/lucene/search/DocIdSetIterator.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/search/DocIdSetIterator.java	(revision 1386719)
+++ lucene/core/src/java/org/apache/lucene/search/DocIdSetIterator.java	(working copy)
@@ -37,8 +37,9 @@
   /**
    * Returns the following:
    * <ul>
-   * <li>-1 or {@link #NO_MORE_DOCS} if {@link #nextDoc()} or
-   * {@link #advance(int)} were not called yet.
+   * <li>-1 if {@link #nextDoc()} or
+   * {@link #advance(int)} were not called yet. Optionally, the iterator
+   * can return {@link #NO_MORE_DOCS} when unpositioned AND exhausted.
    * <li>{@link #NO_MORE_DOCS} if the iterator has exhausted.
    * <li>Otherwise it should return the doc ID it is currently on.
    * </ul>
