Index: TermVectorsReader.java
===================================================================
RCS file: /home/cvspublic/jakarta-lucene/src/java/org/apache/lucene/index/TermVectorsReader.java,v
retrieving revision 1.3
diff -u -r1.3 TermVectorsReader.java
--- TermVectorsReader.java	17 Aug 2004 20:53:16 -0000	1.3
+++ TermVectorsReader.java	18 Aug 2004 19:25:53 -0000
@@ -22,8 +22,6 @@
 import java.io.IOException;
 
 /**
- * FIXME: relax synchro!
- *
  * @version $Id: TermVectorsReader.java,v 1.3 2004/08/17 20:53:16 otis Exp $
  */
 class TermVectorsReader {
@@ -84,7 +82,7 @@
    * @param field The field within the document to retrieve
    * @return The TermFreqVector for the document and field or null
    */ 
-  synchronized TermFreqVector get(int docNum, String field) {
+  TermFreqVector get(int docNum, String field) {
     // Check if no term vectors are available for this segment at all
     int fieldNumber = fieldInfos.fieldNumber(field);
     TermFreqVector result = null;
@@ -137,7 +135,7 @@
 
 
   /** Return all term vectors stored for this document or null if the could not be read in. */
-  synchronized TermFreqVector[] get(int docNum) {
+  TermFreqVector[] get(int docNum) {
     TermFreqVector[] result = null;
     // Check if no term vectors are available for this segment at all
     if (tvx != null) {
