Index: solr/src/test/org/apache/solr/analysis/TestPatternTokenizerFactory.java
===================================================================
--- solr/src/test/org/apache/solr/analysis/TestPatternTokenizerFactory.java	(revision 940668)
+++ solr/src/test/org/apache/solr/analysis/TestPatternTokenizerFactory.java	(working copy)
@@ -26,8 +26,8 @@
 
 import org.apache.lucene.analysis.CharReader;
 import org.apache.lucene.analysis.CharStream;
-import org.apache.lucene.analysis.MappingCharFilter;
-import org.apache.lucene.analysis.NormalizeCharMap;
+import org.apache.lucene.analysis.charfilter.MappingCharFilter;
+import org.apache.lucene.analysis.charfilter.NormalizeCharMap;
 import org.apache.lucene.analysis.TokenStream;
 import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
 
Index: solr/src/test/org/apache/solr/handler/FieldAnalysisRequestHandlerTest.java
===================================================================
--- solr/src/test/org/apache/solr/handler/FieldAnalysisRequestHandlerTest.java	(revision 940668)
+++ solr/src/test/org/apache/solr/handler/FieldAnalysisRequestHandlerTest.java	(working copy)
@@ -312,7 +312,7 @@
     assertNotNull("expecting an index token analysis for field type 'charfilthtmlmap'", indexPart);
     
     assertEquals("  whátëvêr  ", indexPart.get("org.apache.solr.analysis.HTMLStripCharFilter"));
-    assertEquals("  whatever  ", indexPart.get("org.apache.lucene.analysis.MappingCharFilter"));
+    assertEquals("  whatever  ", indexPart.get("org.apache.lucene.analysis.charfilter.MappingCharFilter"));
 
     List<NamedList> tokenList = (List<NamedList>)indexPart.get("org.apache.lucene.analysis.WhitespaceTokenizer");
     assertNotNull("Expecting WhitespaceTokenizer analysis breakdown", tokenList);
Index: solr/src/java/org/apache/solr/analysis/PatternReplaceCharFilter.java
===================================================================
--- solr/src/java/org/apache/solr/analysis/PatternReplaceCharFilter.java	(revision 940668)
+++ solr/src/java/org/apache/solr/analysis/PatternReplaceCharFilter.java	(working copy)
@@ -22,7 +22,7 @@
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-import org.apache.lucene.analysis.BaseCharFilter;
+import org.apache.lucene.analysis.charfilter.BaseCharFilter;
 import org.apache.lucene.analysis.CharStream;
 
 /**
Index: solr/src/java/org/apache/solr/analysis/MappingCharFilterFactory.java
===================================================================
--- solr/src/java/org/apache/solr/analysis/MappingCharFilterFactory.java	(revision 940668)
+++ solr/src/java/org/apache/solr/analysis/MappingCharFilterFactory.java	(working copy)
@@ -25,8 +25,8 @@
 import java.util.regex.Pattern;
 
 import org.apache.lucene.analysis.CharStream;
-import org.apache.lucene.analysis.MappingCharFilter;
-import org.apache.lucene.analysis.NormalizeCharMap;
+import org.apache.lucene.analysis.charfilter.MappingCharFilter;
+import org.apache.lucene.analysis.charfilter.NormalizeCharMap;
 import org.apache.solr.common.ResourceLoader;
 import org.apache.solr.common.util.StrUtils;
 import org.apache.solr.util.plugin.ResourceLoaderAware;
Index: solr/src/java/org/apache/solr/analysis/HTMLStripCharFilter.java
===================================================================
--- solr/src/java/org/apache/solr/analysis/HTMLStripCharFilter.java	(revision 940668)
+++ solr/src/java/org/apache/solr/analysis/HTMLStripCharFilter.java	(working copy)
@@ -24,7 +24,7 @@
 import java.util.HashMap;
 import java.util.Set;
 
-import org.apache.lucene.analysis.BaseCharFilter;
+import org.apache.lucene.analysis.charfilter.BaseCharFilter;
 import org.apache.lucene.analysis.CharReader;
 import org.apache.lucene.analysis.CharStream;
 
Index: lucene/CHANGES.txt
===================================================================
--- lucene/CHANGES.txt	(revision 940668)
+++ lucene/CHANGES.txt	(working copy)
@@ -11,6 +11,9 @@
   - o.a.l.analysis.LengthFilter -> o.a.l.analysis.miscellaneous.LengthFilter
   - o.a.l.analysis.PerFieldAnalyzerWrapper -> o.a.l.analysis.miscellaneous.PerFieldAnalyzerWrapper
   - o.a.l.analysis.TeeSinkTokenFilter -> o.a.l.analysis.sinks.TeeSinkTokenFilter
+  - o.a.l.analysis.BaseCharFilter -> o.a.l.analysis.charfilter.BaseCharFilter
+  - o.a.l.analysis.MappingCharFilter -> o.a.l.analysis.charfilter.MappingCharFilter
+  - o.a.l.analysis.NormalizeCharMap -> o.a.l.analysis.charfilter.NormalizeCharMap
   ... (in progress)
 
 * LUCENE-1458, LUCENE-2111, LUCENE-2354: Changes from flexible indexing:
Index: lucene/src/test/org/apache/lucene/analysis/TestMappingCharFilter.java (deleted)
===================================================================
Index: lucene/src/java/org/apache/lucene/analysis/MappingCharFilter.java (deleted)
===================================================================
Index: lucene/src/java/org/apache/lucene/analysis/BaseCharFilter.java (deleted)
===================================================================
Index: lucene/src/java/org/apache/lucene/analysis/NormalizeCharMap.java (deleted)
===================================================================
Index: lucene/contrib/analyzers/common/src/test/org/apache/lucene/analysis/charfilter/TestMappingCharFilter.java
===================================================================
--- lucene/contrib/analyzers/common/src/test/org/apache/lucene/analysis/charfilter/TestMappingCharFilter.java	(revision 0)
+++ lucene/contrib/analyzers/common/src/test/org/apache/lucene/analysis/charfilter/TestMappingCharFilter.java	(working copy)
@@ -15,10 +15,16 @@
  * limitations under the License.
  */
 
-package org.apache.lucene.analysis;
+package org.apache.lucene.analysis.charfilter;
 
 import java.io.StringReader;
 
+import org.apache.lucene.analysis.BaseTokenStreamTestCase;
+import org.apache.lucene.analysis.CharReader;
+import org.apache.lucene.analysis.CharStream;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.WhitespaceTokenizer;
+
 public class TestMappingCharFilter extends BaseTokenStreamTestCase {
 
   NormalizeCharMap normMap;
Index: lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/charfilter/MappingCharFilter.java
===================================================================
--- lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/charfilter/MappingCharFilter.java	(revision 0)
+++ lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/charfilter/MappingCharFilter.java	(working copy)
@@ -15,12 +15,15 @@
  * limitations under the License.
  */
 
-package org.apache.lucene.analysis;
+package org.apache.lucene.analysis.charfilter;
 
 import java.io.IOException;
 import java.io.Reader;
 import java.util.LinkedList;
 
+import org.apache.lucene.analysis.CharReader;
+import org.apache.lucene.analysis.CharStream;
+
 /**
  * Simplistic {@link CharFilter} that applies the mappings
  * contained in a {@link NormalizeCharMap} to the character
Index: lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/charfilter/NormalizeCharMap.java
===================================================================
--- lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/charfilter/NormalizeCharMap.java	(revision 0)
+++ lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/charfilter/NormalizeCharMap.java	(working copy)
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-package org.apache.lucene.analysis;
+package org.apache.lucene.analysis.charfilter;
 
 import java.util.HashMap;
 import java.util.Map;
Index: lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/charfilter/BaseCharFilter.java
===================================================================
--- lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/charfilter/BaseCharFilter.java	(revision 0)
+++ lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/charfilter/BaseCharFilter.java	(working copy)
@@ -15,11 +15,14 @@
  * limitations under the License.
  */
 
-package org.apache.lucene.analysis;
+package org.apache.lucene.analysis.charfilter;
 
 import java.util.ArrayList;
 import java.util.List;
 
+import org.apache.lucene.analysis.CharFilter;
+import org.apache.lucene.analysis.CharStream;
+
 /**
  * Base utility class for implementing a {@link CharFilter}.
  * You subclass this, and then record mappings by calling
Index: lucene/backwards/src/test/org/apache/lucene/analysis/TestMappingCharFilter.java (deleted)
===================================================================
Index: lucene/backwards/src/java/org/apache/lucene/analysis/MappingCharFilter.java (deleted)
===================================================================
Index: lucene/backwards/src/java/org/apache/lucene/analysis/BaseCharFilter.java (deleted)
===================================================================
Index: lucene/backwards/src/java/org/apache/lucene/analysis/NormalizeCharMap.java (deleted)
===================================================================
