diff --git a/lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/GraphvizFormatter.java b/lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/GraphvizFormatter.java
index de3748b2666..2056f63a761 100644
--- a/lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/GraphvizFormatter.java
+++ b/lucene/analysis/kuromoji/src/java/org/apache/lucene/analysis/ja/GraphvizFormatter.java
@@ -158,6 +158,10 @@ public class GraphvizFormatter {
           sb.append('+');
         }
         sb.append(bgCost);
+        sb.append(' ');
+        sb.append('(');
+        sb.append(dict.getPartOfSpeech(posData.backID[idx]));
+        sb.append(')');
         sb.append("\"");
         sb.append(attrs);
         sb.append("]\n");
diff --git a/lucene/analysis/nori/src/java/org/apache/lucene/analysis/ko/GraphvizFormatter.java b/lucene/analysis/nori/src/java/org/apache/lucene/analysis/ko/GraphvizFormatter.java
index d8e3f7a9faf..1b1685a3096 100644
--- a/lucene/analysis/nori/src/java/org/apache/lucene/analysis/ko/GraphvizFormatter.java
+++ b/lucene/analysis/nori/src/java/org/apache/lucene/analysis/ko/GraphvizFormatter.java
@@ -158,6 +158,11 @@ public class GraphvizFormatter {
           sb.append('+');
         }
         sb.append(bgCost);
+        sb.append(' ');
+        sb.append(dict.getLeftPOS(posData.backID[idx]));
+        sb.append('(');
+        sb.append(dict.getLeftPOS(posData.backID[idx]).description());
+        sb.append(')');
         sb.append("\"");
         sb.append(attrs);
         sb.append("]\n");
