--- /home/angsaa/Downloads/lucene-4.2.0 src-original/join/src/java/org/apache/lucene/search/join/ToChildBlockJoinQuery.java
+++ /home/angsaa/Downloads/lucene-4.2.0 (2)/join/src/java/org/apache/lucene/search/join/ToChildBlockJoinQuery.java
@@ -216,6 +216,13 @@
               childDoc = NO_MORE_DOCS;
               //System.out.println("  END");
               return childDoc;
+            }
+
+            if (acceptDocs != null) {
+	          System.out.println("parent doc " + parentDoc + " is alive: " + acceptDocs.get(parentDoc));
+              if (!acceptDocs.get(parentDoc)) {
+                continue;
+              }
             }
 
             childDoc = 1 + parentBits.prevSetBit(parentDoc-1);
