Index: lucene/expressions/src/java/org/apache/lucene/expressions/js/JavascriptCompiler.java
===================================================================
--- lucene/expressions/src/java/org/apache/lucene/expressions/js/JavascriptCompiler.java	(revision 1694896)
+++ lucene/expressions/src/java/org/apache/lucene/expressions/js/JavascriptCompiler.java	(working copy)
@@ -83,7 +83,7 @@
     }
   }
   
-  private static final int CLASSFILE_VERSION = Opcodes.V1_7;
+  private static final int CLASSFILE_VERSION = Opcodes.V1_8;
   
   // We use the same class name for all generated classes as they all have their own class loader.
   // The source code is displayed as "source file name" in stack trace.
Index: lucene/ivy-versions.properties
===================================================================
--- lucene/ivy-versions.properties	(revision 1694896)
+++ lucene/ivy-versions.properties	(working copy)
@@ -278,7 +278,7 @@
 /org.noggit/noggit = 0.6
 /org.objenesis/objenesis = 1.2
 
-org.ow2.asm.version = 4.1
+org.ow2.asm.version = 5.0.4
 /org.ow2.asm/asm = ${org.ow2.asm.version}
 /org.ow2.asm/asm-commons = ${org.ow2.asm.version}
 
Index: lucene/licenses/asm-4.1.jar.sha1
===================================================================
--- lucene/licenses/asm-4.1.jar.sha1	(revision 1694896)
+++ lucene/licenses/asm-4.1.jar.sha1	(working copy)
@@ -1 +0,0 @@
-ad568238ee36a820bd6c6806807e8a14ea34684d
Index: lucene/licenses/asm-5.0.4.jar.sha1
===================================================================
--- lucene/licenses/asm-5.0.4.jar.sha1	(revision 0)
+++ lucene/licenses/asm-5.0.4.jar.sha1	(working copy)
@@ -0,0 +1 @@
+0da08b8cce7bbf903602a25a3a163ae252435795
Index: lucene/licenses/asm-5.0.4.jar.sha1
===================================================================
--- lucene/licenses/asm-5.0.4.jar.sha1	(revision 0)
+++ lucene/licenses/asm-5.0.4.jar.sha1	(working copy)

Property changes on: lucene/licenses/asm-5.0.4.jar.sha1
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Index: lucene/licenses/asm-commons-4.1.jar.sha1
===================================================================
--- lucene/licenses/asm-commons-4.1.jar.sha1	(revision 1694896)
+++ lucene/licenses/asm-commons-4.1.jar.sha1	(working copy)
@@ -1 +0,0 @@
-f8b86f4ee6e02082f63a658e00eb5506821253c6
Index: lucene/licenses/asm-commons-5.0.4.jar.sha1
===================================================================
--- lucene/licenses/asm-commons-5.0.4.jar.sha1	(revision 0)
+++ lucene/licenses/asm-commons-5.0.4.jar.sha1	(working copy)
@@ -0,0 +1 @@
+5a556786086c23cd689a0328f8519db93821c04c
Index: lucene/licenses/asm-commons-5.0.4.jar.sha1
===================================================================
--- lucene/licenses/asm-commons-5.0.4.jar.sha1	(revision 0)
+++ lucene/licenses/asm-commons-5.0.4.jar.sha1	(working copy)

Property changes on: lucene/licenses/asm-commons-5.0.4.jar.sha1
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Index: solr/licenses/asm-4.1.jar.sha1
===================================================================
--- solr/licenses/asm-4.1.jar.sha1	(revision 1694896)
+++ solr/licenses/asm-4.1.jar.sha1	(working copy)
@@ -1 +0,0 @@
-ad568238ee36a820bd6c6806807e8a14ea34684d
Index: solr/licenses/asm-5.0.4.jar.sha1
===================================================================
--- solr/licenses/asm-5.0.4.jar.sha1	(revision 0)
+++ solr/licenses/asm-5.0.4.jar.sha1	(working copy)
@@ -0,0 +1 @@
+0da08b8cce7bbf903602a25a3a163ae252435795
Index: solr/licenses/asm-5.0.4.jar.sha1
===================================================================
--- solr/licenses/asm-5.0.4.jar.sha1	(revision 0)
+++ solr/licenses/asm-5.0.4.jar.sha1	(working copy)

Property changes on: solr/licenses/asm-5.0.4.jar.sha1
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
Index: solr/licenses/asm-commons-4.1.jar.sha1
===================================================================
--- solr/licenses/asm-commons-4.1.jar.sha1	(revision 1694896)
+++ solr/licenses/asm-commons-4.1.jar.sha1	(working copy)
@@ -1 +0,0 @@
-f8b86f4ee6e02082f63a658e00eb5506821253c6
Index: solr/licenses/asm-commons-5.0.4.jar.sha1
===================================================================
--- solr/licenses/asm-commons-5.0.4.jar.sha1	(revision 0)
+++ solr/licenses/asm-commons-5.0.4.jar.sha1	(working copy)
@@ -0,0 +1 @@
+5a556786086c23cd689a0328f8519db93821c04c
Index: solr/licenses/asm-commons-5.0.4.jar.sha1
===================================================================
--- solr/licenses/asm-commons-5.0.4.jar.sha1	(revision 0)
+++ solr/licenses/asm-commons-5.0.4.jar.sha1	(working copy)

Property changes on: solr/licenses/asm-commons-5.0.4.jar.sha1
___________________________________________________________________
Added: svn:eol-style
## -0,0 +1 ##
+native
\ No newline at end of property
