Index: src/java/org/apache/lucene/index/IndexWriter.java
===================================================================
--- src/java/org/apache/lucene/index/IndexWriter.java	(revision 783557)
+++ src/java/org/apache/lucene/index/IndexWriter.java	(working copy)
@@ -4426,6 +4426,7 @@
             message("now merge\n  merge=" + merge.segString(directory) + "\n  merge=" + merge + "\n  index=" + segString());
 
           mergeMiddle(merge);
+          mergeSuccess(merge);
           success = true;
         } catch (Throwable t) {
           handleMergeException(t, merge);
@@ -4452,7 +4453,10 @@
       handleOOM(oom, "merge");
     }
   }
-
+  
+  void mergeSuccess(MergePolicy.OneMerge merge) {
+  }
+  
   /** Checks whether this merge involves any segments
    *  already participating in a merge.  If not, this merge
    *  is "registered", meaning we record that its segments
