Index: lucene/src/test-framework/java/org/apache/lucene/search/AssertingIndexSearcher.java
===================================================================
--- lucene/src/test-framework/java/org/apache/lucene/search/AssertingIndexSearcher.java	(revision 1203031)
+++ lucene/src/test-framework/java/org/apache/lucene/search/AssertingIndexSearcher.java	(working copy)
@@ -76,7 +76,17 @@
       @Override
       public Scorer scorer(AtomicReaderContext context, boolean scoreDocsInOrder,
           boolean topScorer, Bits acceptDocs) throws IOException {
-        return w.scorer(context, scoreDocsInOrder, topScorer, acceptDocs);
+        Scorer scorer = w.scorer(context, scoreDocsInOrder, topScorer, acceptDocs);
+        if (scorer != null) {
+          // check that scorer obeys disi contract for docID() before next()/advance
+          try {
+            int docid = scorer.docID();
+            assert docid == -1 || docid == DocIdSetIterator.NO_MORE_DOCS;
+          } catch (UnsupportedOperationException ignored) {
+            // from a top-level BS1
+          }
+        }
+        return scorer;
       }
 
       @Override
