Index: lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java	(revision 1572530)
+++ lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java	(working copy)
@@ -98,11 +98,21 @@
   public FacetsConfig() {
   }
 
+  /** Get the default configuration for new dimensions. Useful when
+   *  the dimension is not know on beforehand and may need different 
+   *  default settings, like {@code multivalue = true}.
+   * @return The default configuration to be used for dimensions that 
+   * are not yet used in the {@link FacetsConfig} 
+   */
+  protected DimConfig getDefaultDimConfig(){
+    return DEFAULT_DIM_CONFIG;
+  }
+  
   /** Get the current configuration for a dimension. */
   public synchronized DimConfig getDimConfig(String dimName) {
     DimConfig ft = fieldTypes.get(dimName);
     if (ft == null) {
-      ft = DEFAULT_DIM_CONFIG;
+      ft = getDefaultDimConfig();
     }
     return ft;
   }
