Index: java/org/apache/lucene/store/FSDirectory.java
===================================================================
RCS file: /home/cvspublic/jakarta-lucene/src/java/org/apache/lucene/store/FSDirectory.java,v
retrieving revision 1.30
diff -u -r1.30 FSDirectory.java
--- java/org/apache/lucene/store/FSDirectory.java	11 May 2004 20:01:48 -0000	1.30
+++ java/org/apache/lucene/store/FSDirectory.java	18 May 2004 12:54:42 -0000
@@ -122,6 +122,11 @@
    if (!lockDir.isAbsolute()) {
      lockDir = new File(directory, LOCK_DIR);
    }
+   if (lockDir.exists() == false) {
+       if (lockDir.mkdirs() == false) {
+           throw new IOException("Cannot create temporary directory: " + lockDir);
+       }
+   }
     if (create)
       create();
 
