Index: lucene/contrib/CHANGES.txt
===================================================================
--- lucene/contrib/CHANGES.txt	(revision 949729)
+++ lucene/contrib/CHANGES.txt	(working copy)
@@ -86,6 +86,9 @@
    such as stopword lists and any language-specific processing in addition to
    stemming. Add Turkish and Romanian stopwords lists to support this.
    (Robert Muir, Uwe Schindler, Simon Willnauer)
+
+ * LUCENE-2477: Deprecate MoreLikeThis's DEFAULT_ANALYZER. If you are using stored
+   fields, explicitly set your analyzer instead.  (Robert Muir)
    
 New features
 
Index: lucene/contrib/queries/src/java/org/apache/lucene/search/similar/MoreLikeThis.java
===================================================================
--- lucene/contrib/queries/src/java/org/apache/lucene/search/similar/MoreLikeThis.java	(revision 949729)
+++ lucene/contrib/queries/src/java/org/apache/lucene/search/similar/MoreLikeThis.java	(working copy)
@@ -162,7 +162,10 @@
 	/**
      * Default analyzer to parse source doc with.
 	 * @see #getAnalyzer
+	 * @deprecated This default will be removed in Lucene 4.0 (with the default being null).
+	 * If you are using stored fields, explicitly set your analyzer instead.
      */
+    @Deprecated
     public static final Analyzer DEFAULT_ANALYZER = new StandardAnalyzer(Version.LUCENE_CURRENT);
 
     /**
