Index: lucene/src/test/org/apache/lucene/search/TestFieldCache.java
===================================================================
--- lucene/src/test/org/apache/lucene/search/TestFieldCache.java	(revision 1158642)
+++ lucene/src/test/org/apache/lucene/search/TestFieldCache.java	(working copy)
@@ -277,10 +277,10 @@
         s = unicodeStrings[random.nextInt(i)];
       }
       if (s == null) {
-        s = _TestUtil.randomUnicodeString(random, 250);
+        s = _TestUtil.randomUnicodeString(random);
       }
     } else {
-      s = _TestUtil.randomUnicodeString(random, 250);
+      s = _TestUtil.randomUnicodeString(random);
     }
     return s;
   }
