diff --git a/lucene/misc/src/java/org/apache/lucene/uninverting/FieldCache.java b/lucene/misc/src/java/org/apache/lucene/uninverting/FieldCache.java
index 27d68e04abb067eadd9f271e496c5791aa195f34..11185a903e9152fc5febee028b8e6537c740b7c5 100644
--- a/lucene/misc/src/java/org/apache/lucene/uninverting/FieldCache.java
+++ b/lucene/misc/src/java/org/apache/lucene/uninverting/FieldCache.java
@@ -404,7 +404,7 @@ interface FieldCache {
     
     @Override
     public String toString() {
-      StringBuilder b = new StringBuilder();
+      StringBuilder b = new StringBuilder(250);
       b.append("'").append(getReaderKey()).append("'=>");
       b.append("'").append(getFieldName()).append("',");
       b.append(getCacheType()).append(",").append(getCustom());
