Index: lucene/core/src/java/org/apache/lucene/index/StorableFieldType.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/index/StorableFieldType.java	(revision 1452116)
+++ lucene/core/src/java/org/apache/lucene/index/StorableFieldType.java	(working copy)
@@ -1,33 +0,0 @@
-package org.apache.lucene.index;
-
-import org.apache.lucene.index.FieldInfo.DocValuesType;
-
-/*
- * Licensed to the Apache Software Foundation (ASF) under one or more
- * contributor license agreements.  See the NOTICE file distributed with
- * this work for additional information regarding copyright ownership.
- * The ASF licenses this file to You under the Apache License, Version 2.0
- * (the "License"); you may not use this file except in compliance with
- * the License.  You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-
-// TODO: Move some properties from IndexableFieldType here, those regarding stored fields. 
-
-/** 
- * Describes the properties of a stored field.
- * @lucene.experimental 
- */
-public interface StorableFieldType {
-
-  /** DocValues type; if non-null then the field's value
-   *  will be indexed into docValues */
-  public DocValuesType docValueType();
-}
Index: lucene/core/src/java/org/apache/lucene/document/NumericDocValuesField.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/document/NumericDocValuesField.java	(revision 1452116)
+++ lucene/core/src/java/org/apache/lucene/document/NumericDocValuesField.java	(working copy)
@@ -33,7 +33,7 @@
  * separate {@link StoredField} instance.
  * */
 
-public class NumericDocValuesField extends StoredField {
+public class NumericDocValuesField extends Field {
 
   /**
    * Type for numeric DocValues.
Index: lucene/core/src/java/org/apache/lucene/document/BinaryDocValuesField.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/document/BinaryDocValuesField.java	(revision 1452116)
+++ lucene/core/src/java/org/apache/lucene/document/BinaryDocValuesField.java	(working copy)
@@ -39,7 +39,7 @@
  * 
  * @see BinaryDocValues
  * */
-public class BinaryDocValuesField extends StoredField {
+public class BinaryDocValuesField extends Field {
   
   /**
    * Type for straight bytes DocValues.
Index: lucene/core/src/java/org/apache/lucene/document/SortedDocValuesField.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/document/SortedDocValuesField.java	(revision 1452116)
+++ lucene/core/src/java/org/apache/lucene/document/SortedDocValuesField.java	(working copy)
@@ -36,7 +36,7 @@
  * 
  * */
 
-public class SortedDocValuesField extends StoredField {
+public class SortedDocValuesField extends Field {
 
   /**
    * Type for sorted bytes DocValues
Index: lucene/core/src/java/org/apache/lucene/document/SortedSetDocValuesField.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/document/SortedSetDocValuesField.java	(revision 1452116)
+++ lucene/core/src/java/org/apache/lucene/document/SortedSetDocValuesField.java	(working copy)
@@ -37,7 +37,7 @@
  * 
  * */
 
-public class SortedSetDocValuesField extends StoredField {
+public class SortedSetDocValuesField extends Field {
 
   /**
    * Type for sorted bytes DocValues
