Index: lucene/core/src/java/org/apache/lucene/util/FixedBitSet.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/util/FixedBitSet.java	(revision 1630741)
+++ lucene/core/src/java/org/apache/lucene/util/FixedBitSet.java	(working copy)
@@ -59,24 +59,7 @@
     
     @Override
     public int nextDoc() {
-      if (doc == NO_MORE_DOCS || ++doc >= numBits) {
-        return doc = NO_MORE_DOCS;
-      }
-      int i = doc >> 6;
-      long word = bits[i] >> doc;  // skip all the bits to the right of index
-      
-      if (word != 0) {
-        return doc = doc + Long.numberOfTrailingZeros(word);
-      }
-      
-      while (++i < numWords) {
-        word = bits[i];
-        if (word != 0) {
-          return doc = (i << 6) + Long.numberOfTrailingZeros(word);
-        }
-      }
-      
-      return doc = NO_MORE_DOCS;
+      return advance(doc + 1);
     }
     
     @Override
@@ -91,7 +74,7 @@
     
     @Override
     public int advance(int target) {
-      if (doc == NO_MORE_DOCS || target >= numBits) {
+      if (target >= numBits) {
         return doc = NO_MORE_DOCS;
       }
       int i = target >> 6;
Index: lucene/core/src/java/org/apache/lucene/util/NotDocIdSet.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/util/NotDocIdSet.java	(revision 1630741)
+++ lucene/core/src/java/org/apache/lucene/util/NotDocIdSet.java	(working copy)
@@ -82,9 +82,6 @@
 
       @Override
       public int nextDoc() throws IOException {
-        if (doc == NO_MORE_DOCS) {
-          return NO_MORE_DOCS;
-        }
         return advance(doc + 1);
       }
 
Index: lucene/core/src/java/org/apache/lucene/util/RoaringDocIdSet.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/util/RoaringDocIdSet.java	(revision 1630741)
+++ lucene/core/src/java/org/apache/lucene/util/RoaringDocIdSet.java	(working copy)
@@ -282,9 +282,6 @@
 
     @Override
     public int nextDoc() throws IOException {
-      if (doc == NO_MORE_DOCS) {
-        return NO_MORE_DOCS;
-      }
       final int subNext = sub.nextDoc();
       if (subNext == NO_MORE_DOCS) {
         return firstDocFromNextBlock();
Index: lucene/core/src/java/org/apache/lucene/util/SparseFixedBitSet.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/util/SparseFixedBitSet.java	(revision 1630741)
+++ lucene/core/src/java/org/apache/lucene/util/SparseFixedBitSet.java	(working copy)
@@ -244,14 +244,16 @@
 
     @Override
     public int nextDoc() throws IOException {
-      if (doc == NO_MORE_DOCS || ++doc >= length) {
+      return advance(doc + 1);
+    }
+
+    @Override
+    public int advance(int target) throws IOException {
+      if (target >= length) {
         return doc = NO_MORE_DOCS;
       }
-      return currentOrNextDoc();
-    }
 
-    private int currentOrNextDoc() {
-      final int i4096 = doc >>> 12;
+      final int i4096 = target >>> 12;
       final long index = indices[i4096];
       if (index == 0) {
         // if the index is zero, it means that there is no value in the
@@ -260,7 +262,7 @@
       } else {
         // now we are on a block that contains at least one document
         assert Long.bitCount(index) <= bits[i4096].length;
-        int i64 = doc >>> 6;
+        int i64 = target >>> 6;
         long indexBits = index >>> i64; // shifts are mod 64 in java
         if (indexBits == 0) {
           // if neither the i64-th bit or any other bit on its left is set then
@@ -271,7 +273,7 @@
 
         // We know we still have some 64-bits blocks that have bits set, let's
         // advance to the next one by skipping trailing zeros of the index
-        int i1 = doc & 0x3F;
+        int i1 = target & 0x3F;
         int trailingZeros = Long.numberOfTrailingZeros(indexBits);
         if (trailingZeros != 0) {
           // no bits in the current long, go to the next one
@@ -314,17 +316,6 @@
     }
 
     @Override
-    public int advance(int target) throws IOException {
-      if (target >= length) {
-        return doc = NO_MORE_DOCS;
-      } else {
-        doc = target;
-      }
-
-      return currentOrNextDoc();
-    }
-
-    @Override
     public long cost() {
       // although constant-time, approximateCardinality is a bit expensive so
       // we cache it to avoid performance traps eg. when sorting iterators by
