### Eclipse Workspace Patch 1.0
#P lucene_solr-lucene-3.5.0
Index: src/java/org/apache/lucene/index/IndexWriter.java
===================================================================
--- src/java/org/apache/lucene/index/IndexWriter.java	(revision 1212071)
+++ src/java/org/apache/lucene/index/IndexWriter.java	(working copy)
@@ -4291,6 +4291,8 @@
 
       merge.checkAborted(directory);
 
+      beforeMergeAfterSetup(merge);
+
       // This is where all the work happens:
       mergedDocCount = merge.info.docCount = merger.merge();
 
@@ -4430,6 +4432,14 @@
 
     return mergedDocCount;
   }
+  
+  /**
+   * A hook for extending classes to inject operations after
+   * the {@link MergePolicy.OneMerge} object has been set up
+   * but before the merge has actually begun
+   */
+  void beforeMergeAfterSetup(MergePolicy.OneMerge merge) throws IOException {
+  }
 
   synchronized void addMergeException(MergePolicy.OneMerge merge) {
     assert merge.getException() != null;
