Index: PerFieldAnalyzerWrapper.java
===================================================================
--- PerFieldAnalyzerWrapper.java	(revision 902745)
+++ PerFieldAnalyzerWrapper.java	(working copy)
@@ -118,6 +118,15 @@
       analyzer = defaultAnalyzer;
     return analyzer.getPositionIncrementGap(fieldName);
   }
+
+  /** Return the offsetGap from the analyzer assigned to field */
+  @Override
+  public int getOffsetGap(Fieldable field) {
+    Analyzer analyzer = analyzerMap.get(field.name());
+    if (analyzer == null)
+      analyzer = defaultAnalyzer;
+    return analyzer.getOffsetGap(field);
+  }
   
   @Override
   public String toString() {
