Index: lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestComplexPhraseSpanQuery.java
===================================================================
--- lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestComplexPhraseSpanQuery.java	(revision 1570225)
+++ lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestComplexPhraseSpanQuery.java	(working copy)
@@ -75,9 +75,11 @@
     // checkMatches("\"john -percival\"", "1"); // not logic doesn't work
     checkMatches("\"john percival\"!~2,2", "1"); // not logic doesn't work
 
-    //WAS: leave commented out until LUCENE-5450 is resolved
-//    checkMatches("\"john  nosuchword*\"", ""); // phrases with clauses producing
- //   checkMatches("\"john  nosuchw?rd\"!~2,3", ""); // phrases with clauses producing
+    checkMatches("\"john  nosuchword*\"", ""); // phrases with clauses producing
+    checkMatches("\"john  nosuchw?rd\"!~2,3", "1,3"); // phrases with clauses producing
+    checkMatches("\"nosuchw?rd john\"!~2,3", ""); // phrases with clauses producing
+    checkMatches("\"nosuchw?rd john\"", ""); // phrases with clauses producing
+
     // empty sets
     checkBadQuery("\"jo*  id:1 smith\""); // mixing fields in a phrase is bad
     //WAS:
Index: lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestSpanOnlyQueryParser.java
===================================================================
--- lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestSpanOnlyQueryParser.java	(revision 1570225)
+++ lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestSpanOnlyQueryParser.java	(working copy)
@@ -143,7 +143,7 @@
       writer.addDocument(doc);
     }
     reader = writer.getReader();
-    searcher = new IndexSearcher(reader);
+    searcher = newSearcher(reader);
     writer.close();
   }
 
Index: lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestOverallSpanQueryParser.java
===================================================================
--- lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestOverallSpanQueryParser.java	(revision 1570225)
+++ lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestOverallSpanQueryParser.java	(working copy)
@@ -110,7 +110,7 @@
       writer.addDocument(doc);
     }
     reader = writer.getReader();
-    searcher = new IndexSearcher(reader);
+    searcher = newSearcher(reader);
     writer.close();
 
     parser = new SpanQueryParser(TEST_VERSION_CURRENT, FIELD1, analyzer);
Index: lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestAdvancedAnalyzers.java
===================================================================
--- lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestAdvancedAnalyzers.java	(revision 1570225)
+++ lucene/queryparser/src/test/org/apache/lucene/queryparser/spans/TestAdvancedAnalyzers.java	(working copy)
@@ -145,7 +145,7 @@
       writer.addDocument(doc);
     }
     reader = writer.getReader();
-    searcher = new IndexSearcher(reader);
+    searcher = newSearcher(reader);
     writer.close();
   }
 
