diff --git a/solr/solrj/build.gradle b/solr/solrj/build.gradle
index e8c8a072d05..84c39ce1fe8 100644
--- a/solr/solrj/build.gradle
+++ b/solr/solrj/build.gradle
@@ -54,6 +54,10 @@ dependencies {
     exclude group: "log4j", module: "log4j"
     exclude group: "org.slf4j", module: "slf4j-log4j12"
   })
+  // Zookeeper classes reference this although it's not part of the exported classpath
+  // which causes odd warnings during compilation. Shut it up with an explicit-version
+  // compile-only dependency (!).
+  compileOnly 'com.google.code.findbugs:annotations:3.0.1'
 
   api('org.codehaus.woodstox:woodstox-core-asl', {
     exclude group: "javax.xml.stream", module: "stax-api"
