diff -u --recursive lucene-1.9-200511081055.orig/src/java/org/apache/lucene/search/Searcher.java lucene-1.9-200511081055.work/src/java/org/apache/lucene/search/Searcher.java
--- src/java/org/apache/lucene/search/Searcher.java	2005-10-05 22:14:56.000000000 +0200
+++ src/java/org/apache/lucene/search/Searcher.java	2005-11-09 08:22:18.000000000 +0100
@@ -19,6 +19,7 @@
 import java.io.IOException;
 
 import org.apache.lucene.index.Term;
+import org.apache.lucene.document.Document;
 
 /** An abstract base class for search implementations.
  * Implements the main search methods.
@@ -176,4 +177,20 @@
     }
     return result;
   }
+
+  /* It just seems that gcj doesn't place in abstract classes
+   * implementing an interface methods from the interface that are
+   * not explicitly declared in the abstract class.
+   */
+
+  abstract public void close() throws IOException;
+
+  abstract public int docFreq(Term term) throws IOException;
+
+  abstract public int maxDoc() throws IOException;
+
+  abstract public Document doc(int i) throws IOException;
+
+  abstract public Query rewrite(Query query) throws IOException;
+
 }
