diff --git a/lucene/CHANGES.txt b/lucene/CHANGES.txt
index b5335c8fce..f6958fd7a3 100644
--- a/lucene/CHANGES.txt
+++ b/lucene/CHANGES.txt
@@ -160,6 +160,9 @@ Bug Fixes:
   TokenStream if a token consisting entirely of delimiter characters was 
   encountered, but preserve_original was set. (Alan Woodward)
 
+
+* LUCENE-8398: TieredMergePolicy.getMaxMergedSegmentMB has rounding error (Erick Erickson)
+
 Changes in Runtime Behavior:
 
 * LUCENE-7976: TieredMergePolicy now respects maxSegmentSizeMB by default when executing
diff --git a/lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java b/lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java
index 7f917bba46..c60f578e77 100644
--- a/lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java
+++ b/lucene/core/src/java/org/apache/lucene/index/TieredMergePolicy.java
@@ -192,7 +192,7 @@ public class TieredMergePolicy extends MergePolicy {
    *
    * @see #setMaxMergedSegmentMB */
   public double getMaxMergedSegmentMB() {
-    return maxMergedSegmentBytes/1024/1024.;
+    return maxMergedSegmentBytes/1024.0/1024.0;
   }
 
   /** Controls how aggressively merges that reclaim more
