Index: FieldDoc.java
===================================================================
--- FieldDoc.java	(revision 495792)
+++ FieldDoc.java	(arbetskopia)
@@ -1,5 +1,7 @@
 package org.apache.lucene.search;
 
+import java.util.Arrays;
+
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
@@ -61,4 +63,21 @@
 		super (doc, score);
 		this.fields = fields;
 	}
+
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    if (!super.equals(o)) return false;
+
+    FieldDoc fieldDoc = (FieldDoc) o;
+
+    return Arrays.equals(fields, fieldDoc.fields);
+
+  }
+
+  public int hashCode() {
+    int result = super.hashCode();
+    result = 31 * result + (fields != null ? Arrays.hashCode(fields) : 0);
+    return result;
+  }
 }
\ Inget radslut vid filslut
Index: ScoreDoc.java
===================================================================
--- ScoreDoc.java	(revision 495792)
+++ ScoreDoc.java	(arbetskopia)
@@ -33,4 +33,23 @@
     this.doc = doc;
     this.score = score;
   }
+
+
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+
+    ScoreDoc scoreDoc = (ScoreDoc) o;
+
+    if (doc != scoreDoc.doc) return false;
+    return Float.compare(scoreDoc.score, score) == 0;
+
+  }
+
+  public int hashCode() {
+    int result;
+    result = score != +0.0f ? Float.floatToIntBits(score) : 0;
+    result = 31 * result + doc;
+    return result;
+  }
 }
