Index: src/java/org/apache/lucene/search/FieldCacheImpl.java
===================================================================
--- src/java/org/apache/lucene/search/FieldCacheImpl.java	(revision 488721)
+++ src/java/org/apache/lucene/search/FieldCacheImpl.java	(working copy)
@@ -62,7 +62,7 @@
         }
         if (value == null) {
           value = new CreationPlaceholder();
-          innerCache.put(reader, value);
+          innerCache.put(key, value);
         }
       }
       if (value instanceof CreationPlaceholder) {
