diff --git a/dev/trunk/lucene/src/test/org/apache/lucene/index/TestFilterIndexReader.java b/dev/trunk/lucene/src/test/org/apache/lucene/index/TestFilterIndexReader.java
index 00b7c4f..eee1ff6 100644
--- a/dev/trunk/lucene/src/test/org/apache/lucene/index/TestFilterIndexReader.java
+++ b/dev/trunk/lucene/src/test/org/apache/lucene/index/TestFilterIndexReader.java
@@ -113,7 +113,15 @@ public class TestFilterIndexReader extends LuceneTestCase {
 
     writer.close();
 
+    //IndexReader reader = new TestReader(IndexReader.open(directory, true));
+    RAMDirectory target = new MockRAMDirectory();
+    writer = new IndexWriter(target, new IndexWriterConfig(TEST_VERSION_CURRENT, new MockAnalyzer()));
     IndexReader reader = new TestReader(IndexReader.open(directory, true));
+    writer.addIndexes(reader);
+    writer.close();
+    reader.close();
+    reader = IndexReader.open(target, true);
+    
 
     assertTrue(reader.isOptimized());
     
