Index: D:/Development/workspaceexadel/lucene-dev/contrib/spellchecker/src/java/org/apache/lucene/search/spell/SpellChecker.java
===================================================================
--- D:/Development/workspaceexadel/lucene-dev/contrib/spellchecker/src/java/org/apache/lucene/search/spell/SpellChecker.java	(revision 530684)
+++ D:/Development/workspaceexadel/lucene-dev/contrib/spellchecker/src/java/org/apache/lucene/search/spell/SpellChecker.java	(working copy)
@@ -313,6 +313,18 @@
     // close writer
     writer.optimize();
     writer.close();
+    
+    //close the stale reader
+    if (null != reader)  {
+    	reader.close();
+    	reader = null;
+    }
+    
+    //reopen the stale searcher
+    if (searcher != null) {
+        searcher.close();
+      }
+      searcher = new IndexSearcher(this.spellIndex);
   }
 
   private int getMin(int l) {
