Index: src/java/org/apache/lucene/analysis/NormalizeCharMap.java
===================================================================
--- src/java/org/apache/lucene/analysis/NormalizeCharMap.java	(revision 826183)
+++ src/java/org/apache/lucene/analysis/NormalizeCharMap.java	(working copy)
@@ -27,7 +27,7 @@
 public class NormalizeCharMap {
 
   //Map<Character, NormalizeMap> submap;
-  Map submap;
+  Map<Character, NormalizeCharMap> submap;
   String normStr;
   int diff;
 
@@ -44,9 +44,9 @@
     for(int i = 0; i < singleMatch.length(); i++) {
       char c = singleMatch.charAt(i);
       if (currMap.submap == null) {
-        currMap.submap = new HashMap(1);
+        currMap.submap = new HashMap<Character, NormalizeCharMap>(1);
       }
-      NormalizeCharMap map = (NormalizeCharMap) currMap.submap.get(CharacterCache.valueOf(c));
+      NormalizeCharMap map = currMap.submap.get(CharacterCache.valueOf(c));
       if (map == null) {
         map = new NormalizeCharMap();
         currMap.submap.put(new Character(c), map);
