Index: lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoExactCircleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoExactCircleTest.java	(revision e1851c0ce76064d6b698a64257913ee35010e780)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoExactCircleTest.java	(revision )
@@ -232,4 +232,16 @@
     assertTrue(circle1.getRelationship(circle2) != GeoArea.DISJOINT);
   }
 
+ public void testLUCENE8080() {
+   PlanetModel planetModel = new PlanetModel(1.6304230055804751, 1.0199671157571204);
+   GeoCircle circle;
+   try {
+     circle = GeoCircleFactory.makeExactGeoCircle(planetModel, 0.8853814403571284, 0.9784990176851283, 0.9071033527030907, 1e-11);
+   } catch (IllegalArgumentException e) {
+     circle = GeoCircleFactory.makeExactGeoCircle(planetModel, 0.8853814403571284, 0.9784990176851283, 0.9071033527030907, 1e-7);
+   }
+   GeoPoint point = planetModel.surfacePointOnBearing(circle.getCenter(), 0.9 * circle.getRadius(), 0);
+   assertTrue(circle.isWithin(point));
+ }
+
 }
