Index: lucene/core/src/test/org/apache/lucene/search/TestSloppyPhraseQuery2.java
===================================================================
--- lucene/core/src/test/org/apache/lucene/search/TestSloppyPhraseQuery2.java	(revision 1292983)
+++ lucene/core/src/test/org/apache/lucene/search/TestSloppyPhraseQuery2.java	(working copy)
@@ -174,13 +174,16 @@
   
   /** MultiPhraseQuery~N ⊆ MultiPhraseQuery~N+1 */
   public void testRandomIncreasingSloppiness() throws Exception {
-    long seed = random.nextLong();
-    MultiPhraseQuery q1 = randomPhraseQuery(seed);
-    MultiPhraseQuery q2 = randomPhraseQuery(seed);
-    for (int i = 0; i < 10; i++) {
-      q1.setSlop(i);
-      q2.setSlop(i+1);
-      assertSubsetOf(q1, q2);
+    int num = atLeast(100);
+    for (int k = 0; k < num; k++) {
+      long seed = random.nextLong();
+      MultiPhraseQuery q1 = randomPhraseQuery(seed);
+      MultiPhraseQuery q2 = randomPhraseQuery(seed);
+      for (int i = 0; i < 10; i++) {
+        q1.setSlop(i);
+        q2.setSlop(i+1);
+        assertSubsetOf(q1, q2);
+      }
     }
   }
   
