Index: modules/queryparser/src/test/org/apache/lucene/queryparser/ext/TestExtensions.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/ext/TestExtensions.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/ext/TestExtensions.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.ext;
+package org.apache.lucene.queryparser.ext;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/ext/ExtensionStub.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/ext/ExtensionStub.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/ext/ExtensionStub.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.ext;
+package org.apache.lucene.queryparser.ext;
 
 import org.apache.lucene.index.Term;
 import org.apache.lucene.queryparser.classic.ParseException;
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/ext/TestExtendableQueryParser.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/ext/TestExtendableQueryParser.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/ext/TestExtendableQueryParser.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.ext;
+package org.apache.lucene.queryparser.ext;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/ExceptionQueryTst.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/ExceptionQueryTst.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/ExceptionQueryTst.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
@@ -17,8 +17,8 @@
  * limitations under the License.
  */
 
-import org.apache.lucene.queryParser.surround.parser.ParseException;
-import org.apache.lucene.queryParser.surround.parser.QueryParser;
+import org.apache.lucene.queryparser.surround.parser.ParseException;
+import org.apache.lucene.queryparser.surround.parser.QueryParser;
 
 
 public class ExceptionQueryTst {
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/BooleanQueryTst.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/BooleanQueryTst.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/BooleanQueryTst.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
@@ -25,7 +25,7 @@
 import org.apache.lucene.search.Scorer;
 import org.apache.lucene.search.Query;
 
-import org.apache.lucene.queryParser.surround.parser.QueryParser;
+import org.apache.lucene.queryparser.surround.parser.QueryParser;
 
 import org.junit.Assert;
 
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/Test02Boolean.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/Test02Boolean.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/Test02Boolean.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/Test01Exceptions.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/Test01Exceptions.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/Test01Exceptions.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/SingleFieldTestDb.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/SingleFieldTestDb.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/SingleFieldTestDb.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/Test03Distance.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/Test03Distance.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/surround/query/Test03Distance.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/complexPhrase/TestComplexPhraseQuery.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/complexPhrase/TestComplexPhraseQuery.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/complexPhrase/TestComplexPhraseQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.complexPhrase;
+package org.apache.lucene.queryparser.complexPhrase;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/test/org/apache/lucene/queryparser/analyzing/TestAnalyzingQueryParser.java
===================================================================
--- modules/queryparser/src/test/org/apache/lucene/queryparser/analyzing/TestAnalyzingQueryParser.java	(revision 0)
+++ modules/queryparser/src/test/org/apache/lucene/queryparser/analyzing/TestAnalyzingQueryParser.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.analyzing;
+package org.apache.lucene.queryparser.analyzing;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/ext/Extensions.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/ext/Extensions.java	(revision 0)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/ext/Extensions.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.ext;
+package org.apache.lucene.queryparser.ext;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/ext/ExtensionQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/ext/ExtensionQuery.java	(revision 0)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/ext/ExtensionQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.ext;
+package org.apache.lucene.queryparser.ext;
 
 import org.apache.lucene.queryparser.classic.QueryParser;
 
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/ext/ParserExtension.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/ext/ParserExtension.java	(revision 0)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/ext/ParserExtension.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.ext;
+package org.apache.lucene.queryparser.ext;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/ext/ExtendableQueryParser.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/ext/ExtendableQueryParser.java	(revision 0)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/ext/ExtendableQueryParser.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.ext;
+package org.apache.lucene.queryparser.ext;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
@@ -18,7 +18,7 @@
  */
 
 import org.apache.lucene.analysis.Analyzer;
-import org.apache.lucene.queryParser.ext.Extensions.Pair;
+import org.apache.lucene.queryparser.ext.Extensions.Pair;
 import org.apache.lucene.queryparser.classic.ParseException;
 import org.apache.lucene.queryparser.classic.QueryParser;
 import org.apache.lucene.search.Query;
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/ComposedQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/ComposedQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/ComposedQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/BasicQueryFactory.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/BasicQueryFactory.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/BasicQueryFactory.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/OrQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/OrQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/OrQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndTruncQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndTruncQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndTruncQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndPrefixQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndPrefixQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndPrefixQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/package.html
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/package.html	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/package.html	(working copy)
@@ -22,7 +22,7 @@
   <body>
     This package contains SrndQuery and its subclasses.
     <p/>
-    The parser in the org.apache.lucene.queryParser.surround.parser package
+    The parser in the org.apache.lucene.queryparser.surround.parser package
     normally generates a SrndQuery.
     <p/>
     For searching an org.apache.lucene.search.Query is provided by
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/DistanceQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/DistanceQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/DistanceQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SimpleTerm.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SimpleTerm.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SimpleTerm.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndTermQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndTermQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndTermQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/NotQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/NotQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/NotQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/TooManyBasicQueries.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/TooManyBasicQueries.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/TooManyBasicQueries.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/AndQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/AndQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/AndQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/DistanceSubQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/DistanceSubQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/DistanceSubQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndBooleanQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndBooleanQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SrndBooleanQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SpanNearClauseFactory.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SpanNearClauseFactory.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/SpanNearClauseFactory.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/FieldsQuery.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/FieldsQuery.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/query/FieldsQuery.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.surround.query;
+package org.apache.lucene.queryparser.surround.query;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParserConstants.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParserConstants.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParserConstants.java	(working copy)
@@ -1,5 +1,5 @@
 /* Generated By:JavaCC: Do not edit this line. QueryParserConstants.java */
-package org.apache.lucene.queryParser.surround.parser;
+package org.apache.lucene.queryparser.surround.parser;
 
 
 /**
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/Token.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/Token.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/Token.java	(working copy)
@@ -1,6 +1,6 @@
 /* Generated By:JavaCC: Do not edit this line. Token.java Version 4.1 */
 /* JavaCCOptions:TOKEN_EXTENDS=,KEEP_LINE_COL=null */
-package org.apache.lucene.queryParser.surround.parser;
+package org.apache.lucene.queryparser.surround.parser;
 
 /**
  * Describes the input token stream.
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/TokenMgrError.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/TokenMgrError.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/TokenMgrError.java	(working copy)
@@ -1,6 +1,6 @@
 /* Generated By:JavaCC: Do not edit this line. TokenMgrError.java Version 4.1 */
 /* JavaCCOptions: */
-package org.apache.lucene.queryParser.surround.parser;
+package org.apache.lucene.queryparser.surround.parser;
 
 /** Token Manager Error. */
 @SuppressWarnings("serial")
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParser.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParser.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParser.java	(working copy)
@@ -1,5 +1,5 @@
 /* Generated By:JavaCC: Do not edit this line. QueryParser.java */
-package org.apache.lucene.queryParser.surround.parser;
+package org.apache.lucene.queryparser.surround.parser;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -8,15 +8,15 @@
 
 import org.apache.lucene.analysis.TokenStream;
 
-import org.apache.lucene.queryParser.surround.query.SrndQuery;
-import org.apache.lucene.queryParser.surround.query.FieldsQuery;
-import org.apache.lucene.queryParser.surround.query.OrQuery;
-import org.apache.lucene.queryParser.surround.query.AndQuery;
-import org.apache.lucene.queryParser.surround.query.NotQuery;
-import org.apache.lucene.queryParser.surround.query.DistanceQuery;
-import org.apache.lucene.queryParser.surround.query.SrndTermQuery;
-import org.apache.lucene.queryParser.surround.query.SrndPrefixQuery;
-import org.apache.lucene.queryParser.surround.query.SrndTruncQuery;
+import org.apache.lucene.queryparser.surround.query.SrndQuery;
+import org.apache.lucene.queryparser.surround.query.FieldsQuery;
+import org.apache.lucene.queryparser.surround.query.OrQuery;
+import org.apache.lucene.queryparser.surround.query.AndQuery;
+import org.apache.lucene.queryparser.surround.query.NotQuery;
+import org.apache.lucene.queryparser.surround.query.DistanceQuery;
+import org.apache.lucene.queryparser.surround.query.SrndTermQuery;
+import org.apache.lucene.queryparser.surround.query.SrndPrefixQuery;
+import org.apache.lucene.queryparser.surround.query.SrndTruncQuery;
 
 /**
  * This class is generated by JavaCC.  The only method that clients should need
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParser.jj
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParser.jj	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParser.jj	(working copy)
@@ -28,7 +28,7 @@
 
 PARSER_BEGIN(QueryParser)
 
-package org.apache.lucene.queryParser.surround.parser;
+package org.apache.lucene.queryparser.surround.parser;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -37,15 +37,15 @@
 
 import org.apache.lucene.analysis.TokenStream;
 
-import org.apache.lucene.queryParser.surround.query.SrndQuery;
-import org.apache.lucene.queryParser.surround.query.FieldsQuery;
-import org.apache.lucene.queryParser.surround.query.OrQuery;
-import org.apache.lucene.queryParser.surround.query.AndQuery;
-import org.apache.lucene.queryParser.surround.query.NotQuery;
-import org.apache.lucene.queryParser.surround.query.DistanceQuery;
-import org.apache.lucene.queryParser.surround.query.SrndTermQuery;
-import org.apache.lucene.queryParser.surround.query.SrndPrefixQuery;
-import org.apache.lucene.queryParser.surround.query.SrndTruncQuery;
+import org.apache.lucene.queryparser.surround.query.SrndQuery;
+import org.apache.lucene.queryparser.surround.query.FieldsQuery;
+import org.apache.lucene.queryparser.surround.query.OrQuery;
+import org.apache.lucene.queryparser.surround.query.AndQuery;
+import org.apache.lucene.queryparser.surround.query.NotQuery;
+import org.apache.lucene.queryparser.surround.query.DistanceQuery;
+import org.apache.lucene.queryparser.surround.query.SrndTermQuery;
+import org.apache.lucene.queryparser.surround.query.SrndPrefixQuery;
+import org.apache.lucene.queryparser.surround.query.SrndTruncQuery;
 
 /**
  * This class is generated by JavaCC.  The only method that clients should need
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/FastCharStream.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/FastCharStream.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/FastCharStream.java	(working copy)
@@ -1,5 +1,5 @@
 // FastCharStream.java
-package org.apache.lucene.queryParser.surround.parser;
+package org.apache.lucene.queryparser.surround.parser;
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/CharStream.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/CharStream.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/CharStream.java	(working copy)
@@ -1,6 +1,6 @@
 /* Generated By:JavaCC: Do not edit this line. CharStream.java Version 4.1 */
 /* JavaCCOptions:STATIC=false */
-package org.apache.lucene.queryParser.surround.parser;
+package org.apache.lucene.queryparser.surround.parser;
 
 /**
  * This interface describes a character stream that maintains line and
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/ParseException.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/ParseException.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/ParseException.java	(working copy)
@@ -1,6 +1,6 @@
 /* Generated By:JavaCC: Do not edit this line. ParseException.java Version 4.1 */
 /* JavaCCOptions:KEEP_LINE_COL=null */
-package org.apache.lucene.queryParser.surround.parser;
+package org.apache.lucene.queryparser.surround.parser;
 
 /**
  * This exception is thrown when parse errors are encountered.
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParserTokenManager.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParserTokenManager.java	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/QueryParserTokenManager.java	(working copy)
@@ -1,18 +1,18 @@
 /* Generated By:JavaCC: Do not edit this line. QueryParserTokenManager.java */
-package org.apache.lucene.queryParser.surround.parser;
+package org.apache.lucene.queryparser.surround.parser;
 import java.util.ArrayList;
 import java.util.List;
 import java.io.StringReader;
 import org.apache.lucene.analysis.TokenStream;
-import org.apache.lucene.queryParser.surround.query.SrndQuery;
-import org.apache.lucene.queryParser.surround.query.FieldsQuery;
-import org.apache.lucene.queryParser.surround.query.OrQuery;
-import org.apache.lucene.queryParser.surround.query.AndQuery;
-import org.apache.lucene.queryParser.surround.query.NotQuery;
-import org.apache.lucene.queryParser.surround.query.DistanceQuery;
-import org.apache.lucene.queryParser.surround.query.SrndTermQuery;
-import org.apache.lucene.queryParser.surround.query.SrndPrefixQuery;
-import org.apache.lucene.queryParser.surround.query.SrndTruncQuery;
+import org.apache.lucene.queryparser.surround.query.SrndQuery;
+import org.apache.lucene.queryparser.surround.query.FieldsQuery;
+import org.apache.lucene.queryparser.surround.query.OrQuery;
+import org.apache.lucene.queryparser.surround.query.AndQuery;
+import org.apache.lucene.queryparser.surround.query.NotQuery;
+import org.apache.lucene.queryparser.surround.query.DistanceQuery;
+import org.apache.lucene.queryparser.surround.query.SrndTermQuery;
+import org.apache.lucene.queryparser.surround.query.SrndPrefixQuery;
+import org.apache.lucene.queryparser.surround.query.SrndTruncQuery;
 
 /** Token Manager. */
 public class QueryParserTokenManager implements QueryParserConstants
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/package.html
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/package.html	(revision 1145025)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/surround/parser/package.html	(working copy)
@@ -23,6 +23,6 @@
     This package contains the QueryParser.jj source file for the Surround parser.
     <p/>
     Parsing the text of a query results in a SrndQuery in the
-    org.apache.lucene.queryParser.surround.query package.
+    org.apache.lucene.queryparser.surround.query package.
   </body>
 </html>
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/complexPhrase/ComplexPhraseQueryParser.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/complexPhrase/ComplexPhraseQueryParser.java	(revision 0)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/complexPhrase/ComplexPhraseQueryParser.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.complexPhrase;
+package org.apache.lucene.queryparser.complexPhrase;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/src/java/org/apache/lucene/queryparser/analyzing/AnalyzingQueryParser.java
===================================================================
--- modules/queryparser/src/java/org/apache/lucene/queryparser/analyzing/AnalyzingQueryParser.java	(revision 0)
+++ modules/queryparser/src/java/org/apache/lucene/queryparser/analyzing/AnalyzingQueryParser.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.queryParser.analyzing;
+package org.apache.lucene.queryparser.analyzing;
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: modules/queryparser/build.xml
===================================================================
--- modules/queryparser/build.xml	(revision 1145029)
+++ modules/queryparser/build.xml	(working copy)
@@ -33,10 +33,13 @@
       <fileset dir="src/java/org/apache/lucene/queryparser/classic" includes="*.java">
         <containsregexp expression="Generated.*By.*JavaCC"/>
       </fileset>
+      <fileset dir="src/java/org/apache/lucene/queryparser/surround/parser" includes="*.java">
+        <containsregexp expression="Generated.*By.*JavaCC"/>
+      </fileset>
     </delete>
   </target>
 
-  <target name="javacc" depends="init,javacc-check,clean-javacc,javacc-QueryParser"/>
+  <target name="javacc" depends="init,javacc-check,clean-javacc,javacc-QueryParser,javacc-surround"/>
 
   <target name="javacc-QueryParser" depends="init,javacc-check" if="javacc.present">
     <sequential>
@@ -56,4 +59,10 @@
     </sequential>
   </target>
 
+  <target name="javacc-surround" depends="javacc-check" description="generate surround query parser from jj (requires javacc 4.1">
+  	<invoke-javacc target="src/java/org/apache/lucene/queryparser/surround/parser/QueryParser.jj"
+                   outputDir="src/java/org/apache/lucene/queryparser/surround/parser"
+    />
+  </target>
+
 </project>
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/ext/TestExtensions.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/ext/ExtensionStub.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/ext/TestExtendableQueryParser.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/surround/query/ExceptionQueryTst.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/surround/query/BooleanQueryTst.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/surround/query/Test02Boolean.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/surround/query/Test01Exceptions.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/surround/query/SingleFieldTestDb.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/surround/query/Test03Distance.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/complexPhrase/TestComplexPhraseQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/test/org/apache/lucene/queryParser/analyzing/TestAnalyzingQueryParser.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/ext/Extensions.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/ext/ExtensionQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/ext/ParserExtension.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/ext/ExtendableQueryParser.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/ext/package.html (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/BasicQueryFactory.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/ComposedQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/OrQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/SrndTruncQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/SrndPrefixQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/package.html (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/DistanceQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/SimpleTerm.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/SrndTermQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/TooManyBasicQueries.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/NotQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/AndQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/DistanceSubQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/SrndBooleanQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/SrndQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/SpanNearClauseFactory.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/query/FieldsQuery.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/QueryParserConstants.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/Token.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/TokenMgrError.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/QueryParser.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/QueryParser.jj (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/FastCharStream.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/CharStream.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/ParseException.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/QueryParserTokenManager.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser/package.html (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/package.html (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/complexPhrase/ComplexPhraseQueryParser.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/complexPhrase/package.html (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/analyzing/AnalyzingQueryParser.java (deleted)
===================================================================
Index: lucene/contrib/queryparser/src/java/org/apache/lucene/queryParser/analyzing/package.html (deleted)
===================================================================
Index: lucene/contrib/queryparser/build.xml
===================================================================
--- lucene/contrib/queryparser/build.xml	(revision 1145029)
+++ lucene/contrib/queryparser/build.xml	(working copy)
@@ -37,7 +37,7 @@
   <!--
     NOTE: see the README.javacc for details on how to fully regenerate the parser
   -->
-  <target name="javacc" depends="javacc-flexible,javacc-surround"/>
+  <target name="javacc" depends="javacc-flexible"/>
 
   <target name="javacc-flexible" depends="javacc-check">
     <delete>
@@ -102,12 +102,6 @@
                              byline="true"/>
   </target>
 
-  <target name="javacc-surround" depends="javacc-check" description="generate surround query parser from jj (requires javacc 4.1">
-  	<invoke-javacc target="src/java/org/apache/lucene/queryParser/surround/parser/QueryParser.jj"
-                   outputDir="src/java/org/apache/lucene/queryParser/surround/parser"
-    />
-  </target>
-
   <target name="build-queryparser" unless="queryparser.uptodate">
     <echo>QueryParser Contrib building dependency modules/queryparser</echo>
     <subant target="default">
Index: lucene/build.xml
===================================================================
--- lucene/build.xml	(revision 1145029)
+++ lucene/build.xml	(working copy)
@@ -418,9 +418,6 @@
   <!-- ================================================================== -->
   <target name="clean-javacc">
     <delete>
-      <fileset dir="contrib/queryparser/src/java/org/apache/lucene/queryParser/surround/parser" includes="*.java">
-        <containsregexp expression="Generated.*By.*JavaCC"/>
-      </fileset>
       <fileset dir="contrib/queryparser/src/java/org/apache/lucene/queryParser/standard/parser" includes="*.java">
         <containsregexp expression="Generated.*By.*JavaCC"/>
       </fileset>
