Index: src/java/org/apache/lucene/search/IndexSearcher.java
===================================================================
RCS file: /home/cvspublic/jakarta-lucene/src/java/org/apache/lucene/search/IndexSearcher.java,v
retrieving revision 1.23
diff -u -u -U5 -r1.23 IndexSearcher.java
--- src/java/org/apache/lucene/search/IndexSearcher.java	14 Nov 2004 13:26:28 -0000	1.23
+++ src/java/org/apache/lucene/search/IndexSearcher.java	10 Dec 2004 05:05:01 -0000
@@ -73,10 +73,18 @@
   // inherit javadoc
   public Document doc(int i) throws IOException {
     return reader.document(i);
   }

+  /**
+   * Returns the number of documents in this index.
+   * @see IndexReader#numDocs()
+   **/
+  public int numDocs() throws IOException {
+    return reader.numDocs();
+  }
+
   // inherit javadoc
   public int maxDoc() throws IOException {
     return reader.maxDoc();
   }
