Index: lucene/contrib/analyzers/kuromoji/src/java/org/apache/lucene/analysis/kuromoji/util/CSVUtil.java
===================================================================
--- lucene/contrib/analyzers/kuromoji/src/java/org/apache/lucene/analysis/kuromoji/util/CSVUtil.java	(revision 1232002)
+++ lucene/contrib/analyzers/kuromoji/src/java/org/apache/lucene/analysis/kuromoji/util/CSVUtil.java	(working copy)
@@ -54,7 +54,7 @@
         String value = sb.toString();
         value = unQuoteUnEscape(value);
         result.add(value);
-        sb = new StringBuilder();
+        sb.setLength(0);
         continue;
       }
       
@@ -75,14 +75,18 @@
     String result = original;
     
     // Unquote
-    Matcher m = QUOTE_REPLACE_PATTERN.matcher(original);
-    if(m.matches()) {
-      result = m.group(1);
+    if (result.indexOf('\"') >= 0) {
+      Matcher m = QUOTE_REPLACE_PATTERN.matcher(original);
+      if(m.matches()) {
+        result = m.group(1);
+      }
+    
+      // Unescape
+      if (result.indexOf(ESCAPED_QUOTE) >= 0) {
+        result = result.replace(ESCAPED_QUOTE, "\"");
+      }
     }
     
-    // Unescape
-    result = result.replaceAll(ESCAPED_QUOTE, "\"");
-    
     return result;
     
   }
@@ -92,7 +96,11 @@
    * @param original
    */
   public static String quoteEscape(String original) {
-    String result = original.replaceAll("\"", ESCAPED_QUOTE);
+    String result = original;
+    
+    if (result.indexOf('\"') >= 0) {
+      result.replace("\"", ESCAPED_QUOTE);
+    }
     if(result.indexOf(COMMA) >= 0) {
       result = "\"" + result + "\"";
     }
Index: lucene/contrib/analyzers/kuromoji/build.xml
===================================================================
--- lucene/contrib/analyzers/kuromoji/build.xml	(revision 1232002)
+++ lucene/contrib/analyzers/kuromoji/build.xml	(working copy)
@@ -72,7 +72,7 @@
       <delete verbose="true">
         <fileset dir="src/resources/org/apache/lucene/analysis/kuromoji/dict" includes="**/*"/>
       </delete>
-      <java fork="true" failonerror="true" maxmemory="512m" classname="org.apache.lucene.analysis.kuromoji.util.DictionaryBuilder">
+      <java fork="true" failonerror="true" maxmemory="756m" classname="org.apache.lucene.analysis.kuromoji.util.DictionaryBuilder">
         <classpath>
           <path refid="tools.classpath"/>
           <pathelement path="${build.dir}/classes/tools"/>
