Index: lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java	(revision 1720633)
+++ lucene/core/src/java/org/apache/lucene/search/BooleanQuery.java	(working copy)
@@ -189,6 +189,11 @@
     return clauses;
   }
 
+  /** Return the collection of queries for the given {@link Occur}. */
+  Collection<Query> getClauses(Occur occur) {
+    return clauseSets.get(occur);
+  }
+
   /** Returns an iterator on the clauses in this query. It implements the {@link Iterable} interface to
    * make it possible to do:
    * <pre class="prettyprint">for (BooleanClause clause : booleanQuery) {}</pre>
Index: lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java	(revision 1720633)
+++ lucene/core/src/java/org/apache/lucene/search/BooleanWeight.java	(working copy)
@@ -191,25 +191,27 @@
    *  cannot be used. */
   // pkg-private for forcing use of BooleanScorer in tests
   BulkScorer booleanScorer(LeafReaderContext context) throws IOException {
+    if (query.getClauses(Occur.MUST).isEmpty() == false
+        || query.getClauses(Occur.FILTER).isEmpty() == false) {
+      // TODO: there are some cases where BooleanScorer
+      // would handle conjunctions faster than
+      // BooleanScorer2...
+      return null;
+    } else if (query.getClauses(Occur.MUST_NOT).isEmpty() == false) {
+      // TODO: there are some cases where BooleanScorer could do this faster
+      return null;
+    }
+
     List<BulkScorer> optional = new ArrayList<BulkScorer>();
     Iterator<BooleanClause> cIter = query.iterator();
     for (Weight w  : weights) {
       BooleanClause c =  cIter.next();
+      if (c.getOccur() != Occur.SHOULD) {
+        throw new AssertionError();
+      }
       BulkScorer subScorer = w.bulkScorer(context);
-      
-      if (subScorer == null) {
-        if (c.isRequired()) {
-          return null;
-        }
-      } else if (c.isRequired()) {
-        // TODO: there are some cases where BooleanScorer
-        // would handle conjunctions faster than
-        // BooleanScorer2...
-        return null;
-      } else if (c.isProhibited()) {
-        // TODO: there are some cases where BooleanScorer could do this faster
-        return null;
-      } else {
+
+      if (subScorer != null) {
         optional.add(subScorer);
       }
     }
