Index: /home/dnaber/workspace/Lucene/src/test/org/apache/lucene/index/TestMultiReader.java
===================================================================
--- /home/dnaber/workspace/Lucene/src/test/org/apache/lucene/index/TestMultiReader.java	(revision 499176)
+++ /home/dnaber/workspace/Lucene/src/test/org/apache/lucene/index/TestMultiReader.java	(working copy)
@@ -106,7 +106,7 @@
     assertTrue(reader != null);
   }
   
-  /* known to fail, see https://issues.apache.org/jira/browse/LUCENE-781
+  /* see https://issues.apache.org/jira/browse/LUCENE-781 */
   public void testIsCurrent() throws IOException {
     RAMDirectory ramDir1=new RAMDirectory();
     addDoc(ramDir1, "test foo", true);
@@ -114,11 +114,12 @@
     addDoc(ramDir2, "test blah", true);
     IndexReader[] readers = new IndexReader[]{IndexReader.open(ramDir1), IndexReader.open(ramDir2)};
     MultiReader mr = new MultiReader(readers);
-    assertTrue(mr.isCurrent());   // just opened, must be current
-    addDoc(ramDir1, "more text", false);
-    assertFalse(mr.isCurrent());   // has been modified, not current anymore
-    addDoc(ramDir2, "even more text", false);
-    assertFalse(mr.isCurrent());   // has been modified even more, not current anymore
+    try {
+      mr.isCurrent();
+      fail();
+    } catch (UnsupportedOperationException e) {
+      // expected exception
+    }
     try {
       mr.getVersion();
       fail();
@@ -125,6 +126,15 @@
     } catch (UnsupportedOperationException e) {
       // expected exception
     }
+    try {
+      mr.isOptimized();
+      fail();
+    } catch (UnsupportedOperationException e) {
+      // expected exception
+    }
+    // TODO: what to do here?
+    //mr.setNorm(0, "body", 0.5f);
+    assertNotNull(mr.directory());
     mr.close();
   }
 
@@ -135,6 +145,5 @@
     iw.addDocument(doc);
     iw.close();
   }
-  */
 
 }
