--- TestRussianAnalyzer.java	2008-01-21 07:29:39.000000000 +0600
+++ TestRussianAnalyzer.java.new	2008-02-19 09:57:46.000000000 +0600
@@ -168,4 +168,21 @@
         inWords1251.close();
         sample1251.close();
     }
+    
+    public void testDigitsInRussianCharset() 
+    {
+        Reader reader = new StringReader("text 1000");
+        RussianAnalyzer ra = new RussianAnalyzer();
+        TokenStream stream = ra.tokenStream("", reader);
+
+        try {
+            assertEquals("text", stream.next().termText());
+            assertNotNull("RussianAnalyzer's tokenizer skips numbers from input text", stream.next());
+        }
+        catch (IOException e)
+        {
+            fail("unexpected IOException");
+        }
+    }
+
 }
