diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionDISI.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionDISI.java
index ea1e096622b..40fc1a8c077 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionDISI.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionDISI.java
@@ -30,6 +30,7 @@ import org.apache.lucene.util.CollectionUtil;
  * A conjunction of DocIdSetIterators.
  * This iterates over the doc ids that are present in each given DocIdSetIterator.
  *
+ * @since 7.4
  * @lucene.internal
  */
 final class ConjunctionDISI extends DocIdSetIterator {
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionIntervalIterator.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionIntervalIterator.java
index 8efe4e43b55..a5077603ed9 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionIntervalIterator.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionIntervalIterator.java
@@ -22,6 +22,9 @@ import java.util.List;
 
 import org.apache.lucene.search.DocIdSetIterator;
 
+/**
+ * @since 7.4
+ */
 abstract class ConjunctionIntervalIterator extends IntervalIterator {
 
   final DocIdSetIterator approximation;
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionIntervalsSource.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionIntervalsSource.java
index d2805c97890..7b92d8a3e3f 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionIntervalsSource.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/ConjunctionIntervalsSource.java
@@ -27,6 +27,9 @@ import java.util.stream.Collectors;
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.Term;
 
+/**
+ * @since 7.4
+ */
 class ConjunctionIntervalsSource extends IntervalsSource {
 
   final List<IntervalsSource> subSources;
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DifferenceIntervalFunction.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DifferenceIntervalFunction.java
index 18d4d677e53..a43d9fad6d2 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DifferenceIntervalFunction.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DifferenceIntervalFunction.java
@@ -23,6 +23,8 @@ import java.util.Objects;
 /**
  * A function that takes two interval iterators and combines them to produce a third,
  * generally by computing a difference interval between them
+ *
+ * @since 7.4
  */
 abstract class DifferenceIntervalFunction {
 
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DifferenceIntervalsSource.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DifferenceIntervalsSource.java
index 316b6ff14c4..cc7b4e1dd43 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DifferenceIntervalsSource.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DifferenceIntervalsSource.java
@@ -24,6 +24,9 @@ import java.util.Set;
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.Term;
 
+/**
+ * @since 7.4
+ */
 class DifferenceIntervalsSource extends IntervalsSource {
 
   final IntervalsSource minuend;
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisiPriorityQueue.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisiPriorityQueue.java
index 0be96ab9d46..2435875cbe0 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisiPriorityQueue.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisiPriorityQueue.java
@@ -26,6 +26,8 @@ import org.apache.lucene.util.PriorityQueue;
  * A priority queue of DocIdSetIterators that orders by current doc ID.
  * This specialization is needed over {@link PriorityQueue} because the
  * pluggable comparison function makes the rebalancing quite slow.
+ *
+ * @since 7.4
  * @lucene.internal
  */
 final class DisiPriorityQueue implements Iterable<DisiWrapper> {
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisiWrapper.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisiWrapper.java
index 483c6268cf2..f701e6ea95c 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisiWrapper.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisiWrapper.java
@@ -19,6 +19,9 @@ package org.apache.lucene.search.intervals;
 
 import org.apache.lucene.search.DocIdSetIterator;
 
+/**
+ * @since 7.4
+ */
 class DisiWrapper {
 
   public final DocIdSetIterator iterator;
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisjunctionDISIApproximation.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisjunctionDISIApproximation.java
index d4e8e79a180..49a5cf43995 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisjunctionDISIApproximation.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisjunctionDISIApproximation.java
@@ -23,6 +23,8 @@ import org.apache.lucene.search.DocIdSetIterator;
 /**
  * A {@link DocIdSetIterator} which is a disjunction of the approximations of
  * the provided iterators.
+ *
+ * @since 7.4
  * @lucene.internal
  */
 class DisjunctionDISIApproximation extends DocIdSetIterator {
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisjunctionIntervalsSource.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisjunctionIntervalsSource.java
index c7c7166d4b0..8678224ae50 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisjunctionIntervalsSource.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/DisjunctionIntervalsSource.java
@@ -29,6 +29,9 @@ import org.apache.lucene.index.Term;
 import org.apache.lucene.search.DocIdSetIterator;
 import org.apache.lucene.util.PriorityQueue;
 
+/**
+ * @since 7.4
+ */
 class DisjunctionIntervalsSource extends IntervalsSource {
 
   final List<IntervalsSource> subSources;
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalFilter.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalFilter.java
index 7571fc26bf7..0f0433a05f6 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalFilter.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalFilter.java
@@ -22,6 +22,8 @@ import java.util.Objects;
 
 /**
  * Wraps an {@link IntervalIterator} and passes through those intervals that match the {@link #accept()} function
+ *
+ * @since 7.4
  */
 public abstract class IntervalFilter extends IntervalIterator {
 
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalFunction.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalFunction.java
index a6ce918f517..76102704ca6 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalFunction.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalFunction.java
@@ -24,6 +24,8 @@ import org.apache.lucene.util.PriorityQueue;
 
 /**
  * Combine a list of {@link IntervalIterator}s into another
+ *
+ * @since 7.4
  */
 abstract class IntervalFunction {
 
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalIterator.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalIterator.java
index 242872001d8..dd4c6e9c37b 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalIterator.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalIterator.java
@@ -36,6 +36,8 @@ import org.apache.lucene.search.TwoPhaseIterator;
  *
  * Note that it is possible for a document to return {@link #NO_MORE_INTERVALS}
  * on the first call to {@link #nextInterval()}
+ *
+ * @since 7.4
  */
 public abstract class IntervalIterator extends DocIdSetIterator {
 
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalQuery.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalQuery.java
index cd1dca15db6..331fa3e5519 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalQuery.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalQuery.java
@@ -41,6 +41,8 @@ import org.apache.lucene.search.similarities.Similarity;
  *
  * Static constructor functions for various different sources can be found in the
  * {@link Intervals} class
+ *
+ * @since 7.4
  */
 public final class IntervalQuery extends Query {
 
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalScorer.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalScorer.java
index 8c3c5f291bb..81a47a4b3c5 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalScorer.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalScorer.java
@@ -26,6 +26,9 @@ import org.apache.lucene.search.TwoPhaseIterator;
 import org.apache.lucene.search.Weight;
 import org.apache.lucene.search.similarities.Similarity;
 
+/**
+ * @since 7.4
+ */
 class IntervalScorer extends Scorer {
 
   private final IntervalIterator intervals;
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/Intervals.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/Intervals.java
index 32ea6da6a8e..324b33a9af1 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/Intervals.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/Intervals.java
@@ -27,6 +27,8 @@ import org.apache.lucene.util.BytesRef;
  * These sources implement minimum-interval algorithms taken from the paper
  * <a href="http://vigna.di.unimi.it/ftp/papers/EfficientAlgorithmsMinimalIntervalSemantics.pdf">
  * Efficient Optimally Lazy Algorithms for Minimal-Interval Semantics</a>
+ *
+ * @since 7.4
  */
 public final class Intervals {
 
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalsSource.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalsSource.java
index 9791ff87b6f..06ebf6e3d5e 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalsSource.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/IntervalsSource.java
@@ -29,6 +29,8 @@ import org.apache.lucene.index.Term;
  *
  * Static constructor functions for various different sources can be found in the
  * {@link Intervals} class
+ *
+ * @since 7.4
  */
 public abstract class IntervalsSource {
 
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/LowpassIntervalsSource.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/LowpassIntervalsSource.java
index 4d7846c6f7f..a5324132c52 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/LowpassIntervalsSource.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/LowpassIntervalsSource.java
@@ -24,6 +24,9 @@ import java.util.Set;
 import org.apache.lucene.index.LeafReaderContext;
 import org.apache.lucene.index.Term;
 
+/**
+ * @since 7.4
+ */
 class LowpassIntervalsSource extends IntervalsSource {
 
   final IntervalsSource in;
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/TermIntervalsSource.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/TermIntervalsSource.java
index 81578ef02cb..9196b7205ba 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/TermIntervalsSource.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/TermIntervalsSource.java
@@ -33,6 +33,9 @@ import org.apache.lucene.search.DocIdSetIterator;
 import org.apache.lucene.search.TwoPhaseIterator;
 import org.apache.lucene.util.BytesRef;
 
+/**
+ * @since 7.4
+ */
 class TermIntervalsSource extends IntervalsSource {
 
   final BytesRef term;
diff --git a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/package-info.java b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/package-info.java
index 88d93eaaa75..3d3879d41f7 100644
--- a/lucene/sandbox/src/java/org/apache/lucene/search/intervals/package-info.java
+++ b/lucene/sandbox/src/java/org/apache/lucene/search/intervals/package-info.java
@@ -62,5 +62,7 @@
  *
  * An {@link org.apache.lucene.search.intervals.IntervalQuery} takes a field name and an {@link org.apache.lucene.search.intervals.IntervalsSource},
  * and matches all documents that contain intervals defined by the source in that field.
+ *
+ * @since 7.4
  */
 package org.apache.lucene.search.intervals;
\ No newline at end of file
