diff -r 413596ee2583 src/main/perf/PKLookupTask.java
--- a/src/main/perf/PKLookupTask.java	Fri Apr 03 11:21:02 2015 -0400
+++ b/src/main/perf/PKLookupTask.java	Tue Apr 14 00:58:31 2015 -0400
@@ -86,7 +86,7 @@
       final TermsEnum[] termsEnums = new TermsEnum[subReaders.size()];
       final PostingsEnum[] docsEnums = new PostingsEnum[subReaders.size()];
       for(int subIDX=0;subIDX<subReaders.size();subIDX++) {
-        termsEnums[subIDX] = subReaders.get(subIDX).reader().fields().terms("id").iterator(null);
+        termsEnums[subIDX] = subReaders.get(subIDX).reader().fields().terms("id").iterator();
       }
 
       for(int idx=0;idx<ids.length;idx++) {
diff -r 413596ee2583 src/main/perf/TaskParser.java
--- a/src/main/perf/TaskParser.java	Fri Apr 03 11:21:02 2015 -0400
+++ b/src/main/perf/TaskParser.java	Tue Apr 14 00:58:31 2015 -0400
@@ -31,6 +31,7 @@
 import org.apache.lucene.search.BooleanQuery;
 import org.apache.lucene.search.DisjunctionMaxQuery;
 import org.apache.lucene.search.Filter;
+import org.apache.lucene.search.MultiPhraseQuery;
 import org.apache.lucene.search.NumericRangeQuery;
 import org.apache.lucene.search.Query;
 import org.apache.lucene.search.Sort;
@@ -200,6 +201,30 @@
                                   true);
         sort = null;
         group = null;
+      } else if (text.startsWith("multiPhrase//(")) {
+        int colon = text.indexOf(':');
+        if (colon == -1) {
+          throw new RuntimeException("failed to parse query=" + text);
+        }
+        String field = text.substring("//multiPhrase(".length(), colon);
+        MultiPhraseQuery pq = new MultiPhraseQuery();
+        int endParen = text.indexOf(')');
+        if (endParen == -1) {
+          throw new RuntimeException("failed to parse query=" + text);
+        }
+        String queryText = text.substring(colon+1, endParen);
+        String elements[] = queryText.split("\\s+");
+        for (int i = 0; i < elements.length; i++) {
+          String words[] = elements[i].split("\\|");
+          Term terms[] = new Term[words.length];
+          for (int j = 0; j < words.length; j++) {
+            terms[j] = new Term(field, words[j]);
+          }
+          pq.add(terms);
+        }
+        query = pq;
+        sort = null;
+        group = null;
       } else if (text.startsWith("disjunctionMax//")) {
         final int spot3 = text.indexOf(' ');
         if (spot3 == -1) {
diff -r 413596ee2583 tasks/mpq.tasks
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/tasks/mpq.tasks	Tue Apr 14 00:58:31 2015 -0400
@@ -0,0 +1,6 @@
+MultiPhraseHHH: multiPhrase//(body:in|of the)
+MultiPhraseHHM: multiPhrase//(body:in|of your)
+MultiPhraseHHL: multiPhrase//(body:in|of harvard)
+MultiPhraseMMH: multiPhrase//(body:northern|southern states)
+MultiPhraseMMM: multiPhrase//(body:northern|southern usa)
+MultiPhraseMML: multiPhrase//(body:northern|southern iraq)
