diff --git a/lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilterFactory.java b/lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilterFactory.java
index eb6654c2c3..e0c71871b9 100644
--- a/lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilterFactory.java
+++ b/lucene/analysis/common/src/java/org/apache/lucene/analysis/miscellaneous/StemmerOverrideFilterFactory.java
@@ -61,9 +61,16 @@ public class StemmerOverrideFilterFactory extends TokenFilterFactory implements
         StemmerOverrideFilter.Builder builder = new StemmerOverrideFilter.Builder(ignoreCase);
         for (String file : files) {
           List<String> list = getLines(loader, file.trim());
-          for (String line : list) {
-            String[] mapping = line.split("\t", 2);
-            builder.add(mapping[0], mapping[1]);
+          for (int i = 0; i < list.size(); i++)
+          {
+            String line = list.get(i);
+            String[] mapping = line.split("\\s+", 2);
+            if (mapping.length == 2) {
+              builder.add(mapping[0], mapping[1]);
+            } else {
+              throw new IllegalArgumentException(
+                  "Error reading StemmerOverrideFilter " + file + " dictionary at line " + (i + 1) + ".");
+            }
           }
         }
         dictionary = builder.build();
