Index: src/test/org/apache/lucene/index/TestDocumentWriter.java
===================================================================
--- src/test/org/apache/lucene/index/TestDocumentWriter.java	(revision 608138)
+++ src/test/org/apache/lucene/index/TestDocumentWriter.java	(working copy)
@@ -18,8 +18,11 @@
  */
 
 import org.apache.lucene.analysis.*;
+import org.apache.lucene.analysis.standard.StandardAnalyzer;
 import org.apache.lucene.document.Document;
 import org.apache.lucene.document.Field;
+import org.apache.lucene.document.Field.Index;
+import org.apache.lucene.document.Field.Store;
 import org.apache.lucene.document.Field.TermVector;
 import org.apache.lucene.document.Fieldable;
 import org.apache.lucene.search.Similarity;
@@ -232,4 +235,20 @@
     assertEquals(2, termPositions.nextPosition());
 
   }
+
+  /**
+   * Test that adding two fields with the same name, but with different indexing
+   * 
+   * @throws Exception
+   */
+  public void testTermVector() throws Exception {
+	Document doc = new Document();
+    doc.add(new Field("f1", "v1", Store.YES, Index.TOKENIZED, TermVector.WITH_POSITIONS_OFFSETS));
+    doc.add(new Field("f1", "v2", Store.YES, Index.TOKENIZED, TermVector.NO));
+  
+    RAMDirectory ram = new RAMDirectory();
+    IndexWriter writer = new IndexWriter(ram, new StandardAnalyzer(), true);
+    writer.addDocument(doc);
+    writer.close();
+  }
 }
