diff --git a/gradle/defaults-java.gradle b/gradle/defaults-java.gradle
index 4d0fae7d054..d9110124389 100644
--- a/gradle/defaults-java.gradle
+++ b/gradle/defaults-java.gradle
@@ -25,10 +25,24 @@ allprojects {
     tasks.withType(JavaCompile) {
       options.encoding = "UTF-8"
       options.compilerArgs += [
-        "-Xlint",
         "-Xlint:-deprecation",
         "-Xlint:-serial",
-
+        "-Xlint:cast",
+        "-Xlint:classfile",
+        "-Xlint:dep-ann",
+        "-Xlint:divzero",
+        "-Xlint:empty",
+        "-Xlint:fallthrough",
+        "-Xlint:finally",
+        "-Xlint:options",
+        "-Xlint:overrides",
+        "-Xlint:path",
+        "-Xlint:processing",
+        "-Xlint:rawtypes",
+        "-Xlint:static",
+        "-Xlint:try",
+        "-Xlint:unchecked",
+        "-Xlint:varargs",
         "-Xdoclint:all/protected",
         "-Xdoclint:-missing",
         "-Xdoclint:-accessibility",
