diff --git a/lucene/core/src/java/org/apache/lucene/store/FSDirectory.java b/lucene/core/src/java/org/apache/lucene/store/FSDirectory.java
index 50b52ceb07..158a2e2303 100644
--- a/lucene/core/src/java/org/apache/lucene/store/FSDirectory.java
+++ b/lucene/core/src/java/org/apache/lucene/store/FSDirectory.java
@@ -215,7 +215,7 @@ public abstract class FSDirectory extends BaseDirectory {
     try (DirectoryStream<Path> stream = Files.newDirectoryStream(dir)) {
       for (Path path : stream) {
         String name = path.getFileName().toString();
-        if (skipNames != null && skipNames.contains(name) == false) {
+        if (skipNames == null || skipNames.contains(name) == false) {
           entries.add(name);
         }
       }
diff --git a/lucene/core/src/test/org/apache/lucene/store/TestDirectory.java b/lucene/core/src/test/org/apache/lucene/store/TestDirectory.java
index 5e4a5934d3..d3489ad7fd 100644
--- a/lucene/core/src/test/org/apache/lucene/store/TestDirectory.java
+++ b/lucene/core/src/test/org/apache/lucene/store/TestDirectory.java
@@ -22,7 +22,9 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
 
 import org.apache.lucene.util.LuceneTestCase;
 
@@ -137,5 +139,16 @@ public class TestDirectory extends LuceneTestCase {
       fsDir.close();
     }
   }
+
+  public void testListAll() throws Throwable {
+    Path dir = createTempDir("testdir");
+    Path file1 = Files.createFile(dir.resolve("tempfile1"));
+    Path file2 = Files.createFile(dir.resolve("tempfile2"));
+    Set<String> files = new HashSet<>(Arrays.asList(FSDirectory.listAll(dir)));
+
+    assertTrue(files.size() == 2);
+    assertTrue(files.contains(file1.getFileName().toString()));
+    assertTrue(files.contains(file2.getFileName().toString()));
+  }
 }
 
