Index: lucene/test-framework/src/java/org/apache/lucene/index/AssertingLeafReader.java
===================================================================
--- lucene/test-framework/src/java/org/apache/lucene/index/AssertingLeafReader.java	(revision 1675096)
+++ lucene/test-framework/src/java/org/apache/lucene/index/AssertingLeafReader.java	(working copy)
@@ -797,16 +797,15 @@
     return docsWithField;
   }
 
-  // this is the same hack as FCInvisible
+  // we don't change behavior of the reader: just validate the API.
+
   @Override
   public Object getCoreCacheKey() {
-    return cacheKey;
+    return in.getCoreCacheKey();
   }
 
   @Override
   public Object getCombinedCoreAndDeletesKey() {
-    return cacheKey;
+    return in.getCombinedCoreAndDeletesKey();
   }
-  
-  private final Object cacheKey = new Object();
 }
