Index: lucene/codecs/src/java/org/apache/lucene/codecs/bloom/FuzzySet.java
===================================================================
--- lucene/codecs/src/java/org/apache/lucene/codecs/bloom/FuzzySet.java	(revision 1498027)
+++ lucene/codecs/src/java/org/apache/lucene/codecs/bloom/FuzzySet.java	(working copy)
@@ -99,10 +99,11 @@
   public static int getNearestSetSize(int maxNumberOfBits)
   {
     int result=usableBitSetSizes[0];
-    for (int i = 0; i < usableBitSetSizes.length; i++) {
+    for (int i = usableBitSetSizes.length - 1; i >= 0; i--) {
       if(usableBitSetSizes[i]<=maxNumberOfBits)
       {
         result=usableBitSetSizes[i];
+        break;
       }
     }
     return result;
