diff --git lucene/core/src/java/org/apache/lucene/analysis/tokenattributes/KeywordAttributeImpl.java lucene/core/src/java/org/apache/lucene/analysis/tokenattributes/KeywordAttributeImpl.java
index 481ee74..0c1999c 100644
--- lucene/core/src/java/org/apache/lucene/analysis/tokenattributes/KeywordAttributeImpl.java
+++ lucene/core/src/java/org/apache/lucene/analysis/tokenattributes/KeywordAttributeImpl.java
@@ -45,10 +45,15 @@ public final class KeywordAttributeImpl extends AttributeImpl implements
 
   @Override
   public boolean equals(Object obj) {
-    if (this == obj)
+    if (this == obj) {
       return true;
-    if (getClass() != obj.getClass())
+    }
+    if (obj == null) {
       return false;
+    }
+    if (getClass() != obj.getClass()) {
+      return false;
+    }
     final KeywordAttributeImpl other = (KeywordAttributeImpl) obj;
     return keyword == other.keyword;
   }
diff --git lucene/core/src/java/org/apache/lucene/util/automaton/BasicOperations.java lucene/core/src/java/org/apache/lucene/util/automaton/BasicOperations.java
index 781a4e2..bdba0da 100644
--- lucene/core/src/java/org/apache/lucene/util/automaton/BasicOperations.java
+++ lucene/core/src/java/org/apache/lucene/util/automaton/BasicOperations.java
@@ -486,7 +486,7 @@ final public class BasicOperations {
 
     @Override
     public boolean equals(Object other) {
-      return ((PointTransitions) other).point == point;
+      return other != null && ((PointTransitions) other).point == point;
     }
 
     @Override
diff --git lucene/core/src/java/org/apache/lucene/util/automaton/DaciukMihovAutomatonBuilder.java lucene/core/src/java/org/apache/lucene/util/automaton/DaciukMihovAutomatonBuilder.java
index c6be628..86389f0 100644
--- lucene/core/src/java/org/apache/lucene/util/automaton/DaciukMihovAutomatonBuilder.java
+++ lucene/core/src/java/org/apache/lucene/util/automaton/DaciukMihovAutomatonBuilder.java
@@ -82,6 +82,9 @@ final class DaciukMihovAutomatonBuilder {
      */
     @Override
     public boolean equals(Object obj) {
+      if (obj == null) {
+        return false;
+      }
       final State other = (State) obj;
       return is_final == other.is_final
           && Arrays.equals(this.labels, other.labels)
