Index: src/java/org/apache/lucene/index/FieldInfos.java
===================================================================
--- src/java/org/apache/lucene/index/FieldInfos.java	(revision 191671)
+++ src/java/org/apache/lucene/index/FieldInfos.java	(working copy)
@@ -166,14 +166,9 @@
   }
 
   public int fieldNumber(String fieldName) {
-    try {
       FieldInfo fi = fieldInfo(fieldName);
       if (fi != null)
         return fi.number;
-    }
-    catch (IndexOutOfBoundsException ioobe) {
-      return -1;
-    }
     return -1;
   }
 
@@ -204,12 +199,10 @@
    * doesn't exist.
    */  
   public FieldInfo fieldInfo(int fieldNumber) {
-    try {
+      if (fieldNumber < 0 || fieldNumber >= byNumber.size()) {
+	  return null;
+      }
       return (FieldInfo) byNumber.get(fieldNumber);
-    }
-    catch (IndexOutOfBoundsException ioobe) {
-      return null;
-    }
   }
 
   public int size() {
Index: src/gcj/Makefile
===================================================================
--- src/gcj/Makefile	(revision 191671)
+++ src/gcj/Makefile	(working copy)
@@ -8,7 +8,7 @@
 CORE=$(BUILD)/classes/java
 SRC=.
 
-CORE_OBJ:=$(subst .jar,.a,$(wildcard $(BUILD)/lucene-[0-9]*.jar))
+CORE_OBJ:=$(subst .jar,.a,$(wildcard $(BUILD)/lucene-core-*.jar))
 CORE_JAVA:=$(shell find $(ROOT)/src/java -name '*.java')
 
 CORE_HEADERS=\
@@ -55,7 +55,7 @@
 # yet accept from .class files.
 # NOTE: Change when http://gcc.gnu.org/bugzilla/show_bug.cgi?id=15501 is fixed.
 $(CORE_OBJ) : $(CORE_JAVA)
-	$(GCJ) $(GCJFLAGS) -c -I $(CORE) -o $@ `find $(ROOT)/src/java -name '*.java' -not -name '*Sort*' -not -name 'Span*'` `find $(CORE) -name '*.class' -name '*Sort*' -or -name 'Span*'`
+	$(GCJ) $(GCJFLAGS) -c -I $(CORE) -o $@ `find $(ROOT)/src/java -name '*.java' -not -name '*Sort*' -not -name 'Span*' -not -name 'FieldCacheImpl.java'` `find $(CORE) -name '*.class' -name '*Sort*' -or -name 'Span*' -or -name 'FieldCacheImpl*'`
 
 # generate object code from jar files using gcj
 %.a : %.jar
