Index: /lucene/src/java/org/apache/lucene/analysis/standard/StandardAnalyzer.java
===================================================================
--- /lucene/src/java/org/apache/lucene/analysis/standard/StandardAnalyzer.java	(revision 606641)
+++ /lucene/src/java/org/apache/lucene/analysis/standard/StandardAnalyzer.java	(working copy)
@@ -76,10 +76,11 @@
     return result;
   }
 
-  private class SavedStreams {
+  private static final class SavedStreams {
     StandardTokenizer tokenStream;
     TokenStream filteredTokenStream;
-  };
+  }
+
   public TokenStream reusableTokenStream(String fieldName, Reader reader) throws IOException {
     SavedStreams streams = (SavedStreams) getPreviousTokenStream();
     if (streams == null) {
