Index: TestStandardAnalyzer.java
===================================================================
--- TestStandardAnalyzer.java	(revision 541610)
+++ TestStandardAnalyzer.java	(working copy)
@@ -2,6 +2,7 @@
 
 import junit.framework.TestCase;
 import org.apache.lucene.analysis.standard.StandardAnalyzer;
+import org.apache.lucene.analysis.standard.StandardTokenizer;
 
 import java.io.StringReader;
 
@@ -34,7 +35,21 @@
     ts.close();
   }
 
+  public void testPunctuation() throws Exception {
 
+    TokenStream ts = new StandardTokenizer(new StringReader("hello, world!"));
+    assertEquals("hello", ts.next().termText());
+    assertEquals("<PUNCTUATION>", ts.next().type());
+    assertEquals("world", ts.next().termText());
+    assertEquals("<PUNCTUATION>", ts.next().type());
+    assertNull(ts.next());
+
+    ts = new StandardAnalyzer(new String[]{}).tokenStream(null, new StringReader("hello, world!"));
+    assertEquals("hello", ts.next().termText());
+    assertEquals("world", ts.next().termText());
+    assertNull(ts.next());
+  }
+
   public void testStandard() throws Exception {
     Analyzer a = new StandardAnalyzer();
 
@@ -108,4 +123,8 @@
     assertAnalyzesTo(a, "안녕하세요 한글입니다", new String[]{"안녕하세요", "한글입니다"});
 
   }
+
+
+
+
 }
