Index: src/test/org/apache/lucene/spatial3d/geom/GeoExactCircleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/test/org/apache/lucene/spatial3d/geom/GeoExactCircleTest.java	(date 1511921766000)
+++ src/test/org/apache/lucene/spatial3d/geom/GeoExactCircleTest.java	(revision )
@@ -122,6 +122,15 @@
     assertTrue(success);
   }
 
+  @Test
+  public void testExactCircleDoesNotFit() {
+    GeoCircle circle = GeoCircleFactory.makeExactGeoCircle(PlanetModel.WGS84, 1.5633796542562415, -1.0387149580695152,3.1409865861032844, 1e-12);
+    GeoPoint point = new GeoPoint(PlanetModel.WGS84, -1.5633589814325721, 2.1360815933475457);
+    double distance = PlanetModel.WGS84.surfaceDistance(circle.getCenter(), point);
+    assertEquals(distance < circle.getRadius(), circle.isWithin(point));
+  }
+
+
   /**
    * in LUCENE-8054 we have problems with exact circles that have
    * edges that are close together. This test creates those circles with the same
