Index: src/java/org/apache/lucene/store/instantiated/InstantiatedIndexWriter.java
===================================================================
--- src/java/org/apache/lucene/store/instantiated/InstantiatedIndexWriter.java	(revision 823249)
+++ src/java/org/apache/lucene/store/instantiated/InstantiatedIndexWriter.java	(working copy)
@@ -327,6 +327,13 @@
             if (eFieldSetting_TermDocInfoFactoriesByTermText.getKey().fieldName.equals(e.getKey().field())) {
               TermDocumentInformationFactory factory = eFieldSetting_TermDocInfoFactoriesByTermText.getValue().get(e.getKey().text());
               e.getValue().setTermOffsets(factory.termOffsets.toArray(new TermVectorOffsetInfo[factory.termOffsets.size()]));
+              int[] tps = new int[factory.termPositions.size()];
+              int cpt = 0;
+              for(Integer i : factory.termPositions) {
+            	  tps[cpt++] = i.intValue();
+              }
+              e.getValue().setTermPositions(tps);
+              e.getValue().setFreq(factory.termOffsets.size());
             }
           }
         }
