Index: lucene/src/test/org/apache/lucene/util/TestCharacterUtils.java
===================================================================
--- lucene/src/test/org/apache/lucene/util/TestCharacterUtils.java	(revision 996902)
+++ lucene/src/test/org/apache/lucene/util/TestCharacterUtils.java	(working copy)
@@ -43,7 +43,7 @@
     try {
       java4.codePointAt(highSurrogateAt3, 4);
       fail("array index out of bounds");
-    } catch (ArrayIndexOutOfBoundsException e) {
+    } catch (IndexOutOfBoundsException e) {
     }
 
     CharacterUtils java5 = CharacterUtils.getInstance(TEST_VERSION_CURRENT);
@@ -54,7 +54,7 @@
     try {
       java5.codePointAt(highSurrogateAt3, 4);
       fail("array index out of bounds");
-    } catch (ArrayIndexOutOfBoundsException e) {
+    } catch (IndexOutOfBoundsException e) {
     }
   }
 
@@ -69,7 +69,7 @@
     try {
       java4.codePointAt(highSurrogateAt3, 4);
       fail("string index out of bounds");
-    } catch (StringIndexOutOfBoundsException e) {
+    } catch (IndexOutOfBoundsException e) {
     }
 
     CharacterUtils java5 = CharacterUtils.getInstance(TEST_VERSION_CURRENT);
@@ -80,7 +80,7 @@
     try {
       java5.codePointAt(highSurrogateAt3, 4);
       fail("string index out of bounds");
-    } catch (StringIndexOutOfBoundsException e) {
+    } catch (IndexOutOfBoundsException e) {
     }
 
   }
