Index: lucene/backward-codecs/src/test/org/apache/lucene/index/TestBackwardsCompatibility.java
===================================================================
--- lucene/backward-codecs/src/test/org/apache/lucene/index/TestBackwardsCompatibility.java	(revision 1656127)
+++ lucene/backward-codecs/src/test/org/apache/lucene/index/TestBackwardsCompatibility.java	(working copy)
@@ -305,7 +305,8 @@
       "3.6.1-cfs",
       "3.6.1-nocfs",
       "3.6.2-cfs",
-      "3.6.2-nocfs"
+      "3.6.2-nocfs",
+      "4x-with-3x-segments"
   };
   
   final static String[] oldSingleSegmentNames = {"4.0.0-optimized-cfs",
Index: lucene/backward-codecs/src/test/org/apache/lucene/index/unsupported.4x-with-3x-segments.zip
===================================================================
Cannot display: file marked as a binary type.
svn:mime-type = application/octet-stream
Index: lucene/backward-codecs/src/test/org/apache/lucene/index/unsupported.4x-with-3x-segments.zip
===================================================================
--- lucene/backward-codecs/src/test/org/apache/lucene/index/unsupported.4x-with-3x-segments.zip	(revision 1656127)
+++ lucene/backward-codecs/src/test/org/apache/lucene/index/unsupported.4x-with-3x-segments.zip	(working copy)

Property changes on: lucene/backward-codecs/src/test/org/apache/lucene/index/unsupported.4x-with-3x-segments.zip
___________________________________________________________________
Added: svn:mime-type
## -0,0 +1 ##
+application/octet-stream
\ No newline at end of property
Index: lucene/core/src/java/org/apache/lucene/index/SegmentInfos.java
===================================================================
--- lucene/core/src/java/org/apache/lucene/index/SegmentInfos.java	(revision 1656127)
+++ lucene/core/src/java/org/apache/lucene/index/SegmentInfos.java	(working copy)
@@ -37,6 +37,7 @@
 import org.apache.lucene.codecs.FieldInfosFormat;
 import org.apache.lucene.codecs.LiveDocsFormat;
 import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.DataInput;
 import org.apache.lucene.store.DataOutput;
 import org.apache.lucene.store.Directory;
 import org.apache.lucene.store.IOContext;
@@ -134,6 +135,10 @@
   /** The file format version for the segments_N codec header, since 5.0+ */
   public static final int VERSION_50 = 4;
 
+  private static final String[] unsupportedCodecs = {
+    "Lucene3x"  
+  };
+
   /** Used to name new segments. */
   // TODO: should this be a long ...?
   public int counter;
@@ -322,7 +327,7 @@
         } else {
           segmentID = null;
         }
-        Codec codec = Codec.forName(input.readString());
+        Codec codec = readCodec(input);
         SegmentInfo info = codec.segmentInfoFormat().read(directory, segName, segmentID, IOContext.READ);
         info.setCodec(codec);
         long delGen = input.readLong();
@@ -394,6 +399,27 @@
     }
   }
 
+  private static Codec readCodec(DataInput input) throws IOException {
+    final String name = input.readString();
+    try {
+      return Codec.forName(name);
+    } catch (IllegalArgumentException e) {
+      // give better error messages if we can, first check if this is a legacy codec
+      for (String codec : unsupportedCodecs) {
+        if (codec.equals(name)) {
+          IOException newExc = new IndexFormatTooOldException(input, "Codec '" + name + "' is too old");
+          newExc.initCause(e);
+          throw newExc;
+        }
+      }
+      // or maybe it's an old default codec that moved
+      if (name.startsWith("Lucene")) {
+        throw new IllegalArgumentException("Could not load codec '" + name + "'.  Did you forget to add lucene-backward-codecs.jar?", e);
+      }
+      throw e;
+    }
+  }
+
   /** Find the latest commit ({@code segments_N file}) and
    *  load all {@link SegmentCommitInfo}s. */
   public static final SegmentInfos readLatestCommit(Directory directory) throws IOException {
