Index: CharTermAttributeImpl.java
===================================================================
--- CharTermAttributeImpl.java	(revision 978956)
+++ CharTermAttributeImpl.java	(working copy)
@@ -23,7 +23,6 @@
 import org.apache.lucene.util.ArrayUtil;
 import org.apache.lucene.util.AttributeImpl;
 import org.apache.lucene.util.RamUsageEstimator;
-import org.apache.lucene.util.UnicodeUtil;
 
 /**
  * The term text of a Token.
@@ -244,9 +243,13 @@
   @Override
   public Object clone() {
     CharTermAttributeImpl t = (CharTermAttributeImpl)super.clone();
+    
     // Do a deep clone
-    t.termBuffer = termBuffer.clone();
+    t.termBuffer = new char[this.termLength];
+    System.arraycopy(this.termBuffer, 0, t.termBuffer, 0, this.termLength);
+    
     return t;
+    
   }
   
   @Override
