Index: core/src/java/org/apache/solr/schema/BBoxField.java
===================================================================
--- core/src/java/org/apache/solr/schema/BBoxField.java	(revision 1662301)
+++ core/src/java/org/apache/solr/schema/BBoxField.java	(working copy)
@@ -44,6 +44,7 @@
 
   private String numberTypeName;//required
   private String booleanTypeName = "boolean";
+  private boolean storeSubFields = false;
 
   private IndexSchema schema;
 
@@ -66,6 +67,11 @@
     if (v != null) {
       booleanTypeName = v;
     }
+    
+    v = args.remove("storeSubFields");
+    if (v != null) {
+      storeSubFields = Boolean.valueOf(v);
+    }
   }
 
   @Override
@@ -108,7 +114,15 @@
   // note: Registering the field is probably optional; it makes it show up in the schema browser and may have other
   //  benefits.
   private void register(IndexSchema schema, String name, FieldType fieldType) {
-    SchemaField sf = new SchemaField(name, fieldType);
+    int props = fieldType.properties;
+    props |= INDEXED; // subfields must be indexed
+    if(storeSubFields) {
+      props |= STORED;
+    }
+    else {
+      props &= ~STORED;
+    }
+    SchemaField sf = new SchemaField(name, fieldType, props, null);
     schema.getFields().put(sf.getName(), sf);
   }
 
Index: core/src/test-files/solr/collection1/conf/schema-spatial.xml
===================================================================
--- core/src/test-files/solr/collection1/conf/schema-spatial.xml	(revision 1662301)
+++ core/src/test-files/solr/collection1/conf/schema-spatial.xml	(working copy)
@@ -48,7 +48,7 @@
                numberType="tdouble" distanceUnits="degrees"/>
 
     <fieldType name="bbox" class="solr.BBoxField"
-               numberType="tdoubleDV" distanceUnits="degrees"/>
+               numberType="tdoubleDV" distanceUnits="degrees" storeSubFields="false"/>
   </types>
 
 
Index: core/src/test/org/apache/solr/search/TestSolr4Spatial.java
===================================================================
--- core/src/test/org/apache/solr/search/TestSolr4Spatial.java	(revision 1662301)
+++ core/src/test/org/apache/solr/search/TestSolr4Spatial.java	(working copy)
@@ -26,8 +26,13 @@
 import com.spatial4j.core.distance.DistanceUtils;
 import com.spatial4j.core.shape.Point;
 import com.spatial4j.core.shape.Rectangle;
+
+import org.apache.lucene.spatial.bbox.BBoxStrategy;
 import org.apache.solr.SolrTestCaseJ4;
 import org.apache.solr.common.SolrException;
+import org.apache.solr.schema.BBoxField;
+import org.apache.solr.schema.IndexSchema;
+import org.apache.solr.schema.SchemaField;
 import org.apache.solr.util.SpatialUtils;
 import org.junit.Before;
 import org.junit.BeforeClass;
@@ -366,4 +371,16 @@
         SolrException.ErrorCode.BAD_REQUEST);
   }
 
+
+  @Test
+  public void testSpatialConfig() throws Exception {
+    IndexSchema schema = h.getCoreInc().getLatestSchema();
+
+    // BBox Config
+    // Make sure the subfields are not stored
+    SchemaField sub = schema.getField("bbox"+BBoxStrategy.SUFFIX_MINX);
+    assertFalse(sub.stored());
+    assertTrue(sub.indexed());
+  }
+  
 }
