Index: Hits.java
===================================================================
--- Hits.java	(revision 617952)
+++ Hits.java	(working copy)
@@ -60,7 +60,7 @@
   boolean debugCheckedForDeletions = false; // for test purposes.
 
   Hits(Searcher s, Query q, Filter f) throws IOException {
-    weight = q.weight(s);
+    weight = s.createWeight(q); 
     searcher = s;
     filter = f;
     nDeletions = countDeletions(s);
@@ -69,7 +69,7 @@
   }
 
   Hits(Searcher s, Query q, Filter f, Sort o) throws IOException {
-    weight = q.weight(s);
+    weight = s.createWeight(q); 
     searcher = s;
     filter = f;
     sort = o;
