Index: src/java/org/apache/lucene/index/MultiLevelSkipListReader.java
===================================================================
--- src/java/org/apache/lucene/index/MultiLevelSkipListReader.java	(revision 825193)
+++ src/java/org/apache/lucene/index/MultiLevelSkipListReader.java	(working copy)
@@ -159,7 +159,7 @@
   }
   
   void close() throws IOException {
-    for (int i = 1; i < skipStream.length; i++) {
+    for (int i = 0; i < skipStream.length; i++) {
       if (skipStream[i] != null) {
         skipStream[i].close();
       }
