diff --git lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn01ClassLoad.java lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn01ClassLoad.java
new file mode 100644
index 0000000..51ac285
--- /dev/null
+++ lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn01ClassLoad.java
@@ -0,0 +1,17 @@
+package org.apache.lucene.util.junitcompat.failures;
+
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.Test;
+
+/**
+ * Fail on class loading.
+ */
+public class TestFailOn01ClassLoad extends LuceneTestCase {
+  static {
+    System.out.println("I'll throw ExceptionInInitializer/ArithmeticException." 
+        + 2 / 0);
+  }
+
+  @Test
+  public void testMethod() {}
+}
diff --git lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn02BeforeClass.java lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn02BeforeClass.java
new file mode 100644
index 0000000..43ea47f
--- /dev/null
+++ lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn02BeforeClass.java
@@ -0,0 +1,18 @@
+package org.apache.lucene.util.junitcompat.failures;
+
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/**
+ * Fail on new instance creation.
+ */
+public class TestFailOn02BeforeClass extends LuceneTestCase {
+  @BeforeClass
+  public static void beforeClass() {
+    throw new RuntimeException("Exception on @BeforeClass.");
+  }
+
+  @Test
+  public void testMethod() {}
+}
diff --git lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn03NewInstance.java lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn03NewInstance.java
new file mode 100644
index 0000000..331157a
--- /dev/null
+++ lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn03NewInstance.java
@@ -0,0 +1,17 @@
+package org.apache.lucene.util.junitcompat.failures;
+
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.Test;
+
+/**
+ * Fail on new instance creation.
+ */
+public class TestFailOn03NewInstance extends LuceneTestCase {
+  
+  public TestFailOn03NewInstance() {
+    throw new RuntimeException("Exception on new instance.");
+  }
+
+  @Test
+  public void testMethod() {}
+}
diff --git lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn04Before.java lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn04Before.java
new file mode 100644
index 0000000..bc1e028
--- /dev/null
+++ lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn04Before.java
@@ -0,0 +1,19 @@
+package org.apache.lucene.util.junitcompat.failures;
+
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * Fail on new instance creation.
+ */
+public class TestFailOn04Before extends LuceneTestCase {
+  
+  @Before
+  public void before() {
+    throw new RuntimeException("Exception on new instance.");
+  }
+
+  @Test
+  public void testMethod() {}
+}
diff --git lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn05RuleAfter.java lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn05RuleAfter.java
new file mode 100644
index 0000000..2552fe0
--- /dev/null
+++ lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn05RuleAfter.java
@@ -0,0 +1,28 @@
+package org.apache.lucene.util.junitcompat.failures;
+
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+/**
+ * Fail on new instance creation.
+ */
+public class TestFailOn05RuleAfter extends LuceneTestCase {
+  @Rule
+  public TestRule rule = new TestRule() {
+    public Statement apply(final Statement s, Description d) {
+      return new Statement() {
+        public void evaluate() throws Throwable {
+          s.evaluate();
+          throw new RuntimeException("Failure on @Rule (after test).");
+        }
+      };
+    }
+  };
+
+  @Test
+  public void testMethod() {}
+}
diff --git lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn05RuleBefore.java lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn05RuleBefore.java
new file mode 100644
index 0000000..429d1ab
--- /dev/null
+++ lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn05RuleBefore.java
@@ -0,0 +1,27 @@
+package org.apache.lucene.util.junitcompat.failures;
+
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+/**
+ * Fail on new instance creation.
+ */
+public class TestFailOn05RuleBefore extends LuceneTestCase {
+  @Rule
+  public TestRule rule = new TestRule() {
+    public Statement apply(final Statement s, Description d) {
+      return new Statement() {
+        public void evaluate() throws Throwable {
+          throw new RuntimeException("Failure on @Rule (before test).");
+        }
+      };
+    }
+  };
+
+  @Test
+  public void testMethod() {}
+}
diff --git lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn06Test.java lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn06Test.java
new file mode 100644
index 0000000..97ab8f7
--- /dev/null
+++ lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn06Test.java
@@ -0,0 +1,14 @@
+package org.apache.lucene.util.junitcompat.failures;
+
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.Test;
+
+/**
+ * Fail on new instance creation.
+ */
+public class TestFailOn06Test extends LuceneTestCase {
+  @Test
+  public void testMethod() {
+    throw new RuntimeException("Exception on @Test method.");
+  }
+}
diff --git lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn07After.java lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn07After.java
new file mode 100644
index 0000000..848ef35
--- /dev/null
+++ lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn07After.java
@@ -0,0 +1,18 @@
+package org.apache.lucene.util.junitcompat.failures;
+
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.After;
+import org.junit.Test;
+
+/**
+ * Fail on new instance creation.
+ */
+public class TestFailOn07After extends LuceneTestCase {
+  @After
+  public void after() {
+    throw new RuntimeException("Failure on @After.");
+  }
+
+  @Test
+  public void testMethod() {}
+}
diff --git lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn08AfterClass.java lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn08AfterClass.java
new file mode 100644
index 0000000..8de6d1d
--- /dev/null
+++ lucene/core/src/test/org/apache/lucene/util/junitcompat/failures/TestFailOn08AfterClass.java
@@ -0,0 +1,18 @@
+package org.apache.lucene.util.junitcompat.failures;
+
+import org.apache.lucene.util.LuceneTestCase;
+import org.junit.AfterClass;
+import org.junit.Test;
+
+/**
+ * Fail on new instance creation.
+ */
+public class TestFailOn08AfterClass extends LuceneTestCase {
+  @AfterClass
+  public static void afterClass() {
+    throw new RuntimeException("Failure on @AfterClass.");
+  }
+
+  @Test
+  public void testMethod() {}
+}
