Index: lucene/contrib/CHANGES.txt
===================================================================
--- lucene/contrib/CHANGES.txt	(revision 1161104)
+++ lucene/contrib/CHANGES.txt	(working copy)
@@ -41,6 +41,9 @@
    it has been deprecated and will be removed soon. 
    (Phillipe Ramalho via Adriano Crestani)
 
+ * LUCENE-3400: Deprecated DutchAnalyzer.setStemDictionary since it prevents
+   TokenStream reuse (Chris Male)
+
 Optimizations
 
  * LUCENE-3306: Disabled indexing of positions for spellchecker n-gram
Index: lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/nl/DutchAnalyzer.java
===================================================================
--- lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/nl/DutchAnalyzer.java	(revision 1161104)
+++ lucene/contrib/analyzers/common/src/java/org/apache/lucene/analysis/nl/DutchAnalyzer.java	(working copy)
@@ -220,7 +220,10 @@
    * Reads a stemdictionary file , that overrules the stemming algorithm
    * This is a textfile that contains per line
    * <tt>word<b>\t</b>stem</tt>, i.e: two tab seperated words
+   * @deprecated This prevents reuse of TokenStreams.  If you wish to use a custom
+   * stem dictionary, create your own Analyzer with {@link StemmerOverrideFilter}
    */
+  @Deprecated
   public void setStemDictionary(File stemdictFile) {
     try {
       stemdict = WordlistLoader.getStemDict(stemdictFile);
