Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/NormalDistance.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/NormalDistance.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/NormalDistance.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Normal distance computation style.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoRectangle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoRectangle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoRectangle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Bounding box limited on four sides (top lat, bottom lat, left lon, right lon).
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/XYdZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/XYdZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/XYdZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * 3D rectangle, bounded on six sides by X,Y,Z limits, degenerate in Z
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/Geo3DUtil.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/Geo3DUtil.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/Geo3DUtil.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d;
 
 class Geo3DUtil {
 
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegenerateHorizontalLine.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegenerateHorizontalLine.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoDegenerateHorizontalLine.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Degenerate bounding box limited on two sides (left lon, right lon).
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegenerateLongitudeSlice.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegenerateLongitudeSlice.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoDegenerateLongitudeSlice.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Degenerate longitude slice.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegenerateVerticalLine.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegenerateVerticalLine.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoDegenerateVerticalLine.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Degenerate bounding box limited on two sides (top lat, bottom lat).
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/Vector.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/Vector.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/Vector.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * A 3d vector in space, not necessarily
Index: lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoModelTest.java
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoModelTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoModelTest.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import org.junit.Test;
 
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/PlanetModel.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/PlanetModel.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/PlanetModel.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Holds mathematical constants associated with the model of a planet.
@@ -192,7 +192,7 @@
    * @param p1 is the first point.
    * @param p2 is the second point.
    * @return the adjusted angle, when multiplied by the mean earth radius, yields a surface distance.  This will differ
-   * from GeoPoint.arcDistance() only when the planet model is not a sphere. @see {@link org.apache.lucene.geo3d.GeoPoint#arcDistance(GeoPoint)}
+   * from GeoPoint.arcDistance() only when the planet model is not a sphere. @see {@link GeoPoint#arcDistance(GeoPoint)}
    */
   public double surfaceDistance(final GeoPoint p1, final GeoPoint p2) {
     final double latA = p1.getLatitude();
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/Geo3DPoint.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/Geo3DPoint.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/Geo3DPoint.java	(revision )
@@ -14,14 +14,16 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d;
 
 import org.apache.lucene.document.Field;
 import org.apache.lucene.document.FieldType;
+import org.apache.lucene.spatial3d.geom.GeoPoint;
+import org.apache.lucene.spatial3d.geom.GeoShape;
+import org.apache.lucene.spatial3d.geom.PlanetModel;
 import org.apache.lucene.search.Query;
 import org.apache.lucene.util.BytesRef;
 import org.apache.lucene.util.NumericUtils;
-import org.apache.lucene.util.RamUsageEstimator;
 
 /**
  * Add this to a document to index lat/lon or x/y/z point, indexed as a 3D point.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoStandardCircle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoStandardCircle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoStandardCircle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Circular area with a center and radius.
Index: lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dShape.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dShape.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial-extras/src/java/org/apache/lucene/spatial/spatial4j/Geo3dShape.java	(revision )
@@ -23,12 +23,12 @@
 import org.locationtech.spatial4j.shape.Shape;
 import org.locationtech.spatial4j.shape.SpatialRelation;
 import org.locationtech.spatial4j.shape.impl.RectangleImpl;
-import org.apache.lucene.geo3d.LatLonBounds;
-import org.apache.lucene.geo3d.GeoArea;
-import org.apache.lucene.geo3d.GeoAreaFactory;
-import org.apache.lucene.geo3d.GeoPoint;
-import org.apache.lucene.geo3d.GeoShape;
-import org.apache.lucene.geo3d.PlanetModel;
+import org.apache.lucene.spatial3d.geom.LatLonBounds;
+import org.apache.lucene.spatial3d.geom.GeoArea;
+import org.apache.lucene.spatial3d.geom.GeoAreaFactory;
+import org.apache.lucene.spatial3d.geom.GeoPoint;
+import org.apache.lucene.spatial3d.geom.GeoShape;
+import org.apache.lucene.spatial3d.geom.PlanetModel;
 
 /**
  * A Spatial4j Shape wrapping a {@link GeoShape} ("Geo3D") -- a 3D planar geometry based Spatial4j Shape implementation.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/Membership.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/Membership.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/Membership.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Implemented by Geo3D shapes that can calculate if a point is within it or not.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/XdYdZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/XdYdZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/XdYdZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * 3D rectangle, bounded on six sides by X,Y,Z limits, degenerate in Y and Z.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBBox.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBBox.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoBBox.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * All bounding box shapes have this interface in common.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoSouthLatitudeZone.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoSouthLatitudeZone.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoSouthLatitudeZone.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * This GeoBBox represents an area rectangle limited only in north latitude.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/LatLonBounds.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/LatLonBounds.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/LatLonBounds.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * An object for accumulating latitude/longitude bounds information.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/LinearSquaredDistance.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/LinearSquaredDistance.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/LinearSquaredDistance.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Linear squared distance computation style.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/XdYZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/XdYZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/XdYZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * 3D rectangle, bounded on six sides by X,Y,Z limits, degenerate in Y
Index: lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoBBoxTest.java
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoBBoxTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoBBoxTest.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import java.util.ArrayList;
 import java.util.List;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/SidedPlane.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/SidedPlane.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/SidedPlane.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Combination of a plane, and a sign value indicating what evaluation values are on the correct
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/DistanceStyle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/DistanceStyle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/DistanceStyle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Distance computation styles, supporting various ways of computing
Index: lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoCircleTest.java
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoCircleTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoCircleTest.java	(revision )
@@ -14,15 +14,10 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
-import org.junit.Test;
-
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertTrue;
-
 import org.apache.lucene.util.LuceneTestCase;
+import org.junit.Test;
 
 public class GeoCircleTest extends LuceneTestCase {
 
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoLongitudeSlice.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoLongitudeSlice.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoLongitudeSlice.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Bounding box limited on left and right.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/package-info.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/package-info.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/package-info.java	(revision )
@@ -16,6 +16,6 @@
  */
 
 /**
- * Shapes implemented using 3D planar geometry.
+ * Lucene field &amp; query support for the spatial geometry implemented in {@link org.apache.lucene.spatial3d.geom}.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/XYZBounds.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/XYZBounds.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/XYZBounds.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * An object for accumulating XYZ bounds information.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWorld.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWorld.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoWorld.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Bounding box including the entire world.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/BaseXYZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/BaseXYZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/BaseXYZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Base class of a family of 3D rectangles, bounded on six sides by X,Y,Z limits
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/StandardXYZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/StandardXYZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/StandardXYZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * 3D rectangle, bounded on six sides by X,Y,Z limits
Index: lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/geo3d/GeoPointTest.java
===================================================================
--- lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/geo3d/GeoPointTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoPointTest.java	(revision )
@@ -14,33 +14,30 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.spatial.spatial4j.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
-import org.apache.lucene.geo3d.GeoPoint;
-import org.apache.lucene.geo3d.PlanetModel;
 import org.apache.lucene.util.LuceneTestCase;
 import org.junit.Test;
 
-import org.locationtech.spatial4j.distance.DistanceUtils;
-
 import static com.carrotsearch.randomizedtesting.RandomizedTest.randomFloat;
 
 /**
  * Test basic GeoPoint functionality.
  */
 public class GeoPointTest extends LuceneTestCase {
+  static final double DEGREES_TO_RADIANS =  Math.PI / 180;
 
   @Test
   public void testConversion() {
-    testPointRoundTrip(PlanetModel.SPHERE, 90 * DistanceUtils.DEGREES_TO_RADIANS, 0, 1e-6);
-    testPointRoundTrip(PlanetModel.SPHERE, -90 * DistanceUtils.DEGREES_TO_RADIANS, 0, 1e-6);
-    testPointRoundTrip(PlanetModel.WGS84, 90 * DistanceUtils.DEGREES_TO_RADIANS, 0, 1e-6);
-    testPointRoundTrip(PlanetModel.WGS84, -90 * DistanceUtils.DEGREES_TO_RADIANS, 0, 1e-6);
+    testPointRoundTrip(PlanetModel.SPHERE, 90 * DEGREES_TO_RADIANS, 0, 1e-6);
+    testPointRoundTrip(PlanetModel.SPHERE, -90 * DEGREES_TO_RADIANS, 0, 1e-6);
+    testPointRoundTrip(PlanetModel.WGS84, 90 * DEGREES_TO_RADIANS, 0, 1e-6);
+    testPointRoundTrip(PlanetModel.WGS84, -90 * DEGREES_TO_RADIANS, 0, 1e-6);
 
     final int times = atLeast(100);
     for (int i = 0; i < times; i++) {
-      final double pLat = (randomFloat() * 180.0 - 90.0) * DistanceUtils.DEGREES_TO_RADIANS;
-      final double pLon = (randomFloat() * 360.0 - 180.0) * DistanceUtils.DEGREES_TO_RADIANS;
+      final double pLat = (randomFloat() * 180.0 - 90.0) * DEGREES_TO_RADIANS;
+      final double pLon = (randomFloat() * 360.0 - 180.0) * DEGREES_TO_RADIANS;
       testPointRoundTrip(PlanetModel.SPHERE, pLat, pLon, 1e-6);//1e-6 since there's a square root in there (Karl says)
       testPointRoundTrip(PlanetModel.WGS84, pLat, pLon, 1e-6);
     }
@@ -60,10 +57,10 @@
   public void testSurfaceDistance() {
     final int times = atLeast(100);
     for (int i = 0; i < times; i++) {
-      final double p1Lat = (randomFloat() * 180.0 - 90.0) * DistanceUtils.DEGREES_TO_RADIANS;
-      final double p1Lon = (randomFloat() * 360.0 - 180.0) * DistanceUtils.DEGREES_TO_RADIANS;
-      final double p2Lat = (randomFloat() * 180.0 - 90.0) * DistanceUtils.DEGREES_TO_RADIANS;
-      final double p2Lon = (randomFloat() * 360.0 - 180.0) * DistanceUtils.DEGREES_TO_RADIANS;
+      final double p1Lat = (randomFloat() * 180.0 - 90.0) * DEGREES_TO_RADIANS;
+      final double p1Lon = (randomFloat() * 360.0 - 180.0) * DEGREES_TO_RADIANS;
+      final double p2Lat = (randomFloat() * 180.0 - 90.0) * DEGREES_TO_RADIANS;
+      final double p2Lon = (randomFloat() * 360.0 - 180.0) * DEGREES_TO_RADIANS;
       final GeoPoint p1 = new GeoPoint(PlanetModel.SPHERE, p1Lat, p1Lon);
       final GeoPoint p2 = new GeoPoint(PlanetModel.SPHERE, p2Lat, p2Lon);
       final double arcDistance = p1.arcDistance(p2);
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegenerateLatitudeZone.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegenerateLatitudeZone.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoDegenerateLatitudeZone.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * This GeoBBox represents an area rectangle of one specific latitude with
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoLatitudeZone.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoLatitudeZone.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoLatitudeZone.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * This GeoBBox represents an area rectangle limited only in latitude.
Index: lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoPathTest.java
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoPathTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoPathTest.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import org.junit.Test;
 
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoMembershipShape.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoMembershipShape.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoMembershipShape.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Membership shapes have capabilities of both geohashing and membership
Index: lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dRptTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dRptTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dRptTest.java	(revision )
@@ -32,13 +32,13 @@
 import org.apache.lucene.spatial.prefix.tree.SpatialPrefixTree;
 import org.apache.lucene.spatial.query.SpatialOperation;
 import org.apache.lucene.spatial.serialized.SerializedDVStrategy;
-import org.apache.lucene.geo3d.GeoBBoxFactory;
-import org.apache.lucene.geo3d.GeoStandardCircle;
-import org.apache.lucene.geo3d.GeoPath;
-import org.apache.lucene.geo3d.GeoPoint;
-import org.apache.lucene.geo3d.GeoPolygonFactory;
-import org.apache.lucene.geo3d.GeoShape;
-import org.apache.lucene.geo3d.PlanetModel;
+import org.apache.lucene.spatial3d.geom.GeoBBoxFactory;
+import org.apache.lucene.spatial3d.geom.GeoStandardCircle;
+import org.apache.lucene.spatial3d.geom.GeoPath;
+import org.apache.lucene.spatial3d.geom.GeoPoint;
+import org.apache.lucene.spatial3d.geom.GeoPolygonFactory;
+import org.apache.lucene.spatial3d.geom.GeoShape;
+import org.apache.lucene.spatial3d.geom.PlanetModel;
 import org.junit.Test;
 
 import static org.locationtech.spatial4j.distance.DistanceUtils.DEGREES_TO_RADIANS;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoAreaFactory.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoAreaFactory.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoAreaFactory.java	(revision )
@@ -14,10 +14,10 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
- * Factory for {@link org.apache.lucene.geo3d.GeoArea}.
+ * Factory for {@link GeoArea}.
  *
  * @lucene.experimental
  */
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideNorthRectangle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideNorthRectangle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoWideNorthRectangle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Bounding box wider than PI but limited on three sides (
Index: lucene/spatial3d/src/test/org/apache/lucene/geo3d/XYZSolidTest.java
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/geo3d/XYZSolidTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/XYZSolidTest.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import org.apache.lucene.util.LuceneTestCase;
 import org.junit.Test;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseDistanceShape.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseDistanceShape.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoBaseDistanceShape.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Distance shapes have capabilities of both geohashing and distance
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoPolygon.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoPolygon.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoPolygon.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * GeoPolygon interface description.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoPolygonFactory.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoPolygonFactory.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoPolygonFactory.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import java.util.ArrayList;
 import java.util.BitSet;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoCircle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoCircle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoCircle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Interface describing circular area with a center and radius.
Index: lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoConvexPolygonTest.java
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoConvexPolygonTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoConvexPolygonTest.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import org.junit.Test;
 
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/Bounds.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/Bounds.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/Bounds.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * An interface for accumulating bounds information.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoOutsideDistance.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoOutsideDistance.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoOutsideDistance.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Implemented by Geo3D shapes that can compute the distance from a point to the closest outside edge.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/NormalSquaredDistance.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/NormalSquaredDistance.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/NormalSquaredDistance.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Normal squared distance computation style.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBBoxFactory.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBBoxFactory.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoBBoxFactory.java	(revision )
@@ -14,10 +14,10 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
- * Factory for {@link org.apache.lucene.geo3d.GeoBBox}.
+ * Factory for {@link GeoBBox}.
  *
  * @lucene.experimental
  */
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegeneratePoint.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDegeneratePoint.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoDegeneratePoint.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
     
 /**
  * This class represents a degenerate point bounding box.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/XYZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/XYZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/XYZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Interface for a family of 3D rectangles, bounded on six sides by X,Y,Z limits
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/XYZSolidFactory.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/XYZSolidFactory.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/XYZSolidFactory.java	(revision )
@@ -14,10 +14,10 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
- * Factory for {@link org.apache.lucene.geo3d.XYZSolid}.
+ * Factory for {@link XYZSolid}.
  *
  * @lucene.experimental
  */
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoCircleFactory.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoCircleFactory.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoCircleFactory.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Class which constructs a GeoCircle representing an arbitrary circle.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseMembershipShape.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseMembershipShape.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoBaseMembershipShape.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Membership shapes have capabilities of both geohashing and membership
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoPath.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoPath.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoPath.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import java.util.ArrayList;
 import java.util.Collections;
Index: lucene/spatial3d/src/test/org/apache/lucene/geo3d/PlaneTest.java
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/geo3d/PlaneTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/PlaneTest.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import org.junit.Test;
 
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/dXYdZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/dXYdZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/dXYdZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * 3D rectangle, bounded on six sides by X,Y,Z limits, degenerate in X and Z.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoArea.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoArea.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoArea.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * A GeoArea represents a standard 2-D breakdown of a part of sphere.  It can
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoConvexPolygon.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoConvexPolygon.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoConvexPolygon.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import java.util.ArrayList;
 import java.util.BitSet;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoSouthRectangle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoSouthRectangle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoSouthRectangle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Bounding box limited on three sides (top lat, left lon, right lon).  The
Index: lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeSphereModelRectRelationTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeSphereModelRectRelationTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeSphereModelRectRelationTest.java	(revision )
@@ -20,14 +20,14 @@
 import java.util.List;
 
 import org.locationtech.spatial4j.shape.Rectangle;
-import org.apache.lucene.geo3d.GeoArea;
-import org.apache.lucene.geo3d.GeoBBox;
-import org.apache.lucene.geo3d.GeoBBoxFactory;
-import org.apache.lucene.geo3d.GeoStandardCircle;
-import org.apache.lucene.geo3d.GeoPoint;
-import org.apache.lucene.geo3d.GeoPolygonFactory;
-import org.apache.lucene.geo3d.GeoShape;
-import org.apache.lucene.geo3d.PlanetModel;
+import org.apache.lucene.spatial3d.geom.GeoArea;
+import org.apache.lucene.spatial3d.geom.GeoBBox;
+import org.apache.lucene.spatial3d.geom.GeoBBoxFactory;
+import org.apache.lucene.spatial3d.geom.GeoStandardCircle;
+import org.apache.lucene.spatial3d.geom.GeoPoint;
+import org.apache.lucene.spatial3d.geom.GeoPolygonFactory;
+import org.apache.lucene.spatial3d.geom.GeoShape;
+import org.apache.lucene.spatial3d.geom.PlanetModel;
 import org.junit.Test;
 
 public class Geo3dShapeSphereModelRectRelationTest extends Geo3dShapeRectRelationTestCase {
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/Plane.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/Plane.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/Plane.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * We know about three kinds of planes.  First kind: general plain through two points and origin
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/dXdYdZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/dXdYdZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/dXdYdZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * 3D rectangle, bounded on six sides by X,Y,Z limits, degenerate in all dimensions
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDistanceShape.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDistanceShape.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoDistanceShape.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Distance shapes have capabilities of both geohashing and distance
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/dXYZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/dXYZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/dXYZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * 3D rectangle, bounded on six sides by X,Y,Z limits, degenerate in X.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/Tools.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/Tools.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/Tools.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Static methods globally useful for 3d geometric work.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoNorthRectangle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoNorthRectangle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoNorthRectangle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Bounding box limited on three sides (bottom lat, left lon, right lon), including
Index: lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeRectRelationTestCase.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeRectRelationTestCase.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeRectRelationTestCase.java	(revision )
@@ -25,15 +25,15 @@
 import org.locationtech.spatial4j.shape.Circle;
 import org.locationtech.spatial4j.shape.Point;
 import org.locationtech.spatial4j.shape.RectIntersectionTestHelper;
-import org.apache.lucene.geo3d.LatLonBounds;
-import org.apache.lucene.geo3d.GeoBBox;
-import org.apache.lucene.geo3d.GeoBBoxFactory;
-import org.apache.lucene.geo3d.GeoStandardCircle;
-import org.apache.lucene.geo3d.GeoPath;
-import org.apache.lucene.geo3d.GeoPoint;
-import org.apache.lucene.geo3d.GeoPolygonFactory;
-import org.apache.lucene.geo3d.GeoShape;
-import org.apache.lucene.geo3d.PlanetModel;
+import org.apache.lucene.spatial3d.geom.LatLonBounds;
+import org.apache.lucene.spatial3d.geom.GeoBBox;
+import org.apache.lucene.spatial3d.geom.GeoBBoxFactory;
+import org.apache.lucene.spatial3d.geom.GeoStandardCircle;
+import org.apache.lucene.spatial3d.geom.GeoPath;
+import org.apache.lucene.spatial3d.geom.GeoPoint;
+import org.apache.lucene.spatial3d.geom.GeoPolygonFactory;
+import org.apache.lucene.spatial3d.geom.GeoShape;
+import org.apache.lucene.spatial3d.geom.PlanetModel;
 import org.junit.Rule;
 import org.junit.Test;
 
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoPoint.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoPoint.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoPoint.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
     
 /**
  * This class represents a point on the surface of a sphere or ellipsoid.
@@ -117,7 +117,7 @@
 
   /** Compute an arc distance between two points.
    * Note: this is an angular distance, and not a surface distance, and is therefore independent of planet model.
-   * For surface distance, see {@link org.apache.lucene.geo3d.PlanetModel#surfaceDistance(GeoPoint, GeoPoint)}
+   * For surface distance, see {@link PlanetModel#surfaceDistance(GeoPoint, GeoPoint)}
    * @param v is the second point.
    * @return the angle, in radians, between the two points.
    */
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/PointInGeo3DShapeQuery.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/PointInGeo3DShapeQuery.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/PointInGeo3DShapeQuery.java	(revision )
@@ -14,10 +14,15 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d;
 
 import java.io.IOException;
 
+import org.apache.lucene.spatial3d.geom.BasePlanetObject;
+import org.apache.lucene.spatial3d.geom.GeoArea;
+import org.apache.lucene.spatial3d.geom.GeoAreaFactory;
+import org.apache.lucene.spatial3d.geom.GeoShape;
+import org.apache.lucene.spatial3d.geom.PlanetModel;
 import org.apache.lucene.index.PointValues.IntersectVisitor;
 import org.apache.lucene.index.PointValues;
 import org.apache.lucene.index.PointValues.Relation;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/ArcDistance.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/ArcDistance.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/ArcDistance.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Arc distance computation style.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBasePolygon.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBasePolygon.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoBasePolygon.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * GeoBasePolygon objects are the base class of most GeoPolygon objects.
Index: lucene/spatial3d/src/java/overview.html
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial3d/src/java/overview.html	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/overview.html	(revision )
@@ -23,7 +23,8 @@
   <h1>The Spatial3D Module for Apache Lucene</h1>
 
   <p>
-    APIs for planar spatial3d math.
+    APIs for planar spatial3d math.  It is mostly comprised of computational geometry code in the
+    "org.apache.lucene.spatial3d.geom" package (AKA "Geo3D").
   </p>
   </body>
 </html>
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseBBox.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseBBox.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoBaseBBox.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * All bounding box shapes can derive from this base class, which furnishes
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideSouthRectangle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideSouthRectangle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoWideSouthRectangle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Bounding box wider than PI but limited on three sides (top lat,
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoCompositePolygon.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoCompositePolygon.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoCompositePolygon.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * GeoCompositePolygon is a specific implementation of GeoMembershipShape, which implements GeoPolygon explicitly.
Index: lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoPolygonTest.java
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/geo3d/GeoPolygonTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/geom/GeoPolygonTest.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import java.util.ArrayList;
 import java.util.List;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideDegenerateHorizontalLine.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideDegenerateHorizontalLine.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoWideDegenerateHorizontalLine.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Degenerate bounding box wider than PI and limited on two sides (left lon, right lon).
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/LinearDistance.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/LinearDistance.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/LinearDistance.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Linear distance computation style.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideRectangle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideRectangle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoWideRectangle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Bounding box wider than PI but limited on four sides (top lat,
Index: lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeWGS84ModelRectRelationTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeWGS84ModelRectRelationTest.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial-extras/src/test/org/apache/lucene/spatial/spatial4j/Geo3dShapeWGS84ModelRectRelationTest.java	(revision )
@@ -16,14 +16,14 @@
  */
 package org.apache.lucene.spatial.spatial4j;
 
-import org.apache.lucene.geo3d.GeoArea;
-import org.apache.lucene.geo3d.GeoBBox;
-import org.apache.lucene.geo3d.GeoBBoxFactory;
-import org.apache.lucene.geo3d.GeoCircle;
-import org.apache.lucene.geo3d.GeoStandardCircle;
-import org.apache.lucene.geo3d.GeoPath;
-import org.apache.lucene.geo3d.GeoPoint;
-import org.apache.lucene.geo3d.PlanetModel;
+import org.apache.lucene.spatial3d.geom.GeoArea;
+import org.apache.lucene.spatial3d.geom.GeoBBox;
+import org.apache.lucene.spatial3d.geom.GeoBBoxFactory;
+import org.apache.lucene.spatial3d.geom.GeoCircle;
+import org.apache.lucene.spatial3d.geom.GeoStandardCircle;
+import org.apache.lucene.spatial3d.geom.GeoPath;
+import org.apache.lucene.spatial3d.geom.GeoPoint;
+import org.apache.lucene.spatial3d.geom.PlanetModel;
 import org.junit.Test;
 
 public class Geo3dShapeWGS84ModelRectRelationTest extends Geo3dShapeRectRelationTestCase {
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDistance.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoDistance.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoDistance.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * An implementer of this interface is capable of computing the described "distance" values,
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoShape.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoShape.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoShape.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Generic shape.  This describes methods that help GeoAreas figure out
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideLongitudeSlice.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoWideLongitudeSlice.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoWideLongitudeSlice.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Bounding box wider than PI but limited on left and right sides (
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/BasePlanetObject.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/BasePlanetObject.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/BasePlanetObject.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * All Geo3D shapes can derive from this base class, which furnishes
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoSizeable.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoSizeable.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoSizeable.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Some shapes can compute radii of a geocircle in which they are inscribed.
Index: lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/package-info.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/package-info.java	(revision )
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/package-info.java	(revision )
@@ -0,0 +1,22 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Shapes implemented using 3D planar geometry.  This package has no dependencies aside from Java.
+ * This code was contributed under the name "Geo3D".
+ */
+package org.apache.lucene.spatial3d.geom;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/dXdYZSolid.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/dXdYZSolid.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/dXdYZSolid.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * 3D rectangle, bounded on six sides by X,Y,Z limits, degenerate in X and Y.
Index: lucene/spatial3d/src/test/org/apache/lucene/geo3d/TestGeo3DPoint.java
===================================================================
--- lucene/spatial3d/src/test/org/apache/lucene/geo3d/TestGeo3DPoint.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/test/org/apache/lucene/spatial3d/TestGeo3DPoint.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d;
 
 import java.io.IOException;
 import java.io.PrintWriter;
@@ -36,6 +36,16 @@
 import org.apache.lucene.document.Document;
 import org.apache.lucene.document.Field;
 import org.apache.lucene.document.NumericDocValuesField;
+import org.apache.lucene.spatial3d.geom.GeoArea;
+import org.apache.lucene.spatial3d.geom.GeoAreaFactory;
+import org.apache.lucene.spatial3d.geom.GeoBBoxFactory;
+import org.apache.lucene.spatial3d.geom.GeoCircleFactory;
+import org.apache.lucene.spatial3d.geom.GeoPath;
+import org.apache.lucene.spatial3d.geom.GeoPoint;
+import org.apache.lucene.spatial3d.geom.GeoPolygonFactory;
+import org.apache.lucene.spatial3d.geom.GeoShape;
+import org.apache.lucene.spatial3d.geom.PlanetModel;
+import org.apache.lucene.spatial3d.geom.XYZBounds;
 import org.apache.lucene.index.DirectoryReader;
 import org.apache.lucene.index.IndexReader;
 import org.apache.lucene.index.IndexWriter;
@@ -50,7 +60,6 @@
 import org.apache.lucene.search.Query;
 import org.apache.lucene.search.SimpleCollector;
 import org.apache.lucene.store.Directory;
-import org.apache.lucene.store.MockDirectoryWrapper;
 import org.apache.lucene.util.FixedBitSet;
 import org.apache.lucene.util.IOUtils;
 import org.apache.lucene.util.LuceneTestCase;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoNorthLatitudeZone.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoNorthLatitudeZone.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoNorthLatitudeZone.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * This GeoBBox represents an area rectangle limited only in south latitude.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseShape.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseShape.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoBaseShape.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * Base extended shape object.
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoCompositeMembershipShape.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoCompositeMembershipShape.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoCompositeMembershipShape.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 import java.util.ArrayList;
 import java.util.List;
Index: lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseCircle.java
===================================================================
--- lucene/spatial3d/src/java/org/apache/lucene/geo3d/GeoBaseCircle.java	(revision 9617d3d400ea4d6b3e00cf0b3a2c94598647bb4e)
+++ lucene/spatial3d/src/java/org/apache/lucene/spatial3d/geom/GeoBaseCircle.java	(revision )
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.lucene.geo3d;
+package org.apache.lucene.spatial3d.geom;
 
 /**
  * GeoCircles have all the characteristics of GeoBaseDistanceShapes, plus GeoSizeable.
