Index: lucene/test-framework/src/java/org/apache/lucene/util/TestRuleSetupAndRestoreClassEnv.java
===================================================================
--- lucene/test-framework/src/java/org/apache/lucene/util/TestRuleSetupAndRestoreClassEnv.java	(revision 1523179)
+++ lucene/test-framework/src/java/org/apache/lucene/util/TestRuleSetupAndRestoreClassEnv.java	(working copy)
@@ -253,7 +253,31 @@
           Arrays.toString(avoidCodecs.toArray()));
       throw e;
     }
+    Class target = RandomizedContext.current().getTargetClass();
+    assumeTrue("hack", 
+        target.getName().startsWith("org.apache.lucene.util.junitcompat") 
+        || target.getName().contains("TestMaxFailuresRule")
+        || testsNeededToCrash.contains(target.getSimpleName()));
   }
+  
+  
+  static final Set<String> testsNeededToCrash = new HashSet<String>() {{
+    add("TestHugeRamFile");
+    add("TestTimeLimitingCollector");
+    add("TestLazyProxSkipping");
+    add("TestPostingsOffsets");
+    add("TestRegexpQuery");
+    add("TestTryDelete");
+    add("TestDoubleBarrelLRUCache");
+    add("TestGraphTokenizers");
+    add("TestIndexWriterMerging");
+    add("TestSearchAfter");
+    add("TestNoMergeScheduler");
+    add("TestSearchForDuplicates");
+    add("TestBytesRef");
+    add("TestAttributeSource");
+    add("TestIndexWriterMergePolicy");
+  }};
 
   /**
    * Check codec restrictions.
