Index: lucene/build.xml
===================================================================
--- lucene/build.xml	(revision 1674704)
+++ lucene/build.xml	(working copy)
@@ -53,7 +53,7 @@
   </target>
 
   <!-- "-clover.load" is *not* a useless dependency. do not remove -->
-  <target name="test" depends="-clover.load, -init-totals, test-core, test-modules, -check-totals"
+  <target name="test" depends="-clover.load, -init-totals, test-core, test-test-framework, test-modules, -check-totals"
           description="Runs all unit tests (core, modules and back-compat)"
   />
 
@@ -449,9 +449,18 @@
       <ant dir="core" target="compile-test" inheritall="false">
         <propertyset refid="uptodate.and.compiled.properties"/>
       </ant>
+      <ant dir="test-framework" target="compile-test" inheritall="false">
+        <propertyset refid="uptodate.and.compiled.properties"/>
+      </ant>
       <modules-crawl target="compile-test" failonerror="true"/>
     </sequential>
   </target>
+
+  <target name="test-test-framework">
+      <ant dir="test-framework" target="test" inheritall="false">
+        <propertyset refid="uptodate.and.compiled.properties"/>
+      </ant>
+  </target>
   
   <target name="test-modules">
     <modules-crawl target="test" failonerror="true"/>
@@ -470,6 +479,9 @@
     <ant dir="${common.dir}/core" target="jacoco" inheritAll="false">
       <propertyset refid="uptodate.and.compiled.properties"/>
     </ant>
+    <ant dir="${common.dir}/test-framework" target="jacoco" inheritAll="false">
+      <propertyset refid="uptodate.and.compiled.properties"/>
+    </ant>
     <modules-crawl target="jacoco" failonerror="true"/>
 
     <!-- produce aggregate report -->
@@ -486,7 +498,6 @@
         <classfiles>
           <fileset dir="${common.dir}/build">
              <include name="**/classes/java/**/*.class"/>
-             <exclude name="test-framework/**"/>
              <exclude name="tools/**"/>
           </fileset>
         </classfiles>
Index: lucene/core/src/test/org/apache/lucene/TestWorstCaseTestBehavior.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/analysis/TestGraphTokenizers.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/analysis/TestLookaheadTokenFilter.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/analysis/TestMockAnalyzer.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/analysis/TestMockCharFilter.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/analysis/TestPosition.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/analysis/TrivialLookaheadFilter.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/codecs/asserting/TestAssertingDocValuesFormat.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/codecs/asserting/TestAssertingNormsFormat.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/codecs/asserting/TestAssertingPostingsFormat.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/codecs/asserting/TestAssertingStoredFieldsFormat.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/codecs/asserting/TestAssertingTermVectorsFormat.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/codecs/compressing/TestCompressingStoredFieldsFormat.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/codecs/compressing/TestCompressingTermVectorsFormat.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/mockfile/TestMockFilesystems.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/store/TestDirectory.java
===================================================================
--- lucene/core/src/test/org/apache/lucene/store/TestDirectory.java	(revision 1674704)
+++ lucene/core/src/test/org/apache/lucene/store/TestDirectory.java	(working copy)
@@ -24,30 +24,10 @@
 import java.util.List;
 
 import org.apache.lucene.util.IOUtils;
+import org.apache.lucene.util.LuceneTestCase;
 
-public class TestDirectory extends BaseDirectoryTestCase {
+public class TestDirectory extends LuceneTestCase {
 
-  @Override
-  protected Directory getDirectory(Path path) throws IOException {
-    final Directory dir;
-    if (random().nextBoolean()) {
-      dir = newDirectory();
-    } else {
-      dir = newFSDirectory(path);
-    }
-    if (dir instanceof MockDirectoryWrapper) {
-      // test manipulates directory directly
-      ((MockDirectoryWrapper)dir).setEnableVirusScanner(false);
-    }
-    return dir;
-  }
-
-  // we wrap the directory in slow stuff, so only run nightly
-  @Override @Nightly
-  public void testThreadSafety() throws Exception {
-    super.testThreadSafety();
-  }
-
   // Test that different instances of FSDirectory can coexist on the same
   // path, can read, write, and lock files.
   public void testDirectInstantiation() throws Exception {
Index: lucene/core/src/test/org/apache/lucene/store/TestMockDirectoryWrapper.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/TestMaxFailuresRule.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/TestRamUsageEstimatorOnWildAnimals.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/SorePoint.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/SoreType.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestBeforeAfterOverrides.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestCodecReported.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestExceptionInBeforeClassHooks.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestFailIfDirectoryNotClosed.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestFailIfUnreferencedFiles.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestGroupFiltering.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestJUnitRuleOrder.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestLeaveFilesIfTestFails.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestReproduceMessage.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestReproduceMessageWithRepeated.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestSeedFromUncaught.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/TestSetupTeardownChaining.java (deleted)
===================================================================
Index: lucene/core/src/test/org/apache/lucene/util/junitcompat/WithNestedTests.java (deleted)
===================================================================
Index: lucene/test-framework/build.xml
===================================================================
--- lucene/test-framework/build.xml	(revision 1674704)
+++ lucene/test-framework/build.xml	(working copy)
@@ -35,8 +35,18 @@
     <fileset dir="lib"/>
   </path>
 
-  <path id="test.classpath"/>
+  <path id="test.classpath"> 
+    <pathelement location="${build.dir}/classes/java"/>
+    <pathelement location="${build.dir}/classes/test"/>
+    <path refid="classpath"/>
+    <path refid="junit-path"/>
+  </path>
 
+  <path id="junit.classpath">
+    <path refid="test.classpath"/>
+    <pathelement path="${java.class.path}"/>
+  </path>
+
   <!-- 
       Specialize compile-core to depend on lucene-core and lucene-codecs compilation.
    -->
@@ -43,12 +53,6 @@
   <target name="compile-core" depends="init,compile-lucene-core,compile-codecs,common.compile-core"
           description="Compiles test-framework classes"/>
 
-  <!-- redefine the clover setup, because we dont want to run clover for the test-framework -->
-  <target name="-clover.setup" if="run.clover"/>
-
-  <!-- redefine the test compilation, so it's just a no-op -->
-  <target name="compile-test"/>
-  
   <!-- redefine the forbidden apis for tests, as we check ourselves - no sysout testing -->
   <target name="-check-forbidden-tests" depends="-init-forbidden-apis,compile-core">
     <forbidden-apis signaturesFile="${common.dir}/tools/forbiddenApis/tests.txt" classpathref="forbidden-apis.allclasses.classpath"> 
Index: lucene/test-framework/src/test/org/apache/lucene/codecs/asserting/TestAssertingDocValuesFormat.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/codecs/asserting/TestAssertingDocValuesFormat.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/codecs/asserting/TestAssertingDocValuesFormat.java	(working copy)
@@ -18,10 +18,10 @@
  */
 
 import org.apache.lucene.codecs.Codec;
-import org.apache.lucene.index.BasePostingsFormatTestCase;
+import org.apache.lucene.index.BaseDocValuesFormatTestCase;
 
-/** Test AssertingPostingsFormat directly */
-public class TestAssertingDocValuesFormat extends BasePostingsFormatTestCase {
+/** Test AssertingDocValuesFormat directly */
+public class TestAssertingDocValuesFormat extends BaseDocValuesFormatTestCase {
   private final Codec codec = new AssertingCodec();
   
   @Override
@@ -28,9 +28,4 @@
   protected Codec getCodec() {
     return codec;
   }
-
-  @Override
-  protected boolean isPostingsEnumReuseImplemented() {
-    return false;
-  }
 }
Index: lucene/test-framework/src/test/org/apache/lucene/codecs/asserting/TestAssertingTermVectorsFormat.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/codecs/asserting/TestAssertingTermVectorsFormat.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/codecs/asserting/TestAssertingTermVectorsFormat.java	(working copy)
@@ -18,10 +18,10 @@
  */
 
 import org.apache.lucene.codecs.Codec;
-import org.apache.lucene.index.BaseStoredFieldsFormatTestCase;
+import org.apache.lucene.index.BaseTermVectorsFormatTestCase;
 
-/** Test AssertingStoredFieldsFormat directly */
-public class TestAssertingTermVectorsFormat extends BaseStoredFieldsFormatTestCase {
+/** Test AssertingTermVectorsFormat directly */
+public class TestAssertingTermVectorsFormat extends BaseTermVectorsFormatTestCase {
   private final Codec codec = new AssertingCodec();
   
   @Override
Index: lucene/test-framework/src/test/org/apache/lucene/store/TestMockDirectoryWrapper.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/store/TestMockDirectoryWrapper.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/store/TestMockDirectoryWrapper.java	(working copy)
@@ -18,6 +18,7 @@
  */
 
 import java.io.IOException;
+import java.nio.file.Path;
 
 import org.apache.lucene.document.Document;
 import org.apache.lucene.index.IndexWriter;
@@ -24,10 +25,28 @@
 import org.apache.lucene.index.IndexWriterConfig;
 import org.apache.lucene.index.RandomIndexWriter;
 import org.apache.lucene.util.IOUtils;
-import org.apache.lucene.util.LuceneTestCase;
+import org.apache.lucene.util.LuceneTestCase.Nightly;
 
-public class TestMockDirectoryWrapper extends LuceneTestCase {
+public class TestMockDirectoryWrapper extends BaseDirectoryTestCase {
   
+  @Override
+  protected Directory getDirectory(Path path) throws IOException {
+    final MockDirectoryWrapper dir;
+    if (random().nextBoolean()) {
+      dir = newMockDirectory();
+    } else {
+      dir = newMockFSDirectory(path);
+    }
+    dir.setEnableVirusScanner(false); // test manipulates filesystem directly
+    return dir;
+  }
+  
+  // we wrap the directory in slow stuff, so only run nightly
+  @Override @Nightly
+  public void testThreadSafety() throws Exception {
+    super.testThreadSafety();
+  }
+  
   public void testFailIfIndexWriterNotClosed() throws IOException {
     MockDirectoryWrapper dir = newMockDirectory();
     IndexWriter iw = new IndexWriter(dir, new IndexWriterConfig(null));
@@ -103,6 +122,5 @@
     iw.commit();
     iw.close();
     dir.close();
-  }
-  
+  }  
 }
Index: lucene/test-framework/src/test/org/apache/lucene/util/SorePoint.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/SorePoint.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/SorePoint.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/SoreType.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/SoreType.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/SoreType.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestBeforeAfterOverrides.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestBeforeAfterOverrides.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestBeforeAfterOverrides.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 import org.junit.After;
 import org.junit.Assert;
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestCodecReported.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestCodecReported.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestCodecReported.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 import org.apache.lucene.codecs.Codec;
 import org.junit.Assert;
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestExceptionInBeforeClassHooks.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestExceptionInBeforeClassHooks.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestExceptionInBeforeClassHooks.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestFailIfDirectoryNotClosed.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestFailIfDirectoryNotClosed.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestFailIfDirectoryNotClosed.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestFailIfUnreferencedFiles.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestFailIfUnreferencedFiles.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestFailIfUnreferencedFiles.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestGroupFiltering.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestGroupFiltering.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestGroupFiltering.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 import java.lang.annotation.Documented;
 import java.lang.annotation.Inherited;
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestJUnitRuleOrder.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestJUnitRuleOrder.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestJUnitRuleOrder.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestLeaveFilesIfTestFails.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestLeaveFilesIfTestFails.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestLeaveFilesIfTestFails.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestMaxFailuresRule.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestMaxFailuresRule.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestMaxFailuresRule.java	(working copy)
@@ -19,7 +19,7 @@
 
 import java.util.concurrent.CountDownLatch;
 
-import org.apache.lucene.util.junitcompat.WithNestedTests;
+import org.apache.lucene.util.WithNestedTests;
 import org.junit.Assert;
 import org.junit.BeforeClass;
 import org.junit.Test;
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestRamUsageTesterOnWildAnimals.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestRamUsageTesterOnWildAnimals.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestRamUsageTesterOnWildAnimals.java	(working copy)
@@ -22,7 +22,7 @@
 /**
  * Check large and special graphs. 
  */
-public class TestRamUsageEstimatorOnWildAnimals extends LuceneTestCase {
+public class TestRamUsageTesterOnWildAnimals extends LuceneTestCase {
   public static class ListElement {
     ListElement next;
   }
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestReproduceMessage.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestReproduceMessage.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestReproduceMessage.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestReproduceMessageWithRepeated.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestReproduceMessageWithRepeated.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestReproduceMessageWithRepeated.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestSeedFromUncaught.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestSeedFromUncaught.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestSeedFromUncaught.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestSetupTeardownChaining.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestSetupTeardownChaining.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestSetupTeardownChaining.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
Index: lucene/test-framework/src/test/org/apache/lucene/util/TestWorstCaseTestBehavior.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/TestWorstCaseTestBehavior.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/TestWorstCaseTestBehavior.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene;
+package org.apache.lucene.util;
 
 import org.apache.lucene.util.LuceneTestCase;
 import org.junit.Ignore;
Index: lucene/test-framework/src/test/org/apache/lucene/util/WithNestedTests.java
===================================================================
--- lucene/test-framework/src/test/org/apache/lucene/util/WithNestedTests.java	(revision 1674704)
+++ lucene/test-framework/src/test/org/apache/lucene/util/WithNestedTests.java	(working copy)
@@ -1,4 +1,4 @@
-package org.apache.lucene.util.junitcompat;
+package org.apache.lucene.util;
 
 /*
  * Licensed to the Apache Software Foundation (ASF) under one or more
