From 60c7c65afd9e504471c3253af3be5fcd7aa16d83 Mon Sep 17 00:00:00 2001
From: Florian Diebold <florian.diebold@freiheit.com>
Date: Tue, 4 Jun 2019 16:18:56 +0200
Subject: [PATCH] Fix missing parenthesis in MultiBoolFunction

---
 .../lucene/queries/function/valuesource/MultiBoolFunction.java  | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/lucene/queries/src/java/org/apache/lucene/queries/function/valuesource/MultiBoolFunction.java b/lucene/queries/src/java/org/apache/lucene/queries/function/valuesource/MultiBoolFunction.java
index f368b4a70d..3011cdb485 100644
--- a/lucene/queries/src/java/org/apache/lucene/queries/function/valuesource/MultiBoolFunction.java
+++ b/lucene/queries/src/java/org/apache/lucene/queries/function/valuesource/MultiBoolFunction.java
@@ -68,6 +68,7 @@ public abstract class MultiBoolFunction extends BoolFunction {
           }
           sb.append(dv.toString(doc));
         }
+        sb.append(')');
         return sb.toString();
       }
     };
@@ -86,6 +87,7 @@ public abstract class MultiBoolFunction extends BoolFunction {
       }
       sb.append(source.description());
     }
+    sb.append(')');
     return sb.toString();
   }
 
-- 
2.20.1

