Index: src/java/org/apache/lucene/index/SegmentInfos.java
===================================================================
--- src/java/org/apache/lucene/index/SegmentInfos.java	(revision 637041)
+++ src/java/org/apache/lucene/index/SegmentInfos.java	(working copy)
@@ -70,6 +70,7 @@
    * starting with the current time in milliseconds forces to create unique version numbers.
    */
   private long version = System.currentTimeMillis();
+  private long firstVersion = version;
 
   private long generation = 0;     // generation of the "segments_N" for the next commit
   private long lastGeneration = 0; // generation of the "segments_N" file we last successfully read
@@ -216,6 +217,7 @@
         if (format < CURRENT_FORMAT)
           throw new CorruptIndexException("Unknown format version: " + format);
         version = input.readLong(); // read version
+        firstVersion = version - generation;
         counter = input.readInt(); // read counter
       }
       else{     // file is in old format without explicit format info
@@ -249,6 +251,7 @@
         clear();
       }
     }
+    compareVersionAndGeneration("read",segmentFileName);
   }
 
   /**
@@ -341,8 +344,23 @@
     }
     
     lastGeneration = generation;
+    
+    compareVersionAndGeneration("write",segmentFileName);
   }
 
+  private void compareVersionAndGeneration(String f, String p) {
+    long currGeneraion = this.generation;
+    long currVersionGap = this.version - this.firstVersion;
+    boolean same = (currGeneraion == currVersionGap);
+    String msg = "At "+f+" ("+p+") ver="+this.version+
+        " verGap="+currVersionGap+" gen="+currGeneraion+
+        "  "+(same ? "OK" : " WRONG !!!!!!!!!!!!!!");
+    if (!same) {
+      System.out.println(msg);
+      throw new RuntimeException(msg);
+    }
+  }
+
   /**
    * Returns a copy of this instance, also copying each
    * SegmentInfo.
