Index: lucene/core/src/test/org/apache/lucene/index/TestIndexWriterOnDiskFull.java
===================================================================
--- lucene/core/src/test/org/apache/lucene/index/TestIndexWriterOnDiskFull.java	(revision 1495833)
+++ lucene/core/src/test/org/apache/lucene/index/TestIndexWriterOnDiskFull.java	(working copy)
@@ -462,7 +462,7 @@
 
     @Override
     public void eval(MockDirectoryWrapper dir)  throws IOException {
-      if (!doFail) {
+      if (!doFail || (didFail1 && didFail2)) {
         return;
       }
       StackTraceElement[] trace = new Exception().getStackTrace();
