Index: contrib/spatial/src/java/org/apache/lucene/spatial/tier/DistanceHandler.java
===================================================================
--- contrib/spatial/src/java/org/apache/lucene/spatial/tier/DistanceHandler.java	(revision 813182)
+++ contrib/spatial/src/java/org/apache/lucene/spatial/tier/DistanceHandler.java	(working copy)
@@ -34,9 +34,11 @@
   private Map<Integer,Double> distances;
   public enum Precision {EXACT, TWOFEET, TWENTYFEET, TWOHUNDREDFEET};
   private Precision precise;
+  private Map<String, Double> distanceLookupCache;
   
-  public DistanceHandler (Map<Integer,Double> distances, Precision precise){
+  public DistanceHandler (Map<Integer,Double> distances, Map<String, Double> distanceLookupCache, Precision precise){
     this.distances = distances;
+    this.distanceLookupCache = distanceLookupCache;
     this.precise = precise; 
     
   }
@@ -84,7 +86,7 @@
       
       String k = new Double(xLat).toString() +","+ new Double(xLng).toString();
     
-      Double d = (distances.get(k));
+      Double d = (distanceLookupCache.get(k));
       if (d != null){
         return d.doubleValue();
       }
@@ -96,7 +98,7 @@
   
   
   public static void main(String args[]){ 
-    DistanceHandler db = new DistanceHandler(new HashMap(), Precision.TWOHUNDREDFEET);
+    DistanceHandler db = new DistanceHandler(new HashMap(), new HashMap(), Precision.TWOHUNDREDFEET);
     System.out.println(DistanceHandler.getPrecision(-1234.123456789, db.getPrecision()));
   }
 }

