Index: lucene/facet/src/java/org/apache/lucene/facet/search/TotalFacetCounts.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/facet/search/TotalFacetCounts.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/facet/search/TotalFacetCounts.java	(working copy)
@@ -71,7 +71,6 @@
   
   /** 
    * Construct by key - from index Directory or by recomputing.
-   * @param key the key mapping of this total facet counts (index, taxonomy, category lists...) 
    */
   private TotalFacetCounts (TaxonomyReader taxonomy, FacetIndexingParams facetIndexingParams,
       int[][] counts, CreationType createType4Test) throws IOException, LockObtainFailedException {
Index: lucene/facet/src/java/org/apache/lucene/facet/search/StandardFacetsAccumulator.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/facet/search/StandardFacetsAccumulator.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/facet/search/StandardFacetsAccumulator.java	(working copy)
@@ -230,9 +230,9 @@
   /**
    * Iterate over the documents for this partition and fill the facet arrays with the correct
    * count/complement count/value.
-   * @param internalCollector
+   * @param docids
    * @param facetArrays
-   * @param part
+   * @param partition
    * @throws IOException
    */
   private final void fillArraysForPartition(ScoredDocIDs docids,
Index: lucene/facet/src/java/org/apache/lucene/facet/taxonomy/writercache/cl2o/CompactLabelToOrdinal.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/facet/taxonomy/writercache/cl2o/CompactLabelToOrdinal.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/facet/taxonomy/writercache/cl2o/CompactLabelToOrdinal.java	(working copy)
@@ -449,7 +449,7 @@
 
   /**
    * Opens the file and reloads the CompactLabelToOrdinal. The file it expects
-   * is generated from the {@link #flush()} command.
+   * is generated from the {@link #flush(File)} command.
    */
   static CompactLabelToOrdinal open(File file, float loadFactor,
                                     int numHashArrays) throws IOException {
Index: lucene/facet/src/java/org/apache/lucene/facet/taxonomy/writercache/lru/NameIntCacheLRU.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/facet/taxonomy/writercache/lru/NameIntCacheLRU.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/facet/taxonomy/writercache/lru/NameIntCacheLRU.java	(working copy)
@@ -4,6 +4,7 @@
 import java.util.Iterator;
 import java.util.LinkedHashMap;
 import org.apache.lucene.facet.taxonomy.CategoryPath;
+import org.apache.lucene.facet.taxonomy.directory.DirectoryTaxonomyWriter; // javadocs
 
 /**
  * Licensed to the Apache Software Foundation (ASF) under one or more
@@ -119,7 +120,7 @@
    * If cache is full remove least recently used entries from cache.
    * Return true if anything was removed, false otherwise.
    * 
-   * See comment in {@link DirectoryTaxonomyWriter#addToCache(String, Integer)}
+   * See comment in {@link DirectoryTaxonomyWriter#addToCache(CategoryPath, int)}
    * for an explanation why we clean 2/3rds of the cache, and not just one
    * entry.
    */ 
Index: lucene/facet/src/java/org/apache/lucene/util/collections/ArrayHashMap.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/util/collections/ArrayHashMap.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/util/collections/ArrayHashMap.java	(working copy)
@@ -219,7 +219,7 @@
 
   /**
    * Adds a pair to the map. Takes the first empty position from the
-   * empty-linked-list's head - {@link firstEmpty}. New pairs are always
+   * empty-linked-list's head - {@link #firstEmpty}. New pairs are always
    * inserted to baseHash, and are followed by the old colliding pair.
    */
   private void prvt_put(K key, V value) {
@@ -312,7 +312,7 @@
 
   /**
    * Finds the actual index of a given key with it's baseHashIndex. Some methods
-   * use the baseHashIndex. If those call {@link #find()} there's no need to
+   * use the baseHashIndex. If those call {@link #find} there's no need to
    * re-calculate that hash.
    * 
    * @return the index of the given key, or 0 if the key wasn't found.
Index: lucene/facet/src/java/org/apache/lucene/util/collections/IntToObjectMap.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/util/collections/IntToObjectMap.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/util/collections/IntToObjectMap.java	(working copy)
@@ -242,7 +242,7 @@
 
   /**
    * Adds a pair to the map. Takes the first empty position from the
-   * empty-linked-list's head - {@link firstEmpty}.
+   * empty-linked-list's head - {@link #firstEmpty}.
    * 
    * New pairs are always inserted to baseHash, and are followed by the old
    * colliding pair.
@@ -367,7 +367,7 @@
 
   /**
    * Find the actual index of a given key with it's baseHashIndex.<br>
-   * Some methods use the baseHashIndex. If those call {@link #find()} there's
+   * Some methods use the baseHashIndex. If those call {@link #find} there's
    * no need to re-calculate that hash.
    * 
    * @param key
Index: lucene/facet/src/java/org/apache/lucene/util/collections/ObjectToIntMap.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/util/collections/ObjectToIntMap.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/util/collections/ObjectToIntMap.java	(working copy)
@@ -243,7 +243,7 @@
 
   /**
    * Adds a pair to the map. Takes the first empty position from the
-   * empty-linked-list's head - {@link firstEmpty}.
+   * empty-linked-list's head - {@link #firstEmpty}.
    * 
    * New pairs are always inserted to baseHash, and are followed by the old
    * colliding pair.
@@ -369,7 +369,7 @@
 
   /**
    * Find the actual index of a given key with it's baseHashIndex.<br>
-   * Some methods use the baseHashIndex. If those call {@link #find()} there's
+   * Some methods use the baseHashIndex. If those call {@link #find} there's
    * no need to re-calculate that hash.
    * 
    * @param key
Index: lucene/facet/src/java/org/apache/lucene/util/collections/IntToDoubleMap.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/util/collections/IntToDoubleMap.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/util/collections/IntToDoubleMap.java	(working copy)
@@ -243,7 +243,7 @@
 
   /**
    * Adds a pair to the map. Takes the first empty position from the
-   * empty-linked-list's head - {@link firstEmpty}.
+   * empty-linked-list's head - {@link #firstEmpty}.
    * 
    * New pairs are always inserted to baseHash, and are followed by the old
    * colliding pair.
@@ -368,7 +368,7 @@
 
   /**
    * Find the actual index of a given key with it's baseHashIndex.<br>
-   * Some methods use the baseHashIndex. If those call {@link #find()} there's
+   * Some methods use the baseHashIndex. If those call {@link #find} there's
    * no need to re-calculate that hash.
    * 
    * @param key
Index: lucene/facet/src/java/org/apache/lucene/util/collections/IntToIntMap.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/util/collections/IntToIntMap.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/util/collections/IntToIntMap.java	(working copy)
@@ -241,7 +241,7 @@
 
   /**
    * Adds a pair to the map. Takes the first empty position from the
-   * empty-linked-list's head - {@link firstEmpty}.
+   * empty-linked-list's head - {@link #firstEmpty}.
    * 
    * New pairs are always inserted to baseHash, and are followed by the old
    * colliding pair.
@@ -365,7 +365,7 @@
 
   /**
    * Find the actual index of a given key with it's baseHashIndex.<br>
-   * Some methods use the baseHashIndex. If those call {@link #find()} there's
+   * Some methods use the baseHashIndex. If those call {@link #find} there's
    * no need to re-calculate that hash.
    * 
    * @param key
Index: lucene/facet/src/java/org/apache/lucene/util/collections/IntHashSet.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/util/collections/IntHashSet.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/util/collections/IntHashSet.java	(working copy)
@@ -206,15 +206,13 @@
 
   /**
    * Adds a pair to the map. Takes the first empty position from the
-   * empty-linked-list's head - {@link firstEmpty}.
+   * empty-linked-list's head - {@link #firstEmpty}.
    * 
    * New pairs are always inserted to baseHash, and are followed by the old
    * colliding pair.
    * 
    * @param key
    *            integer which maps the given value
-   * @param e
-   *            value which is being mapped using the given key
    */
   private void prvt_add(int key) {
     // Hash entry to which the new pair would be inserted
@@ -312,7 +310,7 @@
 
   /**
    * Find the actual index of a given key with it's baseHashIndex.<br>
-   * Some methods use the baseHashIndex. If those call {@link #find()} there's
+   * Some methods use the baseHashIndex. If those call {@link #find} there's
    * no need to re-calculate that hash.
    * 
    * @param key
Index: lucene/facet/src/java/org/apache/lucene/util/collections/FloatToObjectMap.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/util/collections/FloatToObjectMap.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/util/collections/FloatToObjectMap.java	(working copy)
@@ -243,7 +243,7 @@
 
   /**
    * Adds a pair to the map. Takes the first empty position from the
-   * empty-linked-list's head - {@link firstEmpty}.
+   * empty-linked-list's head - {@link #firstEmpty}.
    * 
    * New pairs are always inserted to baseHash, and are followed by the old
    * colliding pair.
@@ -367,7 +367,7 @@
 
   /**
    * Find the actual index of a given key with it's baseHashIndex.<br>
-   * Some methods use the baseHashIndex. If those call {@link #find()} there's
+   * Some methods use the baseHashIndex. If those call {@link #find} there's
    * no need to re-calculate that hash.
    * 
    * @param key
Index: lucene/facet/src/java/org/apache/lucene/util/collections/ObjectToFloatMap.java
===================================================================
--- lucene/facet/src/java/org/apache/lucene/util/collections/ObjectToFloatMap.java	(revision 1329012)
+++ lucene/facet/src/java/org/apache/lucene/util/collections/ObjectToFloatMap.java	(working copy)
@@ -243,7 +243,7 @@
 
   /**
    * Adds a pair to the map. Takes the first empty position from the
-   * empty-linked-list's head - {@link firstEmpty}.
+   * empty-linked-list's head - {@link #firstEmpty}.
    * 
    * New pairs are always inserted to baseHash, and are followed by the old
    * colliding pair.
@@ -369,7 +369,7 @@
 
   /**
    * Find the actual index of a given key with it's baseHashIndex.<br>
-   * Some methods use the baseHashIndex. If those call {@link #find()} there's
+   * Some methods use the baseHashIndex. If those call {@link #find} there's
    * no need to re-calculate that hash.
    * 
    * @param key
Index: lucene/analysis/phonetic/src/java/org/apache/lucene/analysis/phonetic/BeiderMorseFilter.java
===================================================================
--- lucene/analysis/phonetic/src/java/org/apache/lucene/analysis/phonetic/BeiderMorseFilter.java	(revision 1329012)
+++ lucene/analysis/phonetic/src/java/org/apache/lucene/analysis/phonetic/BeiderMorseFilter.java	(working copy)
@@ -63,7 +63,7 @@
   private final OffsetAttribute offsetAtt = addAttribute(OffsetAttribute.class);
   
   /** 
-   * Calls {@link #BeiderMorseFilter(TokenStream, PhoneticEngine, Languages.LanguageSet) 
+   * Calls {@link #BeiderMorseFilter(TokenStream, PhoneticEngine, LanguageSet)
    *        BeiderMorseFilter(input, engine, null)}
    */
   public BeiderMorseFilter(TokenStream input, PhoneticEngine engine) {
Index: lucene/memory/src/java/org/apache/lucene/index/memory/MemoryIndex.java
===================================================================
--- lucene/memory/src/java/org/apache/lucene/index/memory/MemoryIndex.java	(revision 1329012)
+++ lucene/memory/src/java/org/apache/lucene/index/memory/MemoryIndex.java	(working copy)
@@ -194,7 +194,7 @@
   /** pos: positions[3*i], startOffset: positions[3*i +1], endOffset: positions[3*i +2] */
   private final int stride;
   
-  /** Could be made configurable; See {@link Document#setBoost(float)} */
+  /** Could be made configurable; */
   private static final float docBoost = 1.0f;
 
   private static final boolean DEBUG = false;
