Index: ISOLatin1AccentFilter.java
===================================================================
--- ISOLatin1AccentFilter.java	(revision 759591)
+++ ISOLatin1AccentFilter.java	(working copy)
@@ -269,6 +269,33 @@
             output[outputPos++] = 's';
             output[outputPos++] = 't';
           break;
+		  
+		//TR
+		case '\u011E': //Ğ
+			output[outputPos++] = 'G';
+					break;
+
+		case '\u011F': //ğ
+			output[outputPos++] = 'g';
+					break;
+
+		case '\u0130': //İ
+			output[outputPos++] = 'I';
+					break;
+
+		case '\u0131': //ı
+			output[outputPos++] = 'i';
+					break;
+
+		case '\u015E': //Ş
+			output[outputPos++] = 'S';
+					break;
+
+		case '\u015F': //ş
+			output[outputPos++] = 's';
+					break;
+		//END-TR
+						
         default :
           output[outputPos++] = c;
           break;
