commit b1f5e734130470521374206da2f9a040f2df27f1
Author: MathBunny <horatiulazu@gmail.com>
Date:   Thu Feb 1 19:18:36 2018 -0500

    LUCENE-8151: Use advanceExact removing redundent conditionals

diff --git a/lucene/join/src/java/org/apache/lucene/search/join/JoinUtil.java b/lucene/join/src/java/org/apache/lucene/search/join/JoinUtil.java
index d42cd2d..00b19ed 100644
--- a/lucene/join/src/java/org/apache/lucene/search/join/JoinUtil.java
+++ b/lucene/join/src/java/org/apache/lucene/search/join/JoinUtil.java
@@ -199,10 +199,7 @@ public final class JoinUtil {
 
         @Override
         public void collect(int doc) throws IOException {
-          if (doc > sortedNumericDocValues.docID()) {
-            sortedNumericDocValues.advance(doc);
-          }
-          if (doc == sortedNumericDocValues.docID()) {
+          if (sortedNumericDocValues.advanceExact(doc)) {
             for (int i = 0; i < sortedNumericDocValues.docValueCount(); i++) {
               long value = sortedNumericDocValues.nextValue();
               joinValues.add(value);
